<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.34">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Probability, inference, and computation: examples ✏️ – Climate Risk Assessment and Management</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../notebooks/extremes-examples.html" rel="next">
<link href="../notebooks/index.html" rel="prev">
<script src="../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-c8ad9e5dbd60b7b70b38521ab19b7da4.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-ff4a0bb939fed34ae4231fc9111ce8ef.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<link href="../site_libs/quarto-contrib/fontawesome6-1.2.0/all.min.css" rel="stylesheet">
<link href="../site_libs/quarto-contrib/fontawesome6-1.2.0/latex-fontsize.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<meta name="citation_title" content="Climate Risk Assessment and Management">
<meta name="citation_author" content="James Doss-Gollin">
<meta name="citation_publication_date" content="2024-12-31">
<meta name="citation_cover_date" content="2024-12-31">
<meta name="citation_year" content="2024">
<meta name="citation_language" content="en">
<meta name="citation_reference" content="citation_title=An Introduction to Earth and Environmental Data Science;,citation_abstract=History: This book grew out of a course developed at Columbia University called Research Computing in Earth Science. It was written mostly by Ryan Abernathey, with significant contributions from Ke...;,citation_author=Ryan Abernathey;,citation_publication_date=2024;,citation_cover_date=2024;,citation_year=2024;,citation_fulltext_html_url=https://earth-env-data-science.github.io/intro.html;">
<meta name="citation_reference" content="citation_title=Differential effects of climate change on average and peak demand for heating and cooling across the contiguous USA;,citation_abstract=While most electricity systems are designed to handle peak demand during summer months, long-term energy pathways consistent with deep decarbonization generally electrify building heating, thus increasing electricity demand during winter. A key question is how climate variability and change will affect peak heating and cooling demand in an electrified future. We conduct a spatially explicit analysis of trends in temperature-based proxies of electricity demand over the past 70 years. Average annual demand for heating (cooling) decreases (increases) over most of the contiguous US. However, while climate change drives robust increases in peak cooling demand, trends in peak heating demand are generally smaller and less robust. Because the distribution of temperature exhibits a long left tail, severe cold snaps dominate the extremes of thermal demand. As building heating electrifies, system operators must account for these events to ensure reliability.;,citation_author=Yash Amonkar;,citation_author=James Doss-Gollin;,citation_author=David J. Farnham;,citation_author=Vijay Modi;,citation_author=Upmanu Lall;,citation_publication_date=2023-11-01;,citation_cover_date=2023-11-01;,citation_year=2023;,citation_fulltext_html_url=https://www.nature.com/articles/s43247-023-01048-1;,citation_issue=1, 1;,citation_doi=10.1038/s43247-023-01048-1;,citation_issn=2662-4435;,citation_volume=4;,citation_journal_title=Communications Earth &amp;amp;amp; Environment;,citation_journal_abbrev=Commun Earth Environ;,citation_publisher=Nature Publishing Group;">
<meta name="citation_reference" content="citation_title=Risk Analysis in the Earth Sciences;,citation_abstract=A free e-textbook that teaches the Earth science and statistical concepts needed to assess climate-related risks, using the R programming environment.;,citation_author=Patrick Applegate;,citation_author=Klaus Keller;,citation_publication_date=2015-10-26;,citation_cover_date=2015-10-26;,citation_year=2015;,citation_fulltext_html_url=https://leanpub.next/raes;">
<meta name="citation_reference" content="citation_title=Climate risk assessment needs urgent improvement;,citation_abstract=Existing constraints in current climate risk assessments make them inappropriate to effectively assess the true exposure of society and businesses to climate-related risk. Using the key constraints to guide a conceptual framework, we identify four cross-cutting and inter-related critical paths for improvement.;,citation_author=Alberto Arribas;,citation_author=Ross Fairgrieve;,citation_author=Trevor Dhu;,citation_author=Juliet Bell;,citation_author=Rosalind Cornforth;,citation_author=Geoff Gooley;,citation_author=Chris J. Hilson;,citation_author=Amy Luers;,citation_author=Theodore G. Shepherd;,citation_author=Roger Street;,citation_author=Nick Wood;,citation_publication_date=2022-08-25;,citation_cover_date=2022-08-25;,citation_year=2022;,citation_fulltext_html_url=https://www.nature.com/articles/s41467-022-31979-w;,citation_issue=1, 1;,citation_doi=10.1038/s41467-022-31979-w;,citation_issn=2041-1723;,citation_volume=13;,citation_journal_title=Nature Communications;,citation_journal_abbrev=Nat Commun;,citation_publisher=Nature Publishing Group;">
<meta name="citation_reference" content="citation_title=Determining benefits and costs for future generations;,citation_abstract=In economic project analysis, the rate at which future benefits and costs are discounted relative to current values often determines whether a project passes the benefit-cost test. This is especially true of projects with long time horizons, such as those to reduce greenhouse gas (GHG) emissions. Whether the benefits of climate policies, which can last for centuries, outweigh the costs, many of which are borne today, is especially sensitive to the rate at which future benefits are discounted. This is also true of other policies, e.g., affecting nuclear waste disposal or the construction of long-lived infrastructure. The United States and others should consider adopting a different approach to estimating costs and benefits in light of uncertainty. The United States and others should consider adopting a different approach to estimating costs and benefits in light of uncertainty.;,citation_author=K. Arrow;,citation_author=M. Cropper;,citation_author=C. Gollier;,citation_author=B. Groom;,citation_author=G. Heal;,citation_author=R. Newell;,citation_author=W. Nordhaus;,citation_author=R. Pindyck;,citation_author=W. Pizer;,citation_author=P. Portney;,citation_author=T. Sterner;,citation_author=R. S. J. Tol;,citation_author=M. Weitzman;,citation_publication_date=2013-07-26;,citation_cover_date=2013-07-26;,citation_year=2013;,citation_fulltext_html_url=https://science.sciencemag.org/content/341/6144/349;,citation_issue=6144;,citation_doi=10.1126/science.1235665;,citation_issn=0036-8075, 1095-9203;,citation_volume=341;,citation_journal_title=Science;,citation_publisher=American Association for the Advancement of Science;">
<meta name="citation_reference" content="citation_title=Social choice and individual values;,citation_author=Kenneth J. Arrow;,citation_publication_date=1951;,citation_cover_date=1951;,citation_year=1951;">
<meta name="citation_reference" content="citation_title=Exploratory modeling for policy analysis;,citation_abstract=Exploratory modeling is using computational experiments to assist in reasoning about systems where there is significant uncertainty. While frequently confused with the use of models to consolidate knowledge into a package that is used to predict system behavior, exploratory modeling is a very different kind of use, requiring a different methodology for model development. This paper distinguishes these two broad classes of model use describes some of the approaches used in exploratory modeling, and suggests some technological innovations needed to facilitate it.;,citation_author=Steve Bankes;,citation_publication_date=1993-06-01;,citation_cover_date=1993-06-01;,citation_year=1993;,citation_fulltext_html_url=https://pubsonline.informs.org/doi/abs/10.1287/opre.41.3.435;,citation_issue=3;,citation_doi=10/c7rgcr;,citation_issn=0030-364X;,citation_volume=41;,citation_journal_title=Operations Research;,citation_journal_abbrev=Operations Research;">
<meta name="citation_reference" content="citation_title=Generative AI without guardrails can harm learning: Evidence from high school mathematics;,citation_abstract=Generative AI is poised to revolutionize how humans work, and has already demonstrated promise in significantly improving human productivity. A key question is how generative AI affects learning—namely, how humans acquire new skills as they perform tasks. Learning is critical to long-term productivity, especially since generative AI is fallible and users must check its outputs. We study this question via a field experiment where we provide nearly a thousand high school math students with access to generative AI tutors. To understand the differential impact of tool design on learning, we deploy two generative AI tutors: one that mimics a standard ChatGPT interface (“GPT Base”) and one with prompts designed to safeguard learning (“GPT Tutor”). Consistent with prior work, our results show that having GPT-4 access while solving problems significantly improves performance (48% improvement in grades for GPT Base and 127% for GPT Tutor). However, we additionally find that when access is subsequently taken away, students actually perform worse than those who never had access (17% reduction in grades for GPT Base)—i.e., unfettered access to GPT-4 can harm educational outcomes. These negative learning effects are largely mitigated by the safeguards in GPT Tutor. Without guardrails, students attempt to use GPT-4 as a “crutch” during practice problem sessions, and subsequently perform worse on their own. Thus, decision-makers must be cautious about design choices underlying generative AI deployments to preserve skill learning and long-term productivity.;,citation_author=Hamsa Bastani;,citation_author=Osbert Bastani;,citation_author=Alp Sungu;,citation_author=Haosen Ge;,citation_author=Özge Kabakcı;,citation_author=Rei Mariman;,citation_publication_date=2025-07;,citation_cover_date=2025-07;,citation_year=2025;,citation_fulltext_html_url=https://www.pnas.org/doi/10.1073/pnas.2422633122;,citation_issue=26;,citation_doi=10.1073/pnas.2422633122;,citation_volume=122;,citation_journal_title=Proceedings of the National Academy of Sciences;,citation_publisher=Proceedings of the National Academy of Sciences;">
<meta name="citation_reference" content="citation_title=Deep Learning: Foundations and Concepts;,citation_author=Christopher M. Bishop;,citation_author=Hugh Bishop;,citation_publication_date=2024;,citation_cover_date=2024;,citation_year=2024;,citation_fulltext_html_url=https://link.springer.com/10.1007/978-3-031-45468-4;,citation_doi=10.1007/978-3-031-45468-4;,citation_isbn=978-3-031-45467-7 978-3-031-45468-4;">
<meta name="citation_reference" content="citation_title=Introduction to Probability, Second Edition;,citation_abstract=Developed from celebrated Harvard statistics lectures, Introduction to Probability provides essential language and tools for understanding statistics, randomness, and uncertainty. The book explores a wide variety of applications and examples, ranging from coincidences and paradoxes to Google PageRank and Markov chain Monte Carlo (MCMC). Additional application areas explored include genetics, medicine, computer science, and information theory. The authors present the material in an accessible style and motivate concepts using real-world examples. Throughout, they use stories to uncover connections between the fundamental distributions in statistics and conditioning to reduce complicated problems to manageable pieces.The book includes many intuitive explanations, diagrams, and practice problems. Each chapter ends with a section showing how to perform relevant simulations and calculations in R, a free statistical software environment. The second edition adds many new examples, exercises, and explanations, to deepen understanding of the ideas, clarify subtle concepts, and respond to feedback from many students and readers. New supplementary online resources have been developed, including animations and interactive visualizations, and the book has been updated to dovetail with these resources. Supplementary material is available on Joseph Blitzstein’s website www. stat110.net. The supplements include:Solutions to selected exercisesAdditional practice problemsHandouts including review material and sample exams Animations and interactive visualizations created in connection with the edX online version of Stat 110.Links to lecture videos available on ITunes U and YouTube There is also a complete instructor’s solutions manual available to instructors who require the book for a course.;,citation_author=Joseph K. Blitzstein;,citation_author=Jessica Hwang;,citation_publication_date=2019-02-08;,citation_cover_date=2019-02-08;,citation_year=2019;,citation_fulltext_html_url=http://probabilitybook.net;,citation_isbn=978-1-138-36991-7;">
<meta name="citation_reference" content="citation_title=A water risk index for portfolio exposure to climatic extremes: Conceptualization and an application to the mining industry;,citation_abstract=Corporations, industries and non-governmental organizations have become increasingly concerned with growing water risks in many parts of the world. Most of the focus has been on water scarcity and competition for the resource between agriculture, urban users, ecology and industry. However, water risks are multi-dimensional. Water-related hazards include flooding due to extreme rainfall, persistent drought and pollution, either due to industrial operations themselves, or to the failure of infrastructure. Most companies have risk management plans at each operational location to address these risks to a certain design level. The residual risk may or may not be managed, and is typically not quantified at a portfolio scale, i.e. across many sites. Given that climate is the driver of many of these extreme events, and there is evidence of quasi-periodic climate regimes at inter-annual and decadal timescales, it is possible that a portfolio is subject to persistent, multi-year exceedances of the design level. In other words, for a multi-national corporation, it is possible that there is correlation in the climate-induced portfolio water risk across its operational sites as multiple sites may experience a hazard beyond the design level in a given year. Therefore, from an investor’s perspective, a need exists for a water risk index that allows for an exploration of the possible space and/or time clustering in exposure across many sites contained in a portfolio. This paper represents a first attempt to develop an index for financial exposure of a geographically diversified, global portfolio to the time-varying risk of climatic extremes using long daily global rainfall datasets derived from climate re-analysis models. Focusing on extreme daily rainfall amounts and using examples from major mining companies, we illustrate how the index can be developed. We discuss how companies can use it to explore their corporate exposure, and what they may need to disclose to investors and regulators to promote transparency as to risk exposure and mitigation efforts. For the examples of mining companies provided, we note that the actual exposure is substantially higher than would be expected in the absence of space and time correlation of risk as is usually tacitly assumed. We also find evidence for the increasing exposure to climate-induced risk, and for decadal variability in exposure. The relative vulnerability of different portfolios to multiple extreme events in a given year is also demonstrated.;,citation_author=Luc Bonnafous;,citation_author=Upmanu Lall;,citation_author=Jason Siegel;,citation_publication_date=2017;,citation_cover_date=2017;,citation_year=2017;,citation_fulltext_html_url=https://www.hydrol-earth-syst-sci.net/21/2075/2017/;,citation_issue=4;,citation_doi=10/f96k67;,citation_volume=21;,citation_journal_title=Hydrology and Earth System Sciences;">
<meta name="citation_reference" content="citation_title=Dismantling Environmental Racism in the USA;,citation_abstract=A growing body of evidence reveals that people of colour and low‐income persons have borne greater environmental and health risks than the society at large in their neighbourhoods, workplaces and playgrounds. Over the past decade or so, grassroots activists have attempted to change the way government implements environmental, health and civil rights laws. A new movement has emerged in opposition to environmental racism and environmental injustice. Over the past two decades or so, grassroots activists have had some success in changing the way the federal government treats communities of colour and their inhabitants. Grassroots groups have also organised, educated and empowered themselves to improve the way health and environmental policies are administered. Environmentalism is now equated with social justice and civil rights.;,citation_author=Robert D. Bullard;,citation_publication_date=1999-02-01;,citation_cover_date=1999-02-01;,citation_year=1999;,citation_fulltext_html_url=https://doi.org/10.1080/13549839908725577;,citation_issue=1;,citation_doi=10.1080/13549839908725577;,citation_issn=1354-9839;,citation_volume=4;,citation_journal_title=Local Environment;,citation_publisher=Routledge;">
<meta name="citation_reference" content="citation_title=Cascading risks: Understanding the 2021 winter blackout in Texas;,citation_abstract=The Texas freeze of February 2021 left more than 4.5 million customers (more than 10 million people) without electricity at its peak, some for several days. The freeze had cascading effects on other services reliant upon electricity including drinking water treatment and medical services. Economic losses from lost output and damage are estimated to be $130 billion in Texas alone. In the wake of the freeze, there has been major fallout among regulators and utilities as actors sought to apportion blame and utilities and generators began to settle up accounts. This piece offers a retrospective on what caused the blackouts and the knock-on effects on other services, the subsequent financial and political effects of the freeze, and the implications for Texas and the country going forward. Texas failed to sufficiently winterize its electricity and gas systems after 2011. Feedback between failures in the two systems made the situation worse. Overall, the state faced outages of 30 GW of electricity as demand reached unprecedented highs. The gap between production and demand forced the non-profit grid manager, the Electric Reliability Council of Texas (ERCOT), to cut off supply to millions of customers or face a systems collapse that by some accounts was minutes away. The 2021 freeze suggests a need to rethink the state’s regulatory approach to energy to avoid future such outcomes. Weatherization, demand response, and expanded interstate interconnections are potential solutions Texas should consider to avoid generation losses, reduce demand, and tap neighboring states’ capacity.;,citation_author=Joshua W. Busby;,citation_author=Kyri Baker;,citation_author=Morgan D. Bazilian;,citation_author=Alex Q. Gilbert;,citation_author=Emily Grubert;,citation_author=Varun Rai;,citation_author=Joshua D. Rhodes;,citation_author=Sarang Shidore;,citation_author=Caitlin A. Smith;,citation_author=Michael E. Webber;,citation_publication_date=2021-07-01;,citation_cover_date=2021-07-01;,citation_year=2021;,citation_fulltext_html_url=outage;,citation_doi=10.1016/j.erss.2021.102106;,citation_issn=2214-6296;,citation_volume=77;,citation_journal_title=Energy Research &amp;amp;amp; Social Science;,citation_journal_abbrev=Energy Research &amp;amp; Social Science;,citation_publisher=Elsevier;">
<meta name="citation_reference" content="citation_title=An introduction to statistical modeling of extreme values;,citation_author=Stuart Coles;,citation_publication_date=2001;,citation_cover_date=2001;,citation_year=2001;,citation_isbn=1-85233-459-2;,citation_series_title=Springer series in statistics;">
<meta name="citation_reference" content="citation_title=Climate services: The business of physical risk;,citation_abstract=A growing number of investors, insurers, financial services providers, and nonprofits rely on information about localized physical climate risks, like floods, hurricanes, and wildfires. The outcomes of these risk projections have significant consequences in the economy, including allocating investment capital, impacting housing prices and demographic shifts, and prioritizing adaptation infrastructure projects. The climate risk information available to individual citizens and municipalities, however, is limited and expensive to access. Further, many providers of climate services use black box models that make overseeing the scientific rigor of their methodologies impossible— a concern given scientific critiques that many may be obfuscating the uncertainty in their projections. Municipalities that want to challenge insurance and bond rating determinations must rally significant resources for modeling and data, a scattershot policing method at best. And when companies have access to sophisticated modeling about future impacts— some of them potentially devastating for entire communities—the decision to share that information has been largely left up to the corporation.This Article argues that actionable and transparent information about our climate-changed future is a public good that the private sector cannot be depended upon to provide equitably or reliably. Further, all private climate services rely on upstream climate data and models that were collected and produced by an enormous network of public institutions. There are important lessons to be learned from the recent success of special interests in pressing for the privatization of weather data and services—a trend that has knock-on effects for weather forecasts globally. This Article urges state and federal governments to invest in their own climate services capacity at a scale not currently contemplated. Risk assessments lacking a scientific basis can lead to maladaptation across the economy. While it is a potentially limited matter of consumer protection or tort liability when a consultancy over-promises its analytical capabilities, it is a much larger problem if regulators themselves misunderstand the limits of uncertainty when designing risk oversight.;,citation_author=Madison Condon;,citation_publication_date=2023-03-22;,citation_cover_date=2023-03-22;,citation_year=2023;,citation_fulltext_html_url=https://papers.ssrn.com/abstract=4396826;">
<meta name="citation_reference" content="citation_title=Market myopia’s climate bubble;,citation_abstract=A growing number of financial institutions, ranging from BlackRock to the Bank of England, have warned that markets may not be accurately incorporating climate change-related risks into asset prices. This Article seeks to explain how this mispricing can exist at the level of individual assets drawing from scholarship on corporate governance and the mechanisms of market (in)efficiency. Market actors: 1. Lack the fine-grained asset-level data they need in order to assess risk exposure; 2. Continue to rely on outdated means of assessing risk; 3. Have misaligned incentives resulting in climate-specific agency costs; 4. Have myopic biases exacerbated by climate change misinformation; and 5. Are impeded by captured regulators distorting the market. Further, trends in institutional share ownership reinforce apathy regarding assessment of firm-specific fundamentals, especially over long-term horizons. This underpricing of corporate climate risk contributes to the negative effects of climate change itself, as the mispricing of risk in the present leads to a misallocation of investment capital, hindering future adaptation and subsidizing future fossil combustion. These risks could accumulate to the macroeconomic scale, generating a systemic risk to the financial system. While a broad array of government interventions are necessary to mitigate climate related financial risks, this Article focuses on proposals for corporate governance and securities regulation—and their limits. Signals from the Biden Administration suggest that mandatory climate risk disclosure regulation from the Securities and Exchange Commission is forthcoming. This Article argues that climate risk disclosure is necessary, though alone not sufficient, to address the widespread disregard of corporate climate exposure.;,citation_author=Madison Condon;,citation_publication_date=2021-02-09;,citation_cover_date=2021-02-09;,citation_year=2021;,citation_fulltext_html_url=https://papers.ssrn.com/abstract=3782675;,citation_issue=2022;,citation_doi=10.2139/ssrn.3782675;,citation_volume=63;,citation_journal_title=Utah Law Review;">
<meta name="citation_reference" content="citation_title=Statistics for spatio-temporal data;,citation_author=Noel A. C. Cressie;,citation_author=Christopher K. Wikle;,citation_publication_date=2011;,citation_cover_date=2011;,citation_year=2011;,citation_isbn=978-0-471-69274-4;">
<meta name="citation_reference" content="citation_title=Understanding Convolutions on Graphs;,citation_abstract=Understanding the building blocks and design choices of graph neural networks.;,citation_author=Ameya Daigavane;,citation_author=Balaraman Ravindran;,citation_author=Gaurav Aggarwal;,citation_publication_date=2021-09-02;,citation_cover_date=2021-09-02;,citation_year=2021;,citation_fulltext_html_url=https://distill.pub/2021/understanding-gnns;,citation_issue=9;,citation_doi=10.23915/distill.00032;,citation_issn=2476-0757;,citation_volume=6;,citation_journal_title=Distill;,citation_journal_abbrev=Distill;">
<meta name="citation_reference" content="citation_title=Theory of value; an axiomatic analysis of economic equilibrium;,citation_author=Gerard Debreu;,citation_publication_date=1959;,citation_cover_date=1959;,citation_year=1959;">
<meta name="citation_reference" content="citation_title=Contribution of Antarctica to past and future sea-level rise;,citation_abstract=Climate and ice-sheet modelling that includes ice fracture dynamics reveals that Antarctica could contribute more than a metre of sea-level rise by 2100 and more than 13 metres by 2500, if greenhouse gas emissions continue unabated.;,citation_author=Robert M. DeConto;,citation_author=David Pollard;,citation_publication_date=2016-03;,citation_cover_date=2016-03;,citation_year=2016;,citation_fulltext_html_url=http://www.nature.com/articles/nature17145;,citation_issue=7596;,citation_doi=10.1038/nature17145;,citation_issn=1476-4687;,citation_volume=531;,citation_journal_title=Nature;,citation_journal_abbrev=Nature;">
<meta name="citation_reference" content="citation_title=Evaluating the effect of flood damage-reducing measures: A case study of the unembanked area of Rotterdam, the Netherlands;,citation_abstract=Empirical evidence of increasing flood damages and the prospect of climatic change has initiated discussions in the flood management community on how to effectively manage flood risks. In the Netherlands, the framework of multi-layer safety (MLS) has been introduced to support this risk-based approach. The MLS framework consists of three layers: (i) prevention, (ii) spatial planning and (iii) evacuation. This paper presents a methodology to evaluate measures in the second layer, such as wet proofing, dry proofing or elevating buildings. The methodology uses detailed land-use data for the area around the city of Rotterdam (up to building level) that has recently become available. The vulnerability of these detailed land-use classes to flooding is assessed using the stage–damage curves from different international models. The methodology is demonstrated using a case study in the unembanked area of Rotterdam in the Netherlands, as measures from the second layer may be particularly effective there. The results show that the flood risk in the region is considerable: EUR 36 million p.a. A large part (almost 60&nbsp;%) of this risk results from industrial land use, emphasising the need to give this category more attention in flood risk assessments. It was found that building level measures could substantially reduce flood risks in the region because of the relatively low inundation levels of buildings. Risk to residential buildings would be reduced by 40&nbsp;% if all buildings would be wet-proofed, by 89&nbsp;% if all buildings would be dry-proofed and elevating buildings over 100&nbsp;cm would render the risk almost zero. While climate change could double the risk in 2100, such building level measures could easily nullify this effect. Despite the high potential of such measures, actual implementation is still limited. This is partly caused by the lack of knowledge regarding these measures by most Dutch companies and the legal impossibility for municipalities to enforce most of these measures as they would go beyond the building codes established at the national level.;,citation_author=Hans Moel;,citation_author=Mathijs Vliet;,citation_author=Jeroen C. J. H. Aerts;,citation_publication_date=2014-06-01;,citation_cover_date=2014-06-01;,citation_year=2014;,citation_fulltext_html_url=https://doi.org/10.1007/s10113-013-0420-z;,citation_issue=3;,citation_doi=10.1007/s10113-013-0420-z;,citation_issn=1436-378X;,citation_volume=14;,citation_journal_title=Regional Environmental Change;,citation_journal_abbrev=Reg Environ Change;">
<meta name="citation_reference" content="citation_title=Real Options by Spreadsheet: Parking Garage Case Example;,citation_author=Richard Neufville;,citation_author=Stefan Scholtes;,citation_author=Tao Wang;,citation_publication_date=2006;,citation_cover_date=2006;,citation_year=2006;,citation_fulltext_html_url=https://ascelibrary.org/doi/abs/10.1061/(ASCE)1076-0342(2006)12:2(107);,citation_issue=2;,citation_doi=10.1061/(asce)1076-0342(2006)12:2(107);,citation_volume=12;,citation_journal_title=Journal of Infrastructure Systems;,citation_journal_abbrev=Journal of Infrastructure Systems;">
<meta name="citation_reference" content="citation_title=Managing uncertainty in flood protection planning with climate projections;,citation_author=Beatrice Dittes;,citation_author=Olga Špačková;,citation_author=Lukas Schoppa;,citation_author=Daniel Straub;,citation_publication_date=2018;,citation_cover_date=2018;,citation_year=2018;,citation_fulltext_html_url=https://www.hydrol-earth-syst-sci.net/22/2511/2018/;,citation_issue=4;,citation_doi=10.5194/hess-22-2511-2018;,citation_volume=22;,citation_journal_title=Hydrology and Earth System Sciences;">
<meta name="citation_reference" content="citation_title=Improving the representation of climate risks in long-term electricity systems planning: A critical review;,citation_abstract=Purpose of Review Electricity systems face substantial climate risks which are escalating due to electrification, renewable energy intermittency, population changes, and the intensifying impacts of climate change such as extreme temperatures and weather-induced infrastructure damage. This critical review investigates climate risks to the electricity sector and scrutinizes the methodologies used to represent climate risk in long-term electricity system planning studies. Recent Findings Climate risks to electricity systems are driven by extreme weather, average weather, technology, and other social and technological factors. All are expected to evolve in the future. Future climate risks to electricity systems depend on interactions between each, and thus assessing future climate risks to electricity systems requires exploring a wide range of possible futures. Summary Many studies rely on weather data and socio-economic scenarios that are inadequate to fully characterize climate risks to present and future electricity systems. We advocate for more holistic assessments that incorporate comprehensive weather data, acknowledge dynamic multi-sector interactions, and employ adaptive and robust methodologies.;,citation_author=James Doss-Gollin;,citation_author=Yash Amonkar;,citation_author=Katlyn Schmeltzer;,citation_author=Daniel Cohan;,citation_publication_date=2023-08-16;,citation_cover_date=2023-08-16;,citation_year=2023;,citation_fulltext_html_url=https://doi.org/10.1007/s40518-023-00224-3;,citation_doi=10.1007/s40518-023-00224-3;,citation_issn=2196-3010;,citation_journal_title=Current Sustainable/Renewable Energy Reports;,citation_journal_abbrev=Curr Sustainable Renewable Energy Rep;">
<meta name="citation_reference" content="citation_title=Robust adaptation to multiscale climate variability;,citation_abstract=The assessment and implementation of structural or financial instruments for climate risk mitigation requires projections of future climate risk over the operational life of each proposed instrument. A point often neglected in the climate adaptation literature is that the physical sources of predictability differ between projects with long and short planning periods: while historical and paleo climate records emphasize modes of variability, anthropogenic climate change is expected to alter their occurrence at longer time scales. In this paper we present a set of stylized experiments to assess the uncertainties and biases involved in estimating future climate risk over a finite future period, given a limited observational record. These experiments consider both quasi-periodic and secular change for the underlying risk, as well as statistical models for estimating this risk from an N-year historical record. The uncertainty of IPCC-like future scenarios is considered through an equivalent sample size N. The relative importance of estimating the short- or long-term risk extremes depends on the investment life M. Shorter design lives are preferred for situations where inter-annual to decadal variability can be successfully identified and predicted, suggesting the importance of sequential investment strategies for adaptation.;,citation_author=James Doss-Gollin;,citation_author=David J. Farnham;,citation_author=Scott Steinschneider;,citation_author=Upmanu Lall;,citation_publication_date=2019-06-07;,citation_cover_date=2019-06-07;,citation_year=2019;,citation_fulltext_html_url=http://agupubs.onlinelibrary.wiley.com/doi/abs/10.1029/2019EF001154;,citation_issue=7;,citation_doi=10.1029/2019ef001154;,citation_issn=2328-4277;,citation_volume=7;,citation_journal_title=Earth’s Future;">
<meta name="citation_reference" content="citation_title=A subjective Bayesian framework for synthesizing deep uncertainties in climate risk management;,citation_abstract=Projections of nonstationary climate risks can vary considerably from one source to another, posing considerable communication and decision-analytical challenges. One such challenge is how to present trade-offs under deep uncertainty in a salient and interpretable manner. Some common approaches include analyzing a small subset of projections or treating all considered projections as equally likely. These approaches can underestimate risks, hide deep uncertainties, and are mostly silent on which assumptions drive decision-relevant outcomes. Here we introduce and demonstrate a transparent Bayesian framework for synthesizing deep uncertainties to inform climate risk management. The first step of this workflow is to generate an ensemble of simulations representing possible futures and analyze them through standard exploratory modeling techniques. Next, a small set of probability distributions representing subjective beliefs about the likelihood of possible futures is used to weight the scenarios. Finally, these weights are used to compute and characterize trade-offs, conduct robustness checks, and reveal implicit assumptions. We demonstrate the framework through a didactic case study analyzing how high to elevate a house to manage coastal flood risks.;,citation_author=James Doss-Gollin;,citation_author=Klaus Keller;,citation_publication_date=2023-01;,citation_cover_date=2023-01;,citation_year=2023;,citation_fulltext_html_url=https://onlinelibrary.wiley.com/doi/abs/10.1029/2022EF003044;,citation_issue=1;,citation_doi=10.1029/2022EF003044;,citation_issn=2328-4277;,citation_volume=11;,citation_journal_title=Earth’s Future;">
<meta name="citation_reference" content="citation_title=How unprecedented was the February 2021 Texas cold snap?;,citation_abstract=Winter storm Uri brought severe cold to the southern United States in February 2021, causing a cascading failure of interdependent systems in Texas where infrastructure was not adequately prepared for such cold. In particular, the failure of interconnected energy systems restricted electricity supply just as demand for heating spiked, leaving millions of Texans without heat or electricity, many for several days. This motivates the question: did historical storms suggest that such temperatures were known to occur, and if so with what frequency? We compute a temperature-based proxy for heating demand and use this metric to answer the question “what would the aggregate demand for heating have been had historic cold snaps occurred with today’s population?”. We find that local temperatures and the inferred demand for heating per capita across the region served by the Texas Interconnection were more severe during a storm in December 1989 than during February 2021, and that cold snaps in 1951 and 1983 were nearly as severe. Given anticipated population growth, future storms may lead to even greater infrastructure failures if adaptive investments are not made. Further, electricity system managers should prepare for trends in electrification of heating to drive peak annual loads on the Texas Interconnection during severe winter storms.;,citation_author=James Doss-Gollin;,citation_author=David J. Farnham;,citation_author=Upmanu Lall;,citation_author=Vijay Modi;,citation_publication_date=2021-06-08;,citation_cover_date=2021-06-08;,citation_year=2021;,citation_fulltext_html_url=e;,citation_doi=10.1088/1748-9326/ac0278;,citation_issn=1748-9326;,citation_journal_title=Environmental Research Letters;,citation_journal_abbrev=Environ. Res. Lett.;">
<meta name="citation_reference" content="citation_title=Think Bayes;,citation_abstract=If you know how to program, you’re ready to tackle Bayesian statistics. With this book, you’ll learn how to solve statistical problems with Python code instead of mathematical formulas, using discrete probability distributions rather than continuous mathematics. Once you get the math out of the way, the Bayesian fundamentals will become clearer and you’ll begin to apply these techniques to real-world problems.Bayesian statistical methods are becoming more common and more important, but there aren’t many resources available to help beginners. Based on undergraduate classes taught by author Allen B. Downey, this book’s computational approach helps you get a solid start.Use your programming skills to learn and understand Bayesian statisticsWork with problems involving estimation, prediction, decision analysis, evidence, and Bayesian hypothesis testingGet started with simple examples, using coins, dice, and a bowl of cookiesLearn computational methods for solving real-world problems;,citation_author=Allen B. Downey;,citation_publication_date=2021-05-18;,citation_cover_date=2021-05-18;,citation_year=2021;,citation_fulltext_html_url=https://allendowney.github.io/ThinkBayes2/;,citation_isbn=978-1-4920-8941-4;">
<meta name="citation_reference" content="citation_title=Risk, ambiguity, and the Savage axioms;,citation_abstract=Abstract. I. Are there uncertainties that are not risks? 643. — II. Uncertainties that are not risks, 647. — III. Why are some uncertainties not risks? — 656.;,citation_author=Daniel Ellsberg;,citation_publication_date=1961-11-01;,citation_cover_date=1961-11-01;,citation_year=1961;,citation_fulltext_html_url=http://academic.oup.com/qje/article/75/4/643/1913802;,citation_issue=4;,citation_doi=10.2307/1884324;,citation_issn=0033-5533;,citation_volume=75;,citation_journal_title=The Quarterly Journal of Economics;,citation_journal_abbrev=Q J Econ;,citation_publisher=Oxford Academic;">
<meta name="citation_reference" content="citation_title=Regional extreme precipitation events: Robust inference from credibly simulated GCM variables;,citation_abstract=eneral circulation models (GCMs) have been demonstrated to produce estimates of precipitation, including the frequency of extreme precipitation, with substantial bias and uncertainty relative to their representation of other fields. Thus, while theory predicts changes in the hydrologic cycle under anthropogenic warming, there is generally low confidence in future projections of extreme precipitation frequency for specific river basins. In this paper, we explore whether a GCM simulates large-scale atmospheric circulation indices that are associated with regional extreme precipitation (REP) days more accurately than it simulates REP days themselves, and thus whether conditional simulation of the precipitation events based on the circulation indices may improve the simulation of REP events. We show that a coupled Geophysical Fluid Dynamics Laboratory GCM simulates too many springtime REP days in the Ohio River Basin in historical (1950–2005) simulations. The GCM, however, does credibly simulate the distributional and persistence properties of several indices (which represent the large-scale atmospheric pressure features, local atmospheric moisture content, and local vertical velocity) that are shown to modulate the likelihood of REP occurrence in the reanalysis/observational record. We show that simulation of REP events based on the GCM-based atmospheric indices greatly reduces the bias of GCM REP frequency relative to the observed record. The simulation is conducted via a Bayesian regression model by imposing the empirical relationship between observed REP occurrence and the reanalysis-based atmospheric indices. Application of this model to future (2006–2100) representative concentration pathway 8.5 scenario suggests an increasing trend in springtime REP incidence in the study region. The proposed approach of simulating precipitation events of interest, particularly those poorly represented in GCMs, with a statistical model based on climate indices that are reasonably simulated by GCMs could be applied to subseasonal to seasonal forecasts as well as future projections.;,citation_author=David J Farnham;,citation_author=James Doss-Gollin;,citation_author=Upmanu Lall;,citation_publication_date=2018;,citation_cover_date=2018;,citation_year=2018;,citation_fulltext_html_url=https://doi.org/10.1002/2017wr021318;,citation_issue=6;,citation_doi=10.1002/2017wr021318;,citation_volume=54;,citation_journal_title=Water Resources Research;">
<meta name="citation_reference" content="citation_title=Learning about climate change uncertainty enables flexible water infrastructure planning;,citation_abstract=Water resources planning requires infrastructure development consider regional climatic uncertainties. Here the authors introduce a new dynamic planning framework that captures opportunities to learn about climate change over time. By applying it to reservoir planning in Kenya, they show the value of flexible approaches in responding to learning.;,citation_author=Sarah Fletcher;,citation_author=Megan Lickley;,citation_author=Kenneth Strzepek;,citation_publication_date=2019-04-16;,citation_cover_date=2019-04-16;,citation_year=2019;,citation_fulltext_html_url=http://www.nature.com/articles/s41467-019-09677-x;,citation_issue=1;,citation_doi=10.1038/s41467-019-09677-x;,citation_issn=2041-1723;,citation_volume=10;,citation_journal_title=Nature Communications;">
<meta name="citation_reference" content="citation_title=Equity in Water Resources Planning: A Path Forward for Decision Support Modelers;,citation_author=Sarah Fletcher;,citation_author=Antonia Hadjimichael;,citation_author=Julianne Quinn;,citation_author=Khalid Osman;,citation_author=Matteo Giuliani;,citation_author=David Gold;,citation_author=Anjuli Jain Figueroa;,citation_author=Bethany Gordon;,citation_publication_date=2022-07;,citation_cover_date=2022-07;,citation_year=2022;,citation_fulltext_html_url=https://ascelibrary.org/doi/10.1061/%28ASCE%29WR.1943-5452.0001573;,citation_issue=7;,citation_doi=10.1061/(ASCE)WR.1943-5452.0001573;,citation_issn=0733-9496, 1943-5452;,citation_volume=148;,citation_journal_title=Journal of Water Resources Planning and Management;,citation_journal_abbrev=J. Water Resour. Plann. Manage.;">
<meta name="citation_reference" content="citation_title=Bold New Jersey Shore Flood Rules Could Be Blueprint for Entire U.S. Coast;,citation_abstract=Coastal flood zones where development is restricted will be based on future climate change projections, not past floods;,citation_author=Thomas Frank;,citation_publication_date=2022-08-22;,citation_cover_date=2022-08-22;,citation_year=2022;,citation_fulltext_html_url=https://www.scientificamerican.com/article/bold-new-jersey-shore-flood-rules-could-be-blueprint-for-entire-u-s-coast/;,citation_journal_title=Scientific American;">
<meta name="citation_reference" content="citation_title=The Elements of Statistical Learning;,citation_author=Jerome Friedman;,citation_author=Trevor Hastie;,citation_author=Robert Tibshirani;,citation_publication_date=2001;,citation_cover_date=2001;,citation_year=2001;,citation_volume=1;">
<meta name="citation_reference" content="citation_title=Using direct policy search to identify robust strategies in adapting to uncertain sea-level rise and storm surge;,citation_abstract=Sea-level rise poses considerable risks to coastal communities, ecosystems, and infrastructure. Decision makers are faced with uncertain sea-level projections when designing a strategy for coastal adaptation. The traditional methods are often silent on tradeoffs as well as the effects of tail-area events and of potential future learning. Here we reformulate a simple sea-level rise adaptation model to address these concerns. We show that Direct Policy Search yields improved solution quality, with respect to Pareto-dominance in the objectives, over the traditional approach under uncertain sea-level rise projections and storm surge. Additionally, the new formulation produces high quality solutions with less computational demands than an intertemporal optimization approach. Our results illustrate the utility of multi-objective adaptive formulations for the example of coastal adaptation and point to wider-ranging application in climate change adaptation decision problems.;,citation_author=Gregory G. Garner;,citation_author=Klaus Keller;,citation_publication_date=2018-09-01;,citation_cover_date=2018-09-01;,citation_year=2018;,citation_fulltext_html_url=http://www.sciencedirect.com/science/article/pii/S1364815217307727;,citation_doi=10.1016/j.envsoft.2018.05.006;,citation_issn=1364-8152;,citation_volume=107;,citation_journal_title=Environmental Modelling &amp;amp;amp; Software;,citation_journal_abbrev=Environmental Modelling &amp;amp; Software;">
<meta name="citation_reference" content="citation_title=Bayesian Data Analysis;,citation_author=Andrew Gelman;,citation_author=John B Carlin;,citation_author=Hal S Stern;,citation_author=Donald B Rubin;,citation_publication_date=2014;,citation_cover_date=2014;,citation_year=2014;">
<meta name="citation_reference" content="citation_title=Philosophy and the practice of Bayesian statistics;,citation_abstract=A substantial school in the philosophy of science identifies Bayesian inference with inductive inference and even rationality as such, and seems to be strengthened by the rise and practical success of Bayesian statistics. We argue that the most successful forms of Bayesian statistics do not actually support that particular philosophy but rather accord much better with sophisticated forms of hypothetico-deductivism. We examine the actual role played by prior distributions in Bayesian models, and the crucial aspects of model checking and model revision, which fall outside the scope of Bayesian confirmation theory. We draw on the literature on the consistency of Bayesian updating and also on our experience of applied work in social science. Clarity about these matters should benefit not just philosophy of science, but also statistical practice. At best, the inductivist view has encouraged researchers to fit and compare models without checking them; at worst, theorists have actively discouraged practitioners from performing model checking because it does not fit into their framework.;,citation_author=Andrew Gelman;,citation_author=Cosma Rohilla Shalizi;,citation_publication_date=2013;,citation_cover_date=2013;,citation_year=2013;,citation_fulltext_html_url=http://onlinelibrary.wiley.com/doi/abs/10.1111/j.2044-8317.2011.02037.x;,citation_issue=1;,citation_doi=10/f4k2h4;,citation_issn=2044-8317;,citation_volume=66;,citation_journal_title=British Journal of Mathematical and Statistical Psychology;">
<meta name="citation_reference" content="citation_title=Regression and other stories;,citation_abstract=&amp;amp;amp;quot;Many textbooks on regression focus on theory and the simplest examples. Real statistical problems, however, are complex and subtle. This is not a book about the theory of regression. It is a book about how to use regression to solve real problems of comparison, estimation, prediction, and causal inference. It focuses on practical issues such as sample size and missing data and a wide range of goals and techniques. It jumps right in to methods and computer code you can use fresh out of the box&amp;quot;–Back cover;,citation_author=Andrew Gelman;,citation_publication_date=2021;,citation_cover_date=2021;,citation_year=2021;,citation_isbn=978-1-107-67651-0;,citation_series_title=Analytical methods for social research;">
<meta name="citation_reference" content="citation_title=Beyond subjective and objective in statistics;,citation_abstract=Decisions in statistical data analysis are often justified, criticized or avoided by using concepts of objectivity and subjectivity. We argue that the words “objective” and “subjective” in statistics discourse are used in a mostly unhelpful way, and we propose to replace each of them with broader collections of attributes, with objectivity replaced by transparency, consensus, impartiality and correspondence to observable reality, and subjectivity replaced by awareness of multiple perspectives and context dependence. Together with stability, these make up a collection of virtues that we think is helpful in discussions of statistical foundations and practice. The advantage of these reformulations is that the replacement terms do not oppose each other and that they give more specific guidance about what statistical science strives to achieve. Instead of debating over whether a given statistical method is subjective or objective (or normatively debating the relative merits of subjectivity and objectivity in statistical practice), we can recognize desirable attributes such as transparency and acknowledgement of multiple perspectives as complementary goals. We demonstrate the implications of our proposal with recent applied examples from pharmacology, election polling and socio-economic stratification. The aim of the paper is to push users and developers of statistical methods towards more effective use of diverse sources of information and more open acknowledgement of assumptions and goals.;,citation_author=Andrew Gelman;,citation_author=Christian Hennig;,citation_publication_date=2017;,citation_cover_date=2017;,citation_year=2017;,citation_fulltext_html_url=https://rss.onlinelibrary.wiley.com/doi/abs/10.1111/rssa.12276;,citation_issue=4;,citation_doi=10.1111/rssa.12276;,citation_issn=1467-985X;,citation_volume=180;,citation_journal_title=Journal of the Royal Statistical Society: Series A (Statistics in Society);">
<meta name="citation_reference" content="citation_title=Bayesian workflow;,citation_abstract=The Bayesian approach to data analysis provides a powerful way to handle uncertainty in all observations, model parameters, and model structure using probability theory. Probabilistic programming languages make it easier to specify and fit Bayesian models, but this still leaves us with many options regarding constructing, evaluating, and using these models, along with many remaining challenges in computation. Using Bayesian inference to solve real-world problems requires not only statistical skills, subject matter knowledge, and programming, but also awareness of the decisions made in the process of data analysis. All of these aspects can be understood as part of a tangled workflow of applied Bayesian statistics. Beyond inference, the workflow also includes iterative model building, model checking, validation and troubleshooting of computational problems, model understanding, and model comparison. We review all these aspects of workflow in the context of several examples, keeping in mind that in practice we will be fitting many models for any given problem, even if only a subset of them will ultimately be relevant for our conclusions.;,citation_author=Andrew Gelman;,citation_author=Aki Vehtari;,citation_author=Daniel Simpson;,citation_author=Charles C. Margossian;,citation_author=Bob Carpenter;,citation_author=Yuling Yao;,citation_author=Lauren Kennedy;,citation_author=Jonah Gabry;,citation_author=Paul-Christian Bürkner;,citation_author=Martin Modrák;,citation_publication_date=2020-11-03;,citation_cover_date=2020-11-03;,citation_year=2020;,citation_fulltext_html_url=http://arxiv.org/abs/2011.01808;,citation_doi=10.48550/arXiv.2011.01808;">
<meta name="citation_reference" content="citation_title=Extreme events: Dynamics, statistics and prediction;,citation_abstract=We review work on extreme events, their causes and consequences, by a group of European and American researchers involved in a three-year project on these topics. The review covers theoretical aspects of time series analysis and of extreme value theory, as well as of the deterministic modeling of extreme events, via continuous and discrete dynamic models. The applications include climatic, seismic and socio-economic events, along with their prediction. $&amp;amp;amp;lt;$br$&amp;gt;&amp;lt;$br$&amp;gt;$ Two important results refer to (i) the complementarity of spectral analysis of a time series in terms of the continuous and the discrete part of its power spectrum; and (ii) the need for coupled modeling of natural and socio-economic systems. Both these results have implications for the study and prediction of natural hazards and their human impacts.;,citation_author=M Ghil;,citation_author=P Yiou;,citation_author=S Hallegatte;,citation_author=B D Malamud;,citation_author=P Naveau;,citation_author=A Soloviev;,citation_author=P Friederichs;,citation_author=V Keilis-Borok;,citation_author=D Kondrashov;,citation_author=V Kossobokov;,citation_author=O Mestre;,citation_author=C Nicolis;,citation_author=H W Rust;,citation_author=P Shebalin;,citation_author=M Vrac;,citation_author=A Witt;,citation_author=I Zaliapin;,citation_publication_date=2011;,citation_cover_date=2011;,citation_year=2011;,citation_fulltext_html_url=http://www.nonlin-processes-geophys.net/18/295/2011/;,citation_issue=3;,citation_doi=10/fvzxvv;,citation_volume=18;,citation_journal_title=Nonlinear Processes in Geophysics;,citation_publisher=Copernicus GmbH;">
<meta name="citation_reference" content="citation_title=Multiple-Hazard Fragility and Restoration Models of Highway Bridges for Regional Risk and Resilience Assessment in the United States: State-of-the-Art Review;,citation_abstract=AbstractHighway bridges are one of the most vulnerable constituents of transportation networks when exposed to one or more natural hazards, such as earthquakes, hurricanes, tsunamis, and riverine floods. To facilitate and enhance prehazard and posthazard ...;,citation_author=Ioannis Gidaris;,citation_author=Jamie E. Padgett;,citation_author=Andre R. Barbosa;,citation_author=Suren Chen;,citation_author=Daniel Cox;,citation_author=Bret Webb;,citation_author=Amy Cerato;,citation_publication_date=2017-03-01;,citation_cover_date=2017-03-01;,citation_year=2017;,citation_fulltext_html_url=https://ascelibrary.org/doi/10.1061/%28ASCE%29ST.1943-541X.0001672;,citation_issue=3;,citation_doi=10.1061/(ASCE)ST.1943-541X.0001672;,citation_issn=1943-541X;,citation_volume=143;,citation_journal_title=Journal of Structural Engineering;,citation_publisher=American Society of Civil Engineers;">
<meta name="citation_reference" content="citation_title=Beyond wickedness: Managing complex systems and climate change;,citation_author=Jonathan M. Gilligan;,citation_author=Michael P. Vandenbergh;,citation_publication_date=2020;,citation_cover_date=2020;,citation_year=2020;,citation_fulltext_html_url=https://wp0.vanderbilt.edu/lawreview/2020/12/beyond-wickedness-managing-complex-systems-and-climate-change/;,citation_volume=73;,citation_journal_title=Vanderbilt Law Review;,citation_journal_abbrev=Vand. L. Rev.;">
<meta name="citation_reference" content="citation_title=Extreme Value Theory and Statistics of Univariate Extremes: A Review;,citation_abstract=SummaryStatistical issues arising in modelling univariate extremes of a random sample have been successfully used in the most diverse fields, such as biometrics, finance, insurance and risk theory. Statistics of univariate extremes (SUE), the subject to be dealt with in this review paper, has recently faced a huge development, partially because rare events can have catastrophic consequences for human activities, through their impact on the natural and constructed environments. In the last decades, there has been a shift from the area of parametric SUE, based on probabilistic asymptotic results in extreme value theory, towards semi-parametric approaches. After a brief reference to Gumbel’s block methodology and more recent improvements in the parametric framework, we present an overview of the developments on the estimation of parameters of extreme events and on the testing of extreme value conditions under a semi-parametric framework. We further discuss a few challenging topics in the area of SUE. © 2014 The Authors. International Statistical Review © 2014 International Statistical Institute;,citation_author=M. Ivette Gomes;,citation_author=Armelle Guillou;,citation_publication_date=2015;,citation_cover_date=2015;,citation_year=2015;,citation_fulltext_html_url=https://onlinelibrary.wiley.com/doi/abs/10.1111/insr.12058;,citation_issue=2;,citation_doi=10.1111/insr.12058;,citation_issn=1751-5823;,citation_volume=83;,citation_journal_title=International Statistical Review;">
<meta name="citation_reference" content="citation_title=Downscaling projections of Indian monsoon rainfall using a non‐homogeneous hidden Markov model;,citation_author=Arthur M Greene;,citation_author=Andrew W Robertson;,citation_author=Padhraic Smyth;,citation_author=Scott Triglia;,citation_publication_date=2011-01;,citation_cover_date=2011-01;,citation_year=2011;,citation_issue=655;,citation_doi=10.1002/qj.788;,citation_volume=137;,citation_journal_title=Quart J Royal Meteoro Soc;,citation_publisher=Wiley Online Library;">
<meta name="citation_reference" content="citation_title=Emissions – the ’business as usual’ story is misleading;,citation_abstract=Stop using the worst-case scenario for climate warming as the most likely outcome — more-realistic baselines make for better policy.;,citation_author=Zeke Hausfather;,citation_author=Glen P. Peters;,citation_publication_date=2020-01;,citation_cover_date=2020-01;,citation_year=2020;,citation_fulltext_html_url=http://www.nature.com/articles/d41586-020-00177-3;,citation_issue=7792, 7792;,citation_doi=10.1038/d41586-020-00177-3;,citation_volume=577;,citation_journal_title=Nature;,citation_publisher=Nature Publishing Group;">
<meta name="citation_reference" content="citation_title=Statistical methods in water resources;,citation_abstract=This text began as a collection of class notes for a course on applied statistical methods for hydrologists taught at the U.S. Geological Survey (USGS) National Training Center. Course material was formalized and organized into a textbook, first published in 1992 by Elsevier as part of their Studies in Environmental Science series. In 2002, the work was made available online as a USGS report.The text has now been updated as a USGS Techniques and Methods Report. It is intended to be a text in applied statistics for hydrology, environmental science, environmental engineering, geology, or biology that addresses distinctive features of environmental data. For example, water resources data tend to have many variables with a lower bound of zero, tend to be more skewed than data from many other disciplines, commonly contain censored data (less than values), and assumptions that the data are normally distributed are not appropriate. Computer-intensive methods (bootstrapping...;,citation_author=Dennis R. Helsel;,citation_author=Robert M. Hirsch;,citation_author=Karen R. Ryberg;,citation_author=Stacey A. Archfield;,citation_author=Edward J. Gilroy;,citation_publication_date=2020;,citation_cover_date=2020;,citation_year=2020;,citation_fulltext_html_url=https://pubs.usgs.gov/publication/tm4A3;,citation_doi=10.3133/tm4A3;,citation_issn=2328-7055;,citation_book_title=Techniques and Methods;">
<meta name="citation_reference" content="citation_title=Climate adaptation as a control problem: Review and perspectives on dynamic water resources planning under uncertainty;,citation_abstract=Climate change introduces substantial uncertainty to water resources planning, and raises the key question: when, or under what conditions, should adaptation occur? A number of recent studies aim to identify policies mapping future observations to actions—in other words, framing climate adaptation as an optimal control problem. This paper uses the control paradigm to review and classify recent dynamic planning studies according to their approaches to uncertainty characterization, policy structure, and solution methods. We propose a set of research gaps and opportunities in this area centered on the challenge of characterizing uncertainty, which prevents the unambiguous application of control methods to this problem. These include: exogenous uncertainty in forcing, model structure, and parameters propagated through a chain of climate and hydrologic models; endogenous uncertainty in human-environmental system dynamics across multiple scales; and sampling uncertainty due to the finite length of historical observations and future projections. Recognizing these challenges, several opportunities exist to improve the use of control methods for climate adaptation, namely: how problem context and understanding of climate processes might assist with uncertainty quantification and experimental design; out-of-sample validation and robustness of optimized adaptation policies; and monitoring and data assimilation, including trend detection, Bayesian inference, and indicator variable selection. We conclude with a summary of recommendations for dynamic water resources planning under climate change through the lens of optimal control.;,citation_author=Jonathan D. Herman;,citation_author=Julianne D. Quinn;,citation_author=Scott Steinschneider;,citation_author=Matteo Giuliani;,citation_author=Sarah Fletcher;,citation_publication_date=2020-01-07;,citation_cover_date=2020-01-07;,citation_year=2020;,citation_fulltext_html_url=http://agupubs.onlinelibrary.wiley.com/doi/abs/10.1029/2019WR025502;,citation_doi=10.1029/2019wr025502;,citation_issn=1944-7973;,citation_journal_title=Water Resources Research;">
<meta name="citation_reference" content="citation_title=How should robustness be defined for water systems planning under change?;,citation_abstract=Water systems planners have long recognized the need for robust solutions capable of withstanding deviations from the conditions for which they were designed. Robustness analyses have shifted from expected utility to exploratory bottom-up approaches which identify vulnerable scenarios prior to assigning likelihoods. Examples include Robust Decision Making (RDM), Decision Scaling, Info-Gap, and Many-Objective Robust Decision Making (MORDM). We propose a taxonomy of robustness frameworks to compare and contrast these approaches based on their methods of (1) alternative generation, (2) sampling of states of the world, (3) quantification of robustness measures, and (4) sensitivity analysis to identify important uncertainties. Building from the proposed taxonomy, we use a regional urban water supply case study in the Research Triangle region of North Carolina to illustrate the decision-relevant consequences that emerge from each of these choices. Results indicate that the methodological choices in the taxonomy lead to the selection of substantially different planning alternatives, underscoring the importance of an informed definition of robustness. Moreover, the results show that some commonly employed methodological choices and definitions of robustness can have undesired consequences when ranking decision alternatives. For the demonstrated test case, recommendations for overcoming these issues include: (1) decision alternatives should be searched rather than prespecified, (2) dominant uncertainties should be discovered through sensitivity analysis rather than assumed, and (3) a carefully elicited multivariate satisficing measure of robustness allows stakeholders to achieve their problem-specific performance requirements. This work emphasizes the importance of an informed problem formulation for systems facing challenging performance tradeoffs and provides a common vocabulary to link the robustness frameworks widely used in the field of water systems planning.;,citation_author=Jonathan D. Herman;,citation_author=Patrick M. Reed;,citation_author=Harrison B. Zeff;,citation_author=Gregory W. Characklis;,citation_publication_date=2015-10-01;,citation_cover_date=2015-10-01;,citation_year=2015;,citation_fulltext_html_url=http://ascelibrary.org/doi/abs/10.1061/(ASCE)WR.1943-5452.0000509;,citation_issue=10;,citation_doi=10.1061/(asce)wr.1943-5452.0000509;,citation_volume=141;,citation_journal_title=Journal of Water Resources Planning and Management;,citation_journal_abbrev=Journal of Water Resources Planning and Management;">
<meta name="citation_reference" content="citation_title=SALib: An open-source Python library for Sensitivity Analysis;,citation_abstract=Herman et al, (2017), SALib: An open-source Python library for Sensitivity Analysis, Journal of Open Source Software, 2(9), 97, doi:10.21105/joss.00097;,citation_author=Jon Herman;,citation_author=Will Usher;,citation_publication_date=2017-01-10;,citation_cover_date=2017-01-10;,citation_year=2017;,citation_fulltext_html_url=https://joss.theoj.org/papers/10.21105/joss.00097;,citation_issue=9;,citation_doi=10.21105/joss.00097;,citation_issn=2475-9066;,citation_volume=2;,citation_journal_title=Journal of Open Source Software;">
<meta name="citation_reference" content="citation_title=Evaluating the distributional fairness of alternative adaptation policies: A case study in Vietnam’s upper Mekong Delta;,citation_abstract=To support equitable adaptation planning, quantitative assessments should consider the fairness of the distribution of outcomes to different people. What constitutes a fair distribution, however, is a normative question. In this study, we explore the use of different moral principles drawn from theories of distributive justice to evaluate fairness. We use adaptation planning in Vietnam Mekong Delta as a case study. We evaluate the preference ranking of six alternative policies for seven moral principles across an ensemble of scenarios. Under the baseline scenario, each principle yields distinctive preference rankings, though most principles identify the same policy as the most preferred one. Across the ensemble of scenarios, the commonly used utilitarian principle yields the most stable ranking, while rankings from other principles are more sensitive to uncertainty. The sufficientarian and the envy-free principles yield the most distinctive ranking of policies, with a median ranking correlation of only 0.07 across all scenarios. Finally, we identify scenarios under which using these two principles results in reversed policy preference rankings. Our study highlights the importance of considering multiple moral principles in evaluating the fairness of adaptation policies, as this would reduce the possibility of maladaptation.;,citation_author=Bramka Arga Jafino;,citation_author=Jan H. Kwakkel;,citation_author=Frans Klijn;,citation_publication_date=2022-08-02;,citation_cover_date=2022-08-02;,citation_year=2022;,citation_fulltext_html_url=https://doi.org/10.1007/s10584-022-03395-y;,citation_issue=3;,citation_doi=10.1007/s10584-022-03395-y;,citation_issn=1573-1480;,citation_volume=173;,citation_journal_title=Climatic Change;,citation_journal_abbrev=Climatic Change;">
<meta name="citation_reference" content="citation_title=An Introduction to Statistical Learning;,citation_author=Gareth James;,citation_author=Daniela Witten;,citation_author=Trevor Hastie;,citation_author=Robert Tibshirani;,citation_publication_date=2013;,citation_cover_date=2013;,citation_year=2013;,citation_isbn=978-1-4614-7137-0;,citation_volume=103;,citation_series_title=Springer Texts in Statistics;">
<meta name="citation_reference" content="citation_title=Probability theory: The logic of science;,citation_author=Edwin T. Jaynes;,citation_publication_date=2003;,citation_cover_date=2003;,citation_year=2003;,citation_isbn=978052159271;">
<meta name="citation_reference" content="citation_title=Global exposure to river and coastal flooding: Long term trends and changes;,citation_author=Brenden Jongman;,citation_author=Philip J Ward;,citation_author=Jeroen C J H Aerts;,citation_publication_date=2012-10;,citation_cover_date=2012-10;,citation_year=2012;,citation_fulltext_html_url=http://linkinghub.elsevier.com/retrieve/pii/S0959378012000830;,citation_issue=4;,citation_doi=10.1016/j.gloenvcha.2012.07.004;,citation_volume=22;,citation_journal_title=Global Environmental Change;">
<meta name="citation_reference" content="citation_title=Many objective robust decision making for complex environmental systems undergoing change;,citation_abstract=This paper introduces many objective robust decision making (MORDM). MORDM combines concepts and methods from many objective evolutionary optimization and robust decision making (RDM), along with extensive use of interactive visual analytics, to facilitate the management of complex environmental systems. Many objective evolutionary search is used to generate alternatives for complex planning problems, enabling the discovery of the key tradeoffs among planning objectives. RDM then determines the robustness of planning alternatives to deeply uncertain future conditions and facilitates decision makers’ selection of promising candidate solutions. MORDM tests each solution under the ensemble of future extreme states of the world (SOW). Interactive visual analytics are used to explore whether solutions of interest are robust to a wide range of plausible future conditions (i.e., assessment of their Pareto satisficing behavior in alternative SOW). Scenario discovery methods that use statistical data mining algorithms are then used to identify what assumptions and system conditions strongly influence the cost-effectiveness, efficiency, and reliability of the robust alternatives. The framework is demonstrated using a case study that examines a single city’s water supply in the Lower Rio Grande Valley (LRGV) in Texas, USA. Results suggest that including robustness as a decision criterion can dramatically change the formulation of complex environmental management problems as well as the negotiated selection of candidate alternatives to implement. MORDM also allows decision makers to characterize the most important vulnerabilities for their systems, which should be the focus of ex post monitoring and identification of triggers for adaptive management.;,citation_author=Joseph R. Kasprzyk;,citation_author=Shanthi Nataraj;,citation_author=Patrick M. Reed;,citation_author=Robert J. Lempert;,citation_publication_date=2013-04-01;,citation_cover_date=2013-04-01;,citation_year=2013;,citation_fulltext_html_url=http://www.sciencedirect.com/science/article/pii/S1364815212003131;,citation_doi=10.1016/j.envsoft.2012.12.007;,citation_issn=1364-8152;,citation_volume=42;,citation_journal_title=Environmental Modelling &amp;amp;amp; Software;,citation_journal_abbrev=Environmental Modelling &amp;amp; Software;">
<meta name="citation_reference" content="citation_title=Climate risk management;,citation_author=Klaus Keller;,citation_author=Casey Helgeson;,citation_author=Vivek Srikrishnan;,citation_publication_date=2021;,citation_cover_date=2021;,citation_year=2021;,citation_issue=1;,citation_doi=10.1146/annurev-earth-080320-055847;,citation_issn=0084-6597;,citation_volume=49;,citation_journal_title=Annual Review of Earth and Planetary Sciences;,citation_publisher=Annual Reviews;">
<meta name="citation_reference" content="citation_title=Adam: A Method for Stochastic Optimization;,citation_abstract=We introduce Adam, an algorithm for first-order gradient-based optimization of stochastic objective functions, based on adaptive estimates of lower-order moments. The method is straightforward to implement, is computationally efficient, has little memory requirements, is invariant to diagonal rescaling of the gradients, and is well suited for problems that are large in terms of data and/or parameters. The method is also appropriate for non-stationary objectives and problems with very noisy and/or sparse gradients. The hyper-parameters have intuitive interpretations and typically require little tuning. Some connections to related algorithms, on which Adam was inspired, are discussed. We also analyze the theoretical convergence properties of the algorithm and provide a regret bound on the convergence rate that is comparable to the best known results under the online convex optimization framework. Empirical results demonstrate that Adam works well in practice and compares favorably to other stochastic optimization methods. Finally, we discuss AdaMax, a variant of Adam based on the infinity norm.;,citation_author=Diederik P. Kingma;,citation_author=Jimmy Ba;,citation_publication_date=2017-01-30;,citation_cover_date=2017-01-30;,citation_year=2017;,citation_fulltext_html_url=http://arxiv.org/abs/1412.6980;,citation_doi=10.48550/arXiv.1412.6980;">
<meta name="citation_reference" content="citation_title=Evolving understanding of Antarctic ice-sheet physics and ambiguity in probabilistic sea-level projections;,citation_abstract=Mechanisms such as ice-shelf hydrofracturing and ice-cliff collapse may rapidly increase discharge from marine-based ice sheets. Here, we link a probabilistic framework for sea-level projections to a small ensemble of Antarctic ice-sheet (AIS) simulations incorporating these physical processes to explore their influence on global-mean sea-level (GMSL) and relative sea-level (RSL). We compare the new projections to past results using expert assessment and structured expert elicitation about AIS changes. Under high greenhouse gas emissions (Representative Concentration Pathway [RCP] 8.5), median projected 21st century GMSL rise increases from 79 to 146 cm. Without protective measures, revised median RSL projections would by 2100 submerge land currently home to 153 million people, an increase of 44 million. The use of a physical model, rather than simple parameterizations assuming constant acceleration of ice loss, increases forcing sensitivity: overlap between the central 90% of simulations for 2100 for RCP 8.5 (93–243 cm) and RCP 2.6 (26–98 cm) is minimal. By 2300, the gap between median GMSL estimates for RCP 8.5 and RCP 2.6 reaches $&amp;amp;amp;gt;$10 m, with median RSL projections for RCP 8.5 jeopardizing land now occupied by 950 million people (versus 167 million for RCP 2.6). The minimal correlation between the contribution of AIS to GMSL by 2050 and that in 2100 and beyond implies current sea-level observations cannot exclude future extreme outcomes. The sensitivity of post-2050 projections to deeply uncertain physics highlights the need for robust decision and adaptive management frameworks.;,citation_author=Robert E. Kopp;,citation_author=Robert M. DeConto;,citation_author=Daniel A. Bader;,citation_author=Carling C. Hay;,citation_author=Radley M. Horton;,citation_author=Scott Kulp;,citation_author=Michael Oppenheimer;,citation_author=David Pollard;,citation_author=Benjamin H. Strauss;,citation_publication_date=2017;,citation_cover_date=2017;,citation_year=2017;,citation_fulltext_html_url=https://doi.org/10.1002/2017ef000663;,citation_issue=12;,citation_doi=10.1002/2017ef000663;,citation_issn=2328-4277;,citation_volume=5;,citation_journal_title=Earth’s Future;">
<meta name="citation_reference" content="citation_title=Probabilistic 21st and 22nd century sea-level projections at a global network of tide-gauge sites;,citation_abstract=Sea-level rise due to both climate change and non-climatic factors threatens coastal settlements, infrastructure, and ecosystems. Projections of mean global sea-level (GSL) rise provide insufficient information to plan adaptive responses; local decisions require local projections that accommodate different risk tolerances and time frames and that can be linked to storm surge projections. Here we present a global set of local sea-level (LSL) projections to inform decisions on timescales ranging from the coming decades through the 22nd century. We provide complete probability distributions, informed by a combination of expert community assessment, expert elicitation, and process modeling. Between the years 2000 and 2100, we project a very likely (90% probability) GSL rise of 0.5–1.2 m under representative concentration pathway (RCP) 8.5, 0.4–0.9 m under RCP 4.5, and 0.3–0.8 m under RCP 2.6. Site-to-site differences in LSL projections are due to varying non-climatic background uplift or subsidence, oceanographic effects, and spatially variable responses of the geoid and the lithosphere to shrinking land ice. The Antarctic ice sheet (AIS) constitutes a growing share of variance in GSL and LSL projections. In the global average and at many locations, it is the dominant source of variance in late 21st century projections, though at some sites oceanographic processes contribute the largest share throughout the century. LSL rise dramatically reshapes flood risk, greatly increasing the expected number of “1-in-10” and “1-in-100” year events.;,citation_author=Robert E. Kopp;,citation_author=Radley M. Horton;,citation_author=Christopher M. Little;,citation_author=Jerry X. Mitrovica;,citation_author=Michael Oppenheimer;,citation_author=D. J. Rasmussen;,citation_author=Benjamin H. Strauss;,citation_author=Claudia Tebaldi;,citation_publication_date=2014;,citation_cover_date=2014;,citation_year=2014;,citation_fulltext_html_url=http://agupubs.onlinelibrary.wiley.com/doi/abs/10.1002/2014EF000239;,citation_issue=8;,citation_doi=10.1002/2014ef000239;,citation_issn=2328-4277;,citation_volume=2;,citation_journal_title=Earth’s Future;">
<meta name="citation_reference" content="citation_title=Your Brain on ChatGPT: Accumulation of Cognitive Debt when Using an AI Assistant for Essay Writing Task;,citation_abstract=This study explores the neural and behavioral consequences of LLM-assisted essay writing. Participants were divided into three groups: LLM, Search Engine, and Brain-only (no tools). Each completed three sessions under the same condition. In a fourth session, LLM users were reassigned to Brain-only group (LLM-to-Brain), and Brain-only users were reassigned to LLM condition (Brain-to-LLM). A total of 54 participants took part in Sessions 1-3, with 18 completing session 4. We used electroencephalography (EEG) to assess cognitive load during essay writing, and analyzed essays using NLP, as well as scoring essays with the help from human teachers and an AI judge. Across groups, NERs, n-gram patterns, and topic ontology showed within-group homogeneity. EEG revealed significant differences in brain connectivity: Brain-only participants exhibited the strongest, most distributed networks; Search Engine users showed moderate engagement; and LLM users displayed the weakest connectivity. Cognitive activity scaled down in relation to external tool use. In session 4, LLM-to-Brain participants showed reduced alpha and beta connectivity, indicating under-engagement. Brain-to-LLM users exhibited higher memory recall and activation of occipito-parietal and prefrontal areas, similar to Search Engine users. Self-reported ownership of essays was the lowest in the LLM group and the highest in the Brain-only group. LLM users also struggled to accurately quote their own work. While LLMs offer immediate convenience, our findings highlight potential cognitive costs. Over four months, LLM users consistently underperformed at neural, linguistic, and behavioral levels. These results raise concerns about the long-term educational implications of LLM reliance and underscore the need for deeper inquiry into AI’s role in learning.;,citation_author=Nataliya Kosmyna;,citation_author=Eugene Hauptmann;,citation_author=Ye Tong Yuan;,citation_author=Jessica Situ;,citation_author=Xian-Hao Liao;,citation_author=Ashly Vivian Beresnitzky;,citation_author=Iris Braunstein;,citation_author=Pattie Maes;,citation_publication_date=2025-06-10;,citation_cover_date=2025-06-10;,citation_year=2025;,citation_fulltext_html_url=http://arxiv.org/abs/2506.08872;,citation_doi=10.48550/arXiv.2506.08872;">
<meta name="citation_reference" content="citation_title=Downscaling and bias-correction contribute considerable uncertainty to local climate projections in CMIP6;,citation_abstract=Efforts to diagnose the risks of a changing climate often rely on downscaled and bias-corrected climate information, making it important to understand the uncertainties and potential biases of this approach. Here, we perform a variance decomposition to partition uncertainty in global climate projections and quantify the relative importance of downscaling and bias-correction. We analyze simple climate metrics such as annual temperature and precipitation averages, as well as several indices of climate extremes. We find that downscaling and bias-correction often contribute substantial uncertainty to local decision-relevant climate outcomes, though our results are strongly heterogeneous across space, time, and climate metrics. Our results can provide guidance to impact modelers and decision-makers regarding the uncertainties associated with downscaling and bias-correction when performing local-scale analyses, as neglecting to account for these uncertainties may risk overconfidence relative to the full range of possible climate futures.;,citation_author=David C. Lafferty;,citation_author=Ryan L. Sriver;,citation_publication_date=2023-09-30;,citation_cover_date=2023-09-30;,citation_year=2023;,citation_fulltext_html_url=https://www.nature.com/articles/s41612-023-00486-0;,citation_issue=1, 1;,citation_doi=10.1038/s41612-023-00486-0;,citation_issn=2397-3722;,citation_volume=6;,citation_journal_title=npj Climate and Atmospheric Science;,citation_journal_abbrev=npj Clim Atmos Sci;,citation_publisher=Nature Publishing Group;">
<meta name="citation_reference" content="citation_title=Chapter 3: Water;,citation_author=Upmanu Lall;,citation_author=Thomas Johnson;,citation_author=Peter Colohan;,citation_author=Amir Aghakouchak;,citation_author=Sankar Arumugam;,citation_author=Casey Brown;,citation_author=Gregory J. Mccabe;,citation_author=Roger S. Pulwarty;,citation_publication_date=2018;,citation_cover_date=2018;,citation_year=2018;,citation_fulltext_html_url=https://nca2018.globalchange.gov/chapter/3/;,citation_doi=10.7930/NCA4.2018.CH3;,citation_inbook_title=Impacts, Risks, and Adaptation in the United States: The Fourth National Climate Assessment, Volume II;">
<meta name="citation_reference" content="citation_title=Robust abatement pathways to tolerable climate futures require immediate global action;,citation_abstract=Uncertainties are often cited as a reason for mitigation inaction. Here, millions of scenarios are evaluated to assess the relative importance of human–earth system uncertainties and policy variables. The growth rate of global abatement is found to be the primary driver of long-term warming.;,citation_author=J. R. Lamontagne;,citation_author=P. M. Reed;,citation_author=G. Marangoni;,citation_author=K. Keller;,citation_author=G. G. Garner;,citation_publication_date=2019-04;,citation_cover_date=2019-04;,citation_year=2019;,citation_fulltext_html_url=http://www.nature.com/articles/s41558-019-0426-8;,citation_issue=4;,citation_doi=10.1038/s41558-019-0426-8;,citation_issn=1758-6798;,citation_volume=9;,citation_journal_title=Nature Climate Change;,citation_journal_abbrev=Nat. Clim. Chang.;">
<meta name="citation_reference" content="citation_title=Some Pitfalls in Statistical Downscaling of Future Climate;,citation_author=John R Lanzante;,citation_author=Keith W Dixon;,citation_author=Mary Jo Nath;,citation_author=Carolyn E Whitlock;,citation_author=Dennis Adams-Smith;,citation_publication_date=2018-04;,citation_cover_date=2018-04;,citation_year=2018;,citation_fulltext_html_url=http://journals.ametsoc.org/doi/10.1175/BAMS-D-17-0046.1;,citation_issue=4;,citation_doi=10.1175/bams-d-17-0046.1;,citation_volume=99;,citation_journal_title=Bulletin of the American Meteorological Society;">
<meta name="citation_reference" content="citation_title=Think Julia: How to think like a computer scientist;,citation_author=Ben Lauwens;,citation_author=Allen B Downey;,citation_publication_date=2019;,citation_cover_date=2019;,citation_year=2019;,citation_fulltext_html_url=https://benlauwens.github.io/ThinkJulia.jl/latest/book.html;,citation_isbn=1-4920-4500-4;">
<meta name="citation_reference" content="citation_title=The Impact of Neglecting Climate Change and Variability on ERCOT’s Forecasts of Electricity Demand in Texas;,citation_abstract=Abstract The Electric Reliability Council of Texas (ERCOT) manages the electric power across most of Texas. They make short-term assessments of electricity demand on the basis of historical weather over the last two decades, thereby ignoring the effects of climate change and the possibility of weather variability outside the recent historical range. In this paper, we develop an empirical method to predict the impact of weather on energy demand. We use that with a large ensemble of climate model runs to construct a probability distribution of power demand on the ERCOT grid for summer and winter 2021. We find that the most severe weather events will use 100% of available power—if anything goes wrong, as it did during the 2021 winter, there will not be sufficient available power. More quantitatively, we estimate a 5% chance that maximum power demand would be within 4.3 and 7.9 GW of ERCOT’s estimate of best-case available resources during summer and winter 2021, respectively, and a 20% chance it would be within 7.1 and 17 GW. The shortage of power on the ERCOT grid is partially hidden by the fact that ERCOTs seasonal assessments, which are based entirely on historical weather, are too low. Prior to the 2021 winter blackout, ERCOT forecast an extreme peak load of 67 GW. In reality, we estimate hourly peak demand was 82 GW, 22% above ERCOT’s most extreme forecast and about equal to the best-case available power. Given the high stakes, ERCOT should develop probabilistic estimates using modern scientific tools to predict the range of power demand more accurately.;,citation_author=Jangho Lee;,citation_author=Andrew E. Dessler;,citation_publication_date=2022-04-01;,citation_cover_date=2022-04-01;,citation_year=2022;,citation_fulltext_html_url=http://journals.ametsoc.org/view/journals/wcas/14/2/WCAS-D-21-0140.1.xml;,citation_issue=2;,citation_doi=10.1175/WCAS-D-21-0140.1;,citation_issn=1948-8327, 1948-8335;,citation_volume=14;,citation_journal_title=Weather, Climate, and Society;,citation_publisher=American Meteorological Society;">
<meta name="citation_reference" content="citation_title=Robust strategies for abating climate change;,citation_abstract=The need for climate-change decisionmakers to craft strategies that are robust in the face of an unpredictable future is addressed. The climate-change policymaking community could do policymakers a great service by examining signposts that might provide a better basis for building near-term climate-change policies.;,citation_author=Robert J. Lempert;,citation_author=Michael E. Schlesinger;,citation_publication_date=2000-06;,citation_cover_date=2000-06;,citation_year=2000;,citation_fulltext_html_url=https://www.proquest.com/docview/198492640/abstract/7534863C60614053PQ/1;,citation_issue=3–4;,citation_doi=10.1023/A:1005698407365;,citation_issn=01650009;,citation_volume=45;,citation_journal_title=Climatic Change;,citation_publisher=Springer Nature B.V.;">
<meta name="citation_reference" content="citation_title=Computing return times or return periods with rare event algorithms;,citation_abstract=The average time between two occurrences of the same event, referred to as its return time (or return period), is a useful statistical concept for practical applications. For instance insurances or public agencies may be interested by the return time of a 10 m flood of the Seine river in Paris. However, due to their scarcity, reliably estimating return times for rare events is very difficult using either observational data or direct numerical simulations. For rare events, an estimator for return times can be built from the extrema of the observable on trajectory blocks. Here, we show that this estimator can be improved to remain accurate for return times of the order of the block size. More importantly, we show that this approach can be generalised to estimate return times from numerical algorithms specifically designed to sample rare events. So far those algorithms often compute probabilities, rather than return times. The approach we propose provides a computationally extremely efficient way to estimate numerically the return times of rare events for a dynamical system, gaining several orders of magnitude of computational costs. We illustrate the method on two kinds of observables, instantaneous and time-averaged, using two different rare event algorithms, for a simple stochastic process, the Ornstein–Uhlenbeck process. As an example of realistic applications to complex systems, we finally discuss extreme values of the drag on an object in a turbulent flow.;,citation_author=Thibault Lestang;,citation_author=Francesco Ragone;,citation_author=Charles-Edouard Bréhier;,citation_author=Corentin Herbert;,citation_author=Freddy Bouchet;,citation_publication_date=2018-04;,citation_cover_date=2018-04;,citation_year=2018;,citation_fulltext_html_url=https://dx.doi.org/10.1088/1742-5468/aab856;,citation_issue=4;,citation_doi=10.1088/1742-5468/aab856;,citation_issn=1742-5468;,citation_volume=2018;,citation_journal_title=Journal of Statistical Mechanics: Theory and Experiment;,citation_journal_abbrev=J. Stat. Mech.;,citation_publisher=IOP Publishing and SISSA;">
<meta name="citation_reference" content="citation_title=Water resource systems planning and management: An introduction to methods, models, and applications;,citation_author=Daniel P. Loucks;,citation_publication_date=2017;,citation_cover_date=2017;,citation_year=2017;,citation_isbn=978-3-319-44234-1;">
<meta name="citation_reference" content="citation_title=Bayesian spatiotemporal nonstationary model quantifies robust increases in daily extreme rainfall across the Western Gulf Coast;,citation_abstract=Precipitation exceedance probabilities are widely used in engineering design, risk assessment, and floodplain management. While common approaches like National Oceanographic and Atmospheric Administration (NOAA) Atlas 14 assume that extreme precipitation characteristics are stationary over time, this assumption may underestimate current and future hazards due to anthropogenic climate change. However, the incorporation of nonstationarity in statistical modeling of extreme precipitation has faced practical challenges, which have restricted its applications. In particular, random sampling variability challenges the reliable estimation of trends and parameters, especially when observational records are limited. To address this methodological gap, we propose the Spatially Varying Covariates Model, a hierarchical Bayesian spatial framework that integrates nonstationarity and regionalization for robust frequency analysis of extreme precipitation. This model draws from extreme value theory, spatial statistics, and Bayesian statistics, and is validated through cross-validation and multiple performance metrics. Applying this framework to a case study of daily rainfall in the Western Gulf Coast, we identify robustly increasing trends in extreme precipitation intensity and variability throughout the study area, with notable spatial heterogeneity. This flexible model accommodates stations with varying observation records, yields smooth return level estimates, and can be straightforwardly adapted to the analysis of precipitation frequencies at different durations and for other regions.;,citation_author=Yuchen Lu;,citation_author=Benjamin Seiyon Lee;,citation_author=James Doss-Gollin;,citation_publication_date=2025-08;,citation_cover_date=2025-08;,citation_year=2025;,citation_fulltext_html_url=https://dx.doi.org/10.1088/2752-5295/adf56e;,citation_issue=3;,citation_doi=10.1088/2752-5295/adf56e;,citation_issn=2752-5295;,citation_volume=4;,citation_journal_title=Environmental Research: Climate;,citation_journal_abbrev=Environ. Res.: Climate;,citation_publisher=IOP Publishing;">
<meta name="citation_reference" content="citation_title=Plotting Positions in Extreme Value Analysis;,citation_abstract=Plotting order-ranked data is a standard technique that is used in estimating the probability of extreme weather events. Typically, observations, say, annual extremes of a period of N years, are ranked in order of magnitude and plotted on probability paper. Some statistical model is then fitted to the order-ranked data by which the return periods of specific extreme events are estimated. A key question in this method is as follows: What is the cumulative probability P that should be associated with the sample of rank m? This issue of the so-called plotting positions has been debated for almost a century, and a number of plotting rules and computational methods have been proposed. Here, it is shown that in estimating the return periods there is only one correct plotting position: P = m/(N + 1). This formula predicts much shorter return periods of extreme events than the other commonly used methods. Thus, many estimates of the weather-related risks should be reevaluated and the related building codes and other related regulations updated.;,citation_author=Lasse Makkonen;,citation_publication_date=2006-02-01;,citation_cover_date=2006-02-01;,citation_year=2006;,citation_fulltext_html_url=https://journals.ametsoc.org/view/journals/apme/45/2/jam2349.1.xml;,citation_issue=2;,citation_doi=10.1175/JAM2349.1;,citation_issn=1558-8424, 1558-8432;,citation_volume=45;,citation_journal_title=Journal of Applied Meteorology and Climatology;,citation_publisher=American Meteorological Society;">
<meta name="citation_reference" content="citation_title=Decision making under deep uncertainty: From theory to practice;,citation_editor=Vincent A. W. J. Marchau;,citation_editor=Warren E. Walker;,citation_editor=Pieter J. T. M. Bloemen;,citation_editor=Steven W. Popper;,citation_publication_date=2019;,citation_cover_date=2019;,citation_year=2019;,citation_isbn=978-3-030-05251-5;">
<meta name="citation_reference" content="citation_title=6. Water-Resource Systems Planning.;,citation_author=Nicholas C Matalas;,citation_author=Myron B Fiering;,citation_publication_date=1977;,citation_cover_date=1977;,citation_year=1977;,citation_fulltext_html_url=https://www.nap.edu/read/185/chapter/11;,citation_inbook_title=Climate, Climatic Change, and Water Supply;">
<meta name="citation_reference" content="citation_title=Statistical rethinking: A Bayesian course with examples in R and Stan;,citation_author=Richard McElreath;,citation_publication_date=2020;,citation_cover_date=2020;,citation_year=2020;,citation_isbn=978-0-429-63914-2;,citation_series_title=Texts in statistical science series;">
<meta name="citation_reference" content="citation_title=Robustness metrics: How are they calculated, when should they be used and why do they give different results?;,citation_abstract=Robustness is being used increasingly for decision analysis in relation to deep uncertainty and many metrics have been proposed for its quantification. Recent studies have shown that the application of different robustness metrics can result in different rankings of decision alternatives, but there has been little discussion of what potential causes for this might be. To shed some light on this issue, we present a unifying framework for the calculation of robustness metrics, which assists with understanding how robustness metrics work, when they should be used, and why they sometimes disagree. The framework categorizes the suitability of metrics to a decision-maker based on (1) the decision-context (i.e., the suitability of using absolute performance or regret), (2) the decision-maker’s preferred level of risk aversion, and (3) the decision-maker’s preference toward maximizing performance, minimizing variance, or some higher-order moment. This article also introduces a conceptual framework describing when relative robustness values of decision alternatives obtained using different metrics are likely to agree and disagree. This is used as a measure of how ?stable? the ranking of decision alternatives is when determined using different robustness metrics. The framework is tested on three case studies, including water supply augmentation in Adelaide, Australia, the operation of a multipurpose regulated lake in Italy, and flood protection for a hypothetical river based on a reach of the river Rhine in the Netherlands. The proposed conceptual framework is confirmed by the case study results, providing insight into the reasons for disagreements between rankings obtained using different robustness metrics.;,citation_author=C. McPhail;,citation_author=H. R. Maier;,citation_author=J. H. Kwakkel;,citation_author=M. Giuliani;,citation_author=A. Castelletti;,citation_author=S. Westra;,citation_publication_date=2019-04-02;,citation_cover_date=2019-04-02;,citation_year=2019;,citation_fulltext_html_url=http://agupubs.onlinelibrary.wiley.com/doi/abs/10.1002/2017EF000649%4010.1002/%28ISSN%292328-4277.RESDEC1;,citation_doi=10.1002/2017ef000649;,citation_issn=2328-4277;,citation_journal_title=Earth’s Future;,citation_journal_abbrev=Earth’s Future;">
<meta name="citation_reference" content="citation_title=Understanding heavy tails of flood peak distributions;,citation_abstract=Statistical distributions of flood peak discharge often show heavy tail behavior, that is, extreme floods are more likely to occur than would be predicted by commonly used distributions that have exponential asymptotic behavior. This heavy tail behavior may surprise flood managers and citizens, as human intuition tends to expect light tail behavior, and the heaviness of the tails is very difficult to predict, which may lead to unnecessarily high flood damage. Despite its high importance, the literature on the heavy tail behavior of flood distributions is rather fragmented. In this review, we provide a coherent overview of the processes causing heavy flood tails and the implications for science and practice. Specifically, we propose nine hypotheses on the mechanisms causing heavy tails in flood peak distributions related to processes in the atmosphere, the catchment, and the river system. We then discuss to which extent the current knowledge supports or contradicts these hypotheses. We also discuss the statistical conditions for the emergence of heavy tail behavior based on derived distribution theory and relate them to the hypotheses and flood generation mechanisms. We review the degree to which the heaviness of the tails can be predicted from process knowledge and data. Finally, we recommend further research toward testing the hypotheses and improving the prediction of heavy tails.;,citation_author=Bruno Merz;,citation_author=Stefano Basso;,citation_author=Svenja Fischer;,citation_author=David Lun;,citation_author=Günter Blöschl;,citation_author=Ralf Merz;,citation_author=Björn Guse;,citation_author=Alberto Viglione;,citation_author=Sergiy Vorogushyn;,citation_author=Elena Macdonald;,citation_author=Luzie Wietzke;,citation_author=Andreas Schumann;,citation_publication_date=2022;,citation_cover_date=2022;,citation_year=2022;,citation_fulltext_html_url=http://onlinelibrary.wiley.com/doi/abs/10.1029/2021WR030506;,citation_issue=6;,citation_doi=10.1029/2021WR030506;,citation_issn=1944-7973;,citation_volume=58;,citation_journal_title=Water Resources Research;">
<meta name="citation_reference" content="citation_title=Floods and climate: Emerging perspectives for flood risk assessment and management;,citation_abstract=Flood estimation and flood management have traditionally been the domain of hydrologists, water resources engineers and statisticians, and disciplinary approaches abound. Dominant views have been shaped; one example is the catchment perspective: floods are formed and influenced by the interaction of local, catchment-specific characteristics, such as meteorology, topography and geology. These traditional views have been beneficial, but they have a narrow framing. In this paper we contrast traditional views with broader perspectives that are emerging from an improved understanding of the climatic context of floods. We come to the following conclusions: (1) extending the traditional system boundaries (local catchment, recent decades, hydrological/hydraulic processes) opens up exciting possibilities for better understanding and improved tools for flood risk assessment and management. (2) Statistical approaches in flood estimation need to be complemented by the search for the causal mechanisms and dominant processes in the atmosphere, catchment and river system that leave their fingerprints on flood characteristics. (3) Natural climate variability leads to time-varying flood characteristics, and this variation may be partially quantifiable and predictable, with the perspective of dynamic, climate-informed flood risk management. (4) Efforts are needed to fully account for factors that contribute to changes in all three risk components (hazard, exposure, vulnerability) and to better understand the interactions between society and floods. (5) Given the global scale and societal importance, we call for the organization of an international multidisciplinary collaboration and data-sharing initiative to further understand the links between climate and flooding and to advance flood research.;,citation_author=Bruno Merz;,citation_author=Jeroen C J H Aerts;,citation_author=Karsten Arnbjerg-Nielsen;,citation_author=M Baldi;,citation_author=A Becker;,citation_author=A Bichet;,citation_author=Günter Blöschl;,citation_author=Laurens M Bouwer;,citation_author=Achim Brauer;,citation_author=F Cioffi;,citation_author=J M Delgado;,citation_author=M Gocht;,citation_author=F Guzzetti;,citation_author=S Harrigan;,citation_author=K Hirschboeck;,citation_author=C Kilsby;,citation_author=W Kron;,citation_author=H H Kwon;,citation_author=Upmanu Lall;,citation_author=R Merz;,citation_author=K Nissen;,citation_author=P Salvatti;,citation_author=Tina Swierczynski;,citation_author=U Ulbrich;,citation_author=A Viglione;,citation_author=P J Ward;,citation_author=M Weiler;,citation_author=B Wilhelm;,citation_author=M Nied;,citation_publication_date=2014;,citation_cover_date=2014;,citation_year=2014;,citation_fulltext_html_url=http://www.nat-hazards-earth-syst-sci.net/14/1921/2014/;,citation_issue=7;,citation_doi=10/gb9nzm;,citation_volume=14;,citation_journal_title=Natural Hazards and Earth System Science;">
<meta name="citation_reference" content="citation_title=Introduction to Catastrophe Risk Modelling: A Physics-based Approach;,citation_abstract=Focusing on the physics of the&nbsp;catastrophe process and addressed directly to advanced students, this innovative textbook quantifies dozens of perils, both natural and man-made, and covers the latest developments in catastrophe modelling. Combining basic statistics, applied physics, natural and environmental sciences, civil engineering, and psychology, the text remains at an introductory level, focusing on fundamental concepts for a comprehensive understanding of catastrophe phenomenology and risk quantification. A broad spectrum of perils are covered, including geophysical, hydrological, meteorological, climatological, biological, extraterrestrial, technological and socio-economic, as well as events caused by domino effects and global warming. Following industry standards, the text provides the necessary tools to develop a CAT model from hazard to loss assessment. Online resources include a&nbsp;CAT risk model starter-kit and a CAT risk modelling ’sandbox’ with Python Jupyter tutorial.&nbsp;Every process, described by equations, (pseudo)codes and illustrations, is fully reproducible, allowing students to solidify knowledge through practice.;,citation_author=Arnaud Mignan;,citation_publication_date=2024;,citation_cover_date=2024;,citation_year=2024;,citation_fulltext_html_url=https://www.cambridge.org/core/product/A3A5B5FB990921422BFEBB07734BF869;,citation_doi=10.1017/9781009437370;,citation_isbn=978-1-009-43738-7;">
<meta name="citation_reference" content="citation_title=Stationarity is dead: Whither water management?;,citation_abstract=Climate change undermines a basic assumption that historically has facilitated management of water supplies, demands, and risks.;,citation_author=P C D Milly;,citation_author=Julio Betancourt;,citation_author=M Falkenmark;,citation_author=R M Hirsch;,citation_author=Z W Kundzewicz;,citation_author=D P Lettenmaier;,citation_author=R J Stouffer;,citation_publication_date=2008-02;,citation_cover_date=2008-02;,citation_year=2008;,citation_fulltext_html_url=http://www.sciencemag.org/cgi/doi/10.1126/science.1151915;,citation_issue=5863;,citation_doi=10.1126/science.1151915;,citation_volume=319;,citation_journal_title=Science;">
<meta name="citation_reference" content="citation_title=Statistical analysis of climate extremes / Manfred Mudelsee.;,citation_abstract=&amp;amp;amp;quot;The big climate question is how climate change affects climate extremes. More hurricanes such as Katrina in 2005? More floods such as that of European river Elbe in 2002? More heatwaves such as in 2003 or 2018? Where to invest resources? All this is not just scientifically challenging. It is also relevant for society and economy to survive. This is the first textbook on statistics and climate extremes. It explains the statistical methods in an accessible language. It provides the necessary software. Case studies on extremes in the three major climate variables (temperature, precipitation and wind speed) show how to use the methods. The book provides the datasets to allow replication of case study calculations. This book is written for students and researchers in climate sciences. It can serve as textbook in university courses. Also risk analysts in the insurance industry benefit from it&amp;quot;– Provided by publisher.;,citation_author=Manfred Mudelsee;,citation_publication_date=2020;,citation_cover_date=2020;,citation_year=2020;,citation_isbn=978-1-108-79146-5;,citation_inbook_title=Statistical analysis of climate extremes;">
<meta name="citation_reference" content="citation_title=Fundamentals of Statistical Hydrology;,citation_editor=Mauro Naghettini;,citation_publication_date=2017;,citation_cover_date=2017;,citation_year=2017;,citation_fulltext_html_url=http://link.springer.com/10.1007/978-3-319-43561-9;,citation_doi=10.1007/978-3-319-43561-9;,citation_isbn=978-3-319-43560-2 978-3-319-43561-9;">
<meta name="citation_reference" content="citation_title=Statistics with Julia: Fundamentals for Data Science, Machine Learning and Artificial Intelligence;,citation_abstract=This monograph uses the Julia language to guide the reader through an exploration of the fundamental concepts of probability and statistics, all with a view of mastering machine learning, data science, and artificial intelligence. The text does not require any prior statistical knowledge and only assumes a basic understanding of programming and mathematical notation. It is accessible to practitioners and researchers in data science, machine learning, bio-statistics, finance, or engineering who may wish to solidify their knowledge of probability and statistics. The book progresses through ten independent chapters starting with an introduction of Julia, and moving through basic probability, distributions, statistical inference, regression analysis, machine learning methods, and the use of Monte Carlo simulation for dynamic stochastic models. Ultimately this text introduces the Julia programming language as a computational tool, uniquely addressing end-users rather than developers. It makes heavy use of over 200 code examples to illustrate dozens of key statistical concepts. The Julia code, written in a simple format with parameters that can be easily modified, is also available for download from the book’s associated GitHub repository online.See what co-creators of the Julia language are saying about the book:Professor Alan Edelman, MIT: With “Statistics with Julia”, Yoni and Hayden have written an easy to read, well organized, modern introduction to statistics. The code may be looked at, and understood on the static pages of a book, or even better, when running live on a computer. Everything you need is here in one nicely written self-contained reference. Dr. Viral Shah, CEO of Julia Computing: Yoni and Hayden provide a modern way to learn statistics with the Julia programming language. This book has been perfected through iteration over several semesters in the classroom. It prepares the reader with two complementary skills - statistical reasoning with hands on experience and working with large datasets through training in Julia.;,citation_author=Yoni Nazarathy;,citation_author=Hayden Klok;,citation_publication_date=2021;,citation_cover_date=2021;,citation_year=2021;,citation_fulltext_html_url=https://www.springer.com/gp/book/9783030709006;,citation_doi=10.1007/978-3-030-70901-3;,citation_isbn=978-3-030-70900-6;,citation_series_title=Springer Series in the Data Sciences;">
<meta name="citation_reference" content="citation_title=Deep uncertainties in sea-level rise and storm surge projections: Implications for coastal flood risk management;,citation_abstract=Sea levels are rising in many areas around the world, posing risks to coastal communities and infrastructures. Strategies for managing these flood risks present decision challenges that require a combination of geophysical, economic, and infrastructure models. Previous studies have broken important new ground on the considerable tensions between the costs of upgrading infrastructure and the damages that could result from extreme flood events. However, many risk-based adaptation strategies remain silent on certain potentially important uncertainties, as well as the tradeoffs between competing objectives. Here, we implement and improve on a classic decision-analytical model (Van Dantzig 1956) to: (i) capture tradeoffs across conflicting stakeholder objectives, (ii) demonstrate the consequences of structural uncertainties in the sea-level rise and storm surge models, and (iii) identify the parametric uncertainties that most strongly influence each objective using global sensitivity analysis. We find that the flood adaptation model produces potentially myopic solutions when formulated using traditional mean-centric decision theory. Moving from a single-objective problem formulation to one with multiobjective tradeoffs dramatically expands the decision space, and highlights the need for compromise solutions to address stakeholder preferences. We find deep structural uncertainties that have large effects on the model outcome, with the storm surge parameters accounting for the greatest impacts. Global sensitivity analysis effectively identifies important parameter interactions that local methods overlook, and that could have critical implications for flood adaptation strategies.;,citation_author=Perry C. Oddo;,citation_author=Ben S. Lee;,citation_author=Gregory G. Garner;,citation_author=Vivek Srikrishnan;,citation_author=Patrick M. Reed;,citation_author=Chris E. Forest;,citation_author=Klaus Keller;,citation_publication_date=2017;,citation_cover_date=2017;,citation_year=2017;,citation_fulltext_html_url=http://onlinelibrary.wiley.com/doi/abs/10.1111/risa.12888;,citation_issue=0;,citation_doi=10/ghkp82;,citation_issn=1539-6924;,citation_volume=0;,citation_journal_title=Risk Analysis;">
<meta name="citation_reference" content="citation_title=Verification, validation, and confirmation of numerical models in the Earth sciences;,citation_abstract=Verification and validation of numerical models of natural systems is impossible. This is because natural systems are never closed and because model results are always nonunique. Models can be confirmed by the demonstration of agreement between ...;,citation_author=Naomi Oreskes;,citation_author=Kristin Shrader-Frechette;,citation_author=Kenneth Belitz;,citation_publication_date=1994-02-04;,citation_cover_date=1994-02-04;,citation_year=1994;,citation_fulltext_html_url=http://www.science.org/doi/abs/10.1126/science.263.5147.641;,citation_doi=10.1126/science.263.5147.641;,citation_journal_title=Science;,citation_publisher=American Association for the Advancement of Science;">
<meta name="citation_reference" content="citation_title=The geography of future water challenges;,citation_author=PBL Netherlands Environmental Assessment Agency;,citation_publication_date=2019;,citation_cover_date=2019;,citation_year=2019;,citation_fulltext_html_url=https://www.pbl.nl/en/publications/the-geography-of-future-water-challenges;,citation_technical_report_institution=PBL Netherlands Environmental Assessment Agency;">
<meta name="citation_reference" content="citation_title=Comparison of Bayesian predictive methods for model selection;,citation_abstract=The goal of this paper is to compare several widely used Bayesian model selection methods in practical model selection problems, highlight their differences and give recommendations about the preferred approaches. We focus on the variable subset selection for regression and classification and perform several numerical experiments using both simulated and real world data. The results show that the optimization of a utility estimate such as the cross-validation (CV) score is liable to finding overfitted models due to relatively high variance in the utility estimates when the data is scarce. This can also lead to substantial selection induced bias and optimism in the performance evaluation for the selected model. From a predictive viewpoint, best results are obtained by accounting for model uncertainty by forming the full encompassing model, such as the Bayesian model averaging solution over the candidate models. If the encompassing model is too complex, it can be robustly simplified by the projection method, in which the information of the full model is projected onto the submodels. This approach is substantially less prone to overfitting than selection based on CV-score. Overall, the projection method appears to outperform also the maximum a posteriori model and the selection of the most probable variables. The study also demonstrates that the model selection can greatly benefit from using cross-validation outside the searching process both for guiding the model size selection and assessing the predictive performance of the finally selected model. Published in: Statistics and Computing, 2017, Volume 27, Issue 3, 711-735;,citation_author=Juho Piironen;,citation_author=Aki Vehtari;,citation_publication_date=2017-05;,citation_cover_date=2017-05;,citation_year=2017;,citation_fulltext_html_url=http://link.springer.com/10.1007/s11222-016-9649-y;,citation_issue=3;,citation_doi=10.1007/s11222-016-9649-y;,citation_volume=27;,citation_journal_title=Statistics and Computing;">
<meta name="citation_reference" content="citation_title=Sequential Decision Analytics and Modeling: Modeling with Python;,citation_abstract=Publishers of Foundations and Trends, making research accessible;,citation_author=Warren B. Powell;,citation_publication_date=2022-11-21;,citation_cover_date=2022-11-21;,citation_year=2022;,citation_fulltext_html_url=https://nowpublishers.com/Article/BookDetails/9781638280828;,citation_issn=978-1-63828-082-8, 978-1-63828-083-5;,citation_publisher=Now Publishers, Inc.;">
<meta name="citation_reference" content="citation_title=Increasing the accuracy and resolution of precipitation forecasts using deep generative models;,citation_abstract=Accurately forecasting extreme rainfall is notoriously difficult, but is also ever more crucial for society as climate change increases the frequency of such extremes. Global numerical weather prediction models often fail to capture extremes, and are produced at too low a resolution to be actionable, while regional, high-resolution models are hugely expensive both in computation and labour. In this paper we explore the use of deep generative models to simultaneously correct and downscale (super-resolve) global ensemble forecasts over the Continental US. Specifically, using fine-grained radar observations as our ground truth, we train a conditional Generative Adversarial Network – coined CorrectorGAN – via a custom training procedure and augmented loss function, to produce ensembles of high-resolution, bias-corrected forecasts based on coarse, global precipitation forecasts in addition to other relevant meteorological fields. Our model outperforms an interpolation baseline, as well as super-resolution-only and CNN-based univariate methods, and approaches the performance of an operational regional high-resolution model across an array of established probabilistic metrics. Crucially, CorrectorGAN, once trained, produces predictions in seconds on a single machine. These results raise exciting questions about the necessity of regional models, and whether data-driven downscaling and correction methods can be transferred to data-poor regions that so far have had no access to high-resolution forecasts.;,citation_author=Ilan Price;,citation_author=Stephan Rasp;,citation_publication_date=2022-03-23;,citation_cover_date=2022-03-23;,citation_year=2022;,citation_fulltext_html_url=http://arxiv.org/abs/2203.12297;,citation_doi=10.48550/arXiv.2203.12297;">
<meta name="citation_reference" content="citation_title=Applied Machine Learning in Python: A Hands-on Guide with Code;,citation_author=Michael J. Pyrcz;,citation_publication_date=2024;,citation_cover_date=2024;,citation_year=2024;,citation_fulltext_html_url=https://geostatsguy.github.io/MachineLearningDemos_Book;">
<meta name="citation_reference" content="citation_title=Direct policy search for robust multi-objective management of deeply uncertain socio-ecological tipping points;,citation_abstract=Managing socio-ecological systems is a challenge wrought by competing societal objectives, deep uncertainties, and potentially irreversible tipping points. A classic, didactic example is the shallow lake problem in which a hypothetical town situated on a lake must develop pollution control strategies to maximize its economic benefits while minimizing the probability of the lake crossing a critical phosphorus (P) threshold, above which it irreversibly transitions into a eutrophic state. Here, we explore the use of direct policy search (DPS) to design robust pollution control rules for the town that account for deeply uncertain system characteristics and conflicting objectives. The closed loop control formulation of DPS improves the quality and robustness of key management tradeoffs, while dramatically reducing the computational complexity of solving the multi-objective pollution control problem relative to open loop control strategies. These insights suggest DPS is a promising tool for managing socio-ecological systems with deeply uncertain tipping points.;,citation_author=Julianne D. Quinn;,citation_author=Patrick M. Reed;,citation_author=Klaus Keller;,citation_publication_date=2017-06-01;,citation_cover_date=2017-06-01;,citation_year=2017;,citation_fulltext_html_url=http://www.sciencedirect.com/science/article/pii/S1364815216302250;,citation_doi=10.1016/j.envsoft.2017.02.017;,citation_issn=1364-8152;,citation_volume=92;,citation_journal_title=Environmental Modelling &amp;amp;amp; Software;,citation_journal_abbrev=Environmental Modelling &amp;amp; Software;">
<meta name="citation_reference" content="citation_title=Universal Differential Equations for Scientific Machine Learning;,citation_abstract=In the context of science, the well-known adage &amp;amp;amp;quot;a picture is worth a thousand words&amp;quot; might well be &quot;a model is worth a thousand datasets.&quot; In this manuscript we introduce the SciML software ecosystem as a tool for mixing the information of physical laws and scientific models with data-driven machine learning approaches. We describe a mathematical object, which we denote universal differential equations (UDEs), as the unifying framework connecting the ecosystem. We show how a wide variety of applications, from automatically discovering biological mechanisms to solving high-dimensional Hamilton-Jacobi-Bellman equations, can be phrased and efficiently handled through the UDE formalism and its tooling. We demonstrate the generality of the software tooling to handle stochasticity, delays, and implicit constraints. This funnels the wide variety of SciML applications into a core set of training mechanisms which are highly optimized, stabilized for stiff equations, and compatible with distributed parallelism and GPU accelerators.;,citation_author=Christopher Rackauckas;,citation_author=Yingbo Ma;,citation_author=Julius Martensen;,citation_author=Collin Warner;,citation_author=Kirill Zubov;,citation_author=Rohit Supekar;,citation_author=Dominic Skinner;,citation_author=Ali Ramadhan;,citation_author=Alan Edelman;,citation_publication_date=2020;,citation_cover_date=2020;,citation_year=2020;,citation_fulltext_html_url=https://arxiv.org/abs/2001.04385;,citation_doi=10.48550/ARXIV.2001.04385;">
<meta name="citation_reference" content="citation_title=The future of sensitivity analysis: An essential discipline for systems modeling and policy support;,citation_abstract=Sensitivity analysis (SA) is en route to becoming an integral part of mathematical modeling. The tremendous potential benefits of SA are, however, yet to be fully realized, both for advancing mechanistic and data-driven modeling of human and natural systems, and in support of decision making. In this perspective paper, a multidisciplinary group of researchers and practitioners revisit the current status of SA, and outline research challenges in regard to both theoretical frameworks and their applications to solve real-world problems. Six areas are discussed that warrant further attention, including (1) structuring and standardizing SA as a discipline, (2) realizing the untapped potential of SA for systems modeling, (3) addressing the computational burden of SA, (4) progressing SA in the context of machine learning, (5) clarifying the relationship and role of SA to uncertainty quantification, and (6) evolving the use of SA in support of decision making. An outlook for the future of SA is provided that underlines how SA must underpin a wide variety of activities to better serve science and society.;,citation_author=Saman Razavi;,citation_author=Anthony Jakeman;,citation_author=Andrea Saltelli;,citation_author=Clémentine Prieur;,citation_author=Bertrand Iooss;,citation_author=Emanuele Borgonovo;,citation_author=Elmar Plischke;,citation_author=Samuele Lo Piano;,citation_author=Takuya Iwanaga;,citation_author=William Becker;,citation_author=Stefano Tarantola;,citation_author=Joseph H. A. Guillaume;,citation_author=John Jakeman;,citation_author=Hoshin Gupta;,citation_author=Nicola Melillo;,citation_author=Giovanni Rabitti;,citation_author=Vincent Chabridon;,citation_author=Qingyun Duan;,citation_author=Xifu Sun;,citation_author=Stefán Smith;,citation_author=Razi Sheikholeslami;,citation_author=Nasim Hosseini;,citation_author=Masoud Asadzadeh;,citation_author=Arnald Puy;,citation_author=Sergei Kucherenko;,citation_author=Holger R. Maier;,citation_publication_date=2020-12-15;,citation_cover_date=2020-12-15;,citation_year=2020;,citation_fulltext_html_url=http://www.sciencedirect.com/science/article/pii/S1364815220310112;,citation_doi=10.1016/j.envsoft.2020.104954;,citation_issn=1364-8152;,citation_journal_title=Environmental Modelling &amp;amp;amp; Software;,citation_journal_abbrev=Environmental Modelling &amp;amp; Software;">
<meta name="citation_reference" content="citation_title=Evolutionary multiobjective optimization in water resources: The past, present, and future;,citation_abstract=This study contributes a rigorous diagnostic assessment of state-of-the-art multiobjective evolutionary algorithms (MOEAs) and highlights key advances that the water resources field can exploit to better discover the critical tradeoffs constraining our systems. This study provides the most comprehensive diagnostic assessment of MOEAs for water resources to date, exploiting more than 100,000 MOEA runs and trillions of design evaluations. The diagnostic assessment measures the effectiveness, efficiency, reliability, and controllability of ten benchmark MOEAs for a representative suite of water resources applications addressing rainfall–runoff calibration, long-term groundwater monitoring (LTM), and risk-based water supply portfolio planning. The suite of problems encompasses a range of challenging problem properties including (1) many-objective formulations with four or more objectives, (2) multi-modality (or false optima), (3) nonlinearity, (4) discreteness, (5) severe constraints, (6) stochastic objectives, and (7) non-separability (also called epistasis). The applications are representative of the dominant problem classes that have shaped the history of MOEAs in water resources and that will be dominant foci in the future. Recommendations are given for the new algorithms that should serve as the benchmarks for innovations in the water resources literature. The future of MOEAs in water resources needs to emphasize self-adaptive search, new technologies for visualizing tradeoffs, and the next generation of computing technologies.;,citation_author=P. M. Reed;,citation_author=D. Hadka;,citation_author=J. D. Herman;,citation_author=J. R. Kasprzyk;,citation_author=J. B. Kollat;,citation_publication_date=2013-01-01;,citation_cover_date=2013-01-01;,citation_year=2013;,citation_fulltext_html_url=http://www.sciencedirect.com/science/article/pii/S0309170812000073;,citation_doi=10.1016/j.advwatres.2012.01.005;,citation_issn=0309-1708;,citation_volume=51;,citation_journal_title=Advances in Water Resources;,citation_journal_abbrev=Advances in Water Resources;,citation_series_title=35th Year Anniversary Issue;">
<meta name="citation_reference" content="citation_title=Multisector Dynamics: Advancing the Science of Complex Adaptive Human-Earth Systems;,citation_abstract=The field of MultiSector Dynamics (MSD) explores the dynamics and co-evolutionary pathways of human and Earth systems with a focus on critical goods, services, and amenities delivered to people through interdependent sectors. This commentary lays out core definitions and concepts, identifies MSD science questions in the context of the current state of knowledge, and describes ongoing activities to expand capacities for open science, leverage revolutions in data and computing, and grow and diversify the MSD workforce. Central to our vision is the ambition of advancing the next generation of complex adaptive human-Earth systems science to better address interconnected risks, increase resilience, and improve sustainability. This will require convergent research and the integration of ideas and methods from multiple disciplines. Understanding the tradeoffs, synergies, and complexities that exist in coupled human-Earth systems is particularly important in the context of energy transitions and increased future shocks.;,citation_author=Patrick M. Reed;,citation_author=Antonia Hadjimichael;,citation_author=Richard H. Moss;,citation_author=Christa Brelsford;,citation_author=Casey D. Burleyson;,citation_author=Stuart Cohen;,citation_author=Ana Dyreson;,citation_author=David F. Gold;,citation_author=Rohini S. Gupta;,citation_author=Klaus Keller;,citation_author=Megan Konar;,citation_author=Erwan Monier;,citation_author=Jennifer Morris;,citation_author=Vivek Srikrishnan;,citation_author=Nathalie Voisin;,citation_author=Jim Yoon;,citation_publication_date=2022;,citation_cover_date=2022;,citation_year=2022;,citation_fulltext_html_url=http://onlinelibrary.wiley.com/doi/abs/10.1029/2021EF002621;,citation_issue=3;,citation_doi=10.1029/2021EF002621;,citation_issn=2328-4277;,citation_volume=10;,citation_journal_title=Earth’s Future;">
<meta name="citation_reference" content="citation_title=Techniques for assessing water infrastructure for nonstationary extreme events: A review;,citation_author=J D Salas;,citation_author=J Obeysekera;,citation_author=R M Vogel;,citation_publication_date=2018;,citation_cover_date=2018;,citation_year=2018;,citation_fulltext_html_url=https://www.tandfonline.com/doi/full/10.1080/02626667.2018.1426858;,citation_issue=3;,citation_doi=10.1080/02626667.2018.1426858;,citation_volume=63;,citation_journal_title=Hydrological Sciences Journal;">
<meta name="citation_reference" content="citation_title=Global sensitivity analysis: The primer;,citation_author=Andrea Saltelli;,citation_author=Marco Ratto;,citation_author=Terry Andres;,citation_author=Francesca Campolongo;,citation_author=Jessica Cariboni;,citation_author=Debora Gatelli;,citation_author=Michaela Saisana;,citation_author=Stefano Tarantola;,citation_publication_date=2008;,citation_cover_date=2008;,citation_year=2008;,citation_fulltext_html_url=http://onlinelibrary.wiley.com/doi/abs/10.1002/9780470725184.ch1;,citation_isbn=978-0-470-72518-4;">
<meta name="citation_reference" content="citation_title=A Gentle Introduction to Graph Neural Networks;,citation_abstract=What components are needed for building learning algorithms that leverage the structure and properties of graphs?;,citation_author=Benjamin Sanchez-Lengeling;,citation_author=Emily Reif;,citation_author=Adam Pearce;,citation_author=Alexander B. Wiltschko;,citation_publication_date=2021-09-02;,citation_cover_date=2021-09-02;,citation_year=2021;,citation_fulltext_html_url=https://distill.pub/2021/gnn-intro;,citation_issue=9;,citation_doi=10.23915/distill.00033;,citation_issn=2476-0757;,citation_volume=6;,citation_journal_title=Distill;,citation_journal_abbrev=Distill;">
<meta name="citation_reference" content="citation_title=Boomtown, flood town;,citation_abstract=Climate change will bring more frequent and fierce rainstorms to cities like Houston. But unchecked development remains a priority in the famously un-zoned city, creating short-term economic gains for some while increasing flood risks for everyone.;,citation_author=Nina Satija;,citation_author=Kiah Collier;,citation_author=Al Shaw;,citation_publication_date=2016-12-07;,citation_cover_date=2016-12-07;,citation_year=2016;,citation_fulltext_html_url=https://projects.propublica.org/houston-cypress/;,citation_journal_title=ProPublica: Hell and High Water;">
<meta name="citation_reference" content="citation_title=Foundations of statistics.;,citation_author=L. J. Savage;,citation_publication_date=1954;,citation_cover_date=1954;,citation_year=1954;">
<meta name="citation_reference" content="citation_title=Incorporating non-stationarity from climate change into rainfall frequency and intensity-duration-frequency (IDF) curves;,citation_abstract=Intensity-duration-frequency (IDF) curves – sometimes also called precipitation frequency estimates – are used to design urban drainage systems for stormwater control and, when combined with hydrologic modeling, to design flood control infrastructure and other structures. However, common approaches to estimating IDF curves need to be revised to account for non-stationarity from global climate change. This review synthesizes the current knowledge and on-going research regarding IDF curves under non-stationarity. In particular, this review briefly summarizes known and projected changes in extreme precipitation at the global scale, describes approaches for IDF curve estimation under non-stationarity (focusing on the covariate-based approach and including a brief overview of the specific challenges facing snow-dominated regions), addresses the topic of regionalization under non-stationarity (which has been overlooked in previous reviews), explores the challenges of design values and uncertainty in the context of non-stationarity, provides details on these topics in the context of the United States, and finishes by enumerating needed avenues of future research.;,citation_author=Katherine E. Schlef;,citation_author=Kenneth E. Kunkel;,citation_author=Casey Brown;,citation_author=Yonas Demissie;,citation_author=Dennis P. Lettenmaier;,citation_author=Anna Wagner;,citation_author=Mark S. Wigmosta;,citation_author=Thomas R. Karl;,citation_author=David R. Easterling;,citation_author=Kimberly J. Wang;,citation_author=Baptiste François;,citation_author=Eugene Yan;,citation_publication_date=2023-01-01;,citation_cover_date=2023-01-01;,citation_year=2023;,citation_fulltext_html_url=https://www.sciencedirect.com/science/article/pii/S0022169422013270;,citation_doi=10.1016/j.jhydrol.2022.128757;,citation_issn=0022-1694;,citation_volume=616;,citation_journal_title=Journal of Hydrology;,citation_journal_abbrev=Journal of Hydrology;">
<meta name="citation_reference" content="citation_title=Can we estimate the likelihood of climatic changes at 2100?;,citation_abstract=Schneider points out some of the consequences of the deliberate choice not to discuss the possibilities of CO2 emissions and how that can–and has–led to confusion on the part of the media and policy makers over the likelihood of potential dangerous anthropogenic climate change in the next century.;,citation_author=Stephen H. Schneider;,citation_publication_date=2002-03;,citation_cover_date=2002-03;,citation_year=2002;,citation_fulltext_html_url=https://www.proquest.com/docview/198493931/abstract/7F4C45C74324C3EPQ/1;,citation_issue=4;,citation_doi=http://dx.doi.org/10.1023/A:1014276210717;,citation_issn=01650009;,citation_volume=52;,citation_journal_title=Climatic Change;,citation_publisher=Springer Nature B.V.;">
<meta name="citation_reference" content="citation_title=Optimizing Scale for Decentralized Wastewater Treatment: A Tool to Address Failing Wastewater Infrastructure in the United States;,citation_abstract=Wastewater systems (sewered or on-site septic tanks) are failing across the U.S. Economically disadvantaged populations, communities of color, tribal lands, and rural/peri-urban areas are especially vulnerable. Efficient deployment of public and private capital to assure appropriate service levels and affordability is a critical need. We present a modeling framework to identify economically optimal wastewater network layout and component sizes. Six system configurations are considered, which include gravity versus pressurized collection system flow and treatment of septic effluent versus raw wastewater. A case study for Uniontown, Alabama─a community that has been in national news for the failure of their wastewater infrastructure─is presented. We find that a decentralized network that separates and stores solids on-site, prior to conveyance to a cluster-scale treatment site, has a fraction of the capital and operating cost of a centralized system that is currently proposed. Broader implications are discussed.;,citation_author=Sara E. Schwetschenau;,citation_author=Yunus Kovankaya;,citation_author=Mark A. Elliott;,citation_author=Maura Allaire;,citation_author=Kevin D. White;,citation_author=Upmanu Lall;,citation_publication_date=2023-01-13;,citation_cover_date=2023-01-13;,citation_year=2023;,citation_fulltext_html_url=https://doi.org/10.1021/acsestengg.2c00188;,citation_issue=1;,citation_doi=10.1021/acsestengg.2c00188;,citation_volume=3;,citation_journal_title=ACS ES&amp;amp;amp;T Engineering;,citation_journal_abbrev=ACS EST Eng.;,citation_publisher=American Chemical Society;">
<meta name="citation_reference" content="citation_title=Operationalizing equity for integrated water resources management;,citation_abstract=Advancing social equity has been implicitly and explicitly central to water resources policy for decades. Yet, equity remains largely outside of standard water resources planning and management practices. Inclusion of equity within water resources infrastructure is inhibited by barriers including an incomplete conceptual understanding of equity, a perceived lack of quantitative and qualitative equity metrics, unclear connections between equity and standard project planning frameworks, and the absence of concrete examples. To facilitate greater practical inclusion of social equity in water resources practices, we describe equity relative to dimensions of distribution, procedure, and recognition and identify metrics associated with each. We then map these dimensions of equity onto different stages of a water resources project life cycle. We discuss how inequities are often perpetuated by current approaches and highlight case studies that promote one or more of the facets of equity. Rather than providing a prescriptive solution to “achieve” equity within water resources practices, we emphasize the need for contextualized approaches that include pragmatic steps toward more equitable practices and outcomes.;,citation_author=Cydney K. Seigerman;,citation_author=S. Kyle McKay;,citation_author=Raul Basilio;,citation_author=Shelly A. Biesel;,citation_author=Jon Hallemeier;,citation_author=Andressa V. Mansur;,citation_author=Candice Piercy;,citation_author=Sebastian Rowan;,citation_author=Bruno Ubiali;,citation_author=Elissa Yeates;,citation_author=Donald R. Nelson;,citation_publication_date=2023;,citation_cover_date=2023;,citation_year=2023;,citation_fulltext_html_url=https://onlinelibrary.wiley.com/doi/abs/10.1111/1752-1688.13086;,citation_issue=2;,citation_doi=10.1111/1752-1688.13086;,citation_issn=1752-1688;,citation_volume=59;,citation_journal_title=JAWRA Journal of the American Water Resources Association;">
<meta name="citation_reference" content="citation_title=Weather and climate extreme events in a changing climate;,citation_author=S. I. Seneviratne;,citation_author=X. Zhang;,citation_author=M. Adnan;,citation_author=W. Badi;,citation_author=C. Dereczynski;,citation_author=A. Di Luca;,citation_author=S. Ghosh;,citation_author=I. Iskandar;,citation_author=J. Kossin;,citation_author=S. Lewis;,citation_author=F. Otto;,citation_author=I. Pinto;,citation_author=M. Satoh;,citation_author=S. M. Vicente-Serrano;,citation_author=M. Wehner;,citation_author=B. Zhou;,citation_editor=V. Masson-Delmotte;,citation_editor=P. Zhai;,citation_editor=A. Pirani;,citation_editor=S. L. Connors;,citation_editor=C. Péan;,citation_editor=S. Berger;,citation_editor=N. Caud;,citation_editor=Y. Chen;,citation_editor=L. Goldfarb;,citation_editor=M. I. Gomis;,citation_editor=M. Huang;,citation_editor=K. Leitzell;,citation_editor=E. Lonnoy;,citation_editor=J. B. R. Matthews;,citation_editor=T. K. Maycock;,citation_editor=T. Waterfield;,citation_editor=O. Yelekçi;,citation_editor=R. Yu;,citation_editor=B. Zhou;,citation_publication_date=2021;,citation_cover_date=2021;,citation_year=2021;,citation_fulltext_html_url=https://www.ipcc.ch/report/ar6/wg1/downloads/report/IPCC_AR6_WGI_Chapter11.pdf;,citation_doi=10.1017/9781009157896.013;,citation_inbook_title=Climate change 2021: The physical science basis. Contribution of working group I to the sixth assessment report of the intergovernmental panel on climate change;">
<meta name="citation_reference" content="citation_title=Multiobjective optimization and Pareto front visualization techniques applied to normal conducting rf accelerating structures;,citation_abstract=There has been a renewed interest in applying multiobjective (MO) optimization methods to a number of problems in the physical sciences, including to rf structure design. The results of these optimizations generate large datasets, which makes visualizing the data and selecting individual solutions difficult. Using the generated results, Pareto fronts can be found giving the trade-off between different objectives, allowing one to utilize this key information in design decisions. Although various visualization techniques exist, it can be difficult to know which technique is appropriate and how to apply them successfully to the problem at hand. First, we present the setup and execution of MO optimizations of one standing wave and one traveling wave accelerating cavity, including constraint handling and an algorithm comparison. In order to understand the generated Pareto frontiers, we discuss several visualization techniques, applying them to the problem, and give the benefits and drawbacks of each. We found that the best techniques involve clustering the resulting data first to narrow down the possible choices and then using multidimensional visualization methods such as parallel coordinate plots and decision maps to view the clustered results and select individual solutions. Finally, we give some examples of the application of these methods and the cavities selected based on arbitrary design requirements.;,citation_author=S. Smith;,citation_author=M. Southerby;,citation_author=S. Setiniyaz;,citation_author=R. Apsimon;,citation_author=G. Burt;,citation_publication_date=2022-06-14;,citation_cover_date=2022-06-14;,citation_year=2022;,citation_fulltext_html_url=https://link.aps.org/doi/10.1103/PhysRevAccelBeams.25.062002;,citation_issue=6;,citation_doi=10.1103/PhysRevAccelBeams.25.062002;,citation_volume=25;,citation_journal_title=Physical Review Accelerators and Beams;,citation_journal_abbrev=Phys. Rev. Accel. Beams;,citation_publisher=American Physical Society;">
<meta name="citation_reference" content="citation_title=Near-term tropical cyclone risk and coupled Earth system model biases;,citation_abstract=Most current climate models predict that the equatorial Pacific will evolve under greenhouse gas–induced warming to a more El Niño-like state over the next several decades, with a reduced zonal sea surface temperature gradient and weakened atmospheric Walker circulation. Yet, observations over the last 50 y show the opposite trend, toward a more La Niña-like state. Recent research provides evidence that the discrepancy cannot be dismissed as due to internal variability but rather that the models are incorrectly simulating the equatorial Pacific response to greenhouse gas warming. This implies that projections of regional tropical cyclone activity may be incorrect as well, perhaps even in the direction of change, in ways that can be understood by analogy to historical El Niño and La Niña events: North Pacific tropical cyclone projections will be too active, North Atlantic ones not active enough, for example. Other perils, including severe convective storms and droughts, will also be projected erroneously. While it can be argued that these errors are transient, such that the models’ responses to greenhouse gases may be correct in equilibrium, the transient response is relevant for climate adaptation in the next several decades. Given the urgency of understanding regional patterns of climate risk in the near term, it would be desirable to develop projections that represent a broader range of possible future tropical Pacific warming scenarios—including some in which recent historical trends continue—even if such projections cannot currently be produced using existing coupled earth system models.;,citation_author=Adam H. Sobel;,citation_author=Chia-Ying Lee;,citation_author=Steven G. Bowen;,citation_author=Suzana J. Camargo;,citation_author=Mark A. Cane;,citation_author=Amy Clement;,citation_author=Boniface Fosu;,citation_author=Megan Hart;,citation_author=Kevin A. Reed;,citation_author=Richard Seager;,citation_author=Michael K. Tippett;,citation_publication_date=2023-08-15;,citation_cover_date=2023-08-15;,citation_year=2023;,citation_fulltext_html_url=https://www.pnas.org/doi/10.1073/pnas.2209631120;,citation_issue=33;,citation_doi=10.1073/pnas.2209631120;,citation_volume=120;,citation_journal_title=Proceedings of the National Academy of Sciences;,citation_publisher=Proceedings of the National Academy of Sciences;">
<meta name="citation_reference" content="citation_title=An unexpected item is blocking cities’ climate change prep: Obsolete rainfall records;,citation_abstract=Cities are experiencing heavier storms and flooding as the climate gets hotter. But due to outdated rainfall records, many are still building infrastructure for the climate of the past.;,citation_author=Lauren Sommer;,citation_publication_date=2022-02-09;,citation_cover_date=2022-02-09;,citation_year=2022;,citation_fulltext_html_url=https://www.npr.org/2022/02/09/1078261183/an-unexpected-item-is-blocking-cities-climate-change-prep-obsolete-rainfall-reco;,citation_journal_title=NPR: Climate;">
<meta name="citation_reference" content="citation_title=Expanded decision-scaling framework to select robust long-term water-system plans under hydroclimatic uncertainties;,citation_author=Scott Steinschneider;,citation_author=Rachel McCrary;,citation_author=Sungwook Wi;,citation_author=Kevin Mulligan;,citation_author=Linda O Mearns;,citation_author=Casey M Brown;,citation_publication_date=2015-11;,citation_cover_date=2015-11;,citation_year=2015;,citation_fulltext_html_url=http://ascelibrary.org/doi/10.1061/%28ASCE%29WR.1943-5452.0000536;,citation_issue=11;,citation_doi=10.1061/(asce)wr.1943-5452.0000536;,citation_volume=141;,citation_journal_title=Journal of Water Resources Planning and Management;">
<meta name="citation_reference" content="citation_title=Reinforcement learning: An Introduction;,citation_author=Richard S Sutton;,citation_author=Andrew G Barto;,citation_publication_date=2018;,citation_cover_date=2018;,citation_year=2018;,citation_isbn=0-262-03924-9;">
<meta name="citation_reference" content="citation_title=Global and regional sea level rise scenarios for the United States;,citation_author=W. V. Sweet;,citation_author=B. D. Hamlington;,citation_author=R. E. Kopp;,citation_author=C. P. Weaver;,citation_author=P. L. Barnard;,citation_author=D. Bekaert;,citation_author=W. Brooks;,citation_author=M. Craghan;,citation_author=G. Dusek;,citation_publication_date=2022;,citation_cover_date=2022;,citation_year=2022;,citation_fulltext_html_url=https://oceanservice.noaa.gov/hazards/sealevelrise/sealevelrise-tech-report-sections.html;,citation_technical_report_institution=National Oceanic and Atmospheric Administration, National Ocean Service;,citation_technical_report_number=NOS 01;">
<meta name="citation_reference" content="citation_title=Exposure of real estate properties to the 2018 Hurricane Florence flooding;,citation_abstract=Quantifying the potential exposure of property to damages associated with storm surges, extreme weather and hurricanes is fundamental to developing frameworks that can be used to conceive and implement mitigation plans as well as support urban development that accounts for such events. In this study, we aim at quantifying the total value and area of properties exposed to the flooding associated with Hurricane Florence that occurred in September 2018. To this aim, we implement an approach for the identification of affected areas by generating a map of the maximum flood extent obtained from a combination of the flood extent produced by the Federal Emergency Management Agency’s (FEMA’s) water marks with those obtained from spaceborne radar remote-sensing data. The use of radar in the creation of the flood extent allows for those properties commonly missed by FEMA’s interpolation methods, especially from pluvial or non-fluvial sources, and can be used in more accurately estimating the exposure and market value of properties to event-specific flooding. Lastly, we study and quantify how the urban development over the past decades in the regions flooded by Hurricane Florence might have impacted the exposure of properties to present-day storms and floods. This approach is conceptually similar to what experts are addressing as the “expanding bull’s eye effect”, in which “targets” of geophysical hazards, such as people and their built environments, enlarge as populations grow and spread. Our results indicate that the total value of property exposed to flooding during Hurricane Florence was USD&amp;amp;amp;thinsp;52 billion (in 2018 USD), with this value increasing from USD&amp;thinsp;∼10 billion at the beginning of the past century to the final amount based on the expansion of the number of properties exposed. We also found that, despite the decrease in the number of properties built during the decade before Florence, much of the new construction was in proximity to permanent water bodies, hence increasing exposure to flooding. Ultimately, the results of this paper provide a new tool for shedding light on the relationships between urban development in coastal areas and the flooding of those areas, which is estimated to increase in view of projected increasing sea level rise, storm surges and the strength of storms.;,citation_author=Marco Tedesco;,citation_author=Steven McAlpine;,citation_author=Jeremy R. Porter;,citation_publication_date=2020-04-01;,citation_cover_date=2020-04-01;,citation_year=2020;,citation_fulltext_html_url=https://www.nat-hazards-earth-syst-sci.net/20/907/2020/;,citation_issue=3;,citation_doi=10.5194/nhess-20-907-2020;,citation_issn=1561-8633;,citation_volume=20;,citation_journal_title=Natural Hazards and Earth System Sciences;,citation_publisher=Copernicus GmbH;">
<meta name="citation_reference" content="citation_title=Systemic Financial Risk Arising From Residential Flood Losses;,citation_abstract=Direct damage from flooding at residential properties has typically been categorized as insured, with liabilities accruing to insurers, or uninsured, with costs accruing to property owners. However, residential flooding can also expose lenders and local governments to financial risk, though the distribution of this risk is not well understood. Flood losses are not limited to direct damages, but also include indirect effects such as decreases in property values, which can be substantial, though are rarely well quantified. The combination of direct damage and property value decrease influences rates of mortgage default and property abandonment in the wake of a flood, creating financial risk. In this research, property-level data on sales, mortgages, and insurance claims are used in combination with machine learning techniques and geostatistical methods to provide estimates of flood losses that are then utilized to evaluate the risk of default and abandonment in eastern North Carolina following Hurricane Florence (2018). Within the study area, Hurricane Florence generated \366M in observed insured damages and an estimated \1.77B in combined uninsured damages and property value decreases. Property owners, lenders, and local governments were exposed to an additional $562M in potential losses due to increased rates of default and abandonment. Areas with lower pre-flood property values were exposed to greater risk than areas with higher valued properties. Results suggest more highly resolved estimates of a flooding event’s systemic financial risk may be useful in developing improved flood resilience strategies.;,citation_author=Hope Thomson;,citation_author=Harrison B. Zeff;,citation_author=Rachel Kleiman;,citation_author=Antonia Sebastian;,citation_author=Gregory W. Characklis;,citation_publication_date=2023;,citation_cover_date=2023;,citation_year=2023;,citation_fulltext_html_url=https://onlinelibrary.wiley.com/doi/abs/10.1029/2022EF003206;,citation_issue=4;,citation_doi=10.1029/2022EF003206;,citation_issn=2328-4277;,citation_volume=11;,citation_journal_title=Earth’s Future;">
<meta name="citation_reference" content="citation_title=Physics-based deep learning;,citation_author=N. Thuerey;,citation_author=B. Holzschuh;,citation_author=P. Holl;,citation_author=G. Kohl;,citation_author=M. Lino;,citation_author=Q. Liu;,citation_author=P. Schnell;,citation_author=F. Trost;,citation_publication_date=2024;,citation_cover_date=2024;,citation_year=2024;,citation_fulltext_html_url=https://physicsbaseddeeplearning.org;">
<meta name="citation_reference" content="citation_title=Impacts of future weather and climate extremes on United States infrastructure: Assessing and prioritizing adaptation actions;,citation_author=Mari R. Tye;,citation_author=Jason P. Giovannettone;,citation_publication_date=2021-10-13;,citation_cover_date=2021-10-13;,citation_year=2021;,citation_fulltext_html_url=https://ascelibrary.org/doi/book/10.1061/9780784415863;,citation_doi=10.1061/9780784415863;,citation_isbn=978-0-7844-1586-3 978-0-7844-8372-5;">
<meta name="citation_reference" content="citation_title=Evaluation of flood risk reduction strategies through combinations of interventions;,citation_abstract=Large, complex coastal regions often require a combination of interventions to lower the risk of flooding to an acceptable level. In practice, a limited number of strategies are considered and interdependencies between interventions are often simplified. This paper presents the Multiple Lines of Defence Optimization System (MODOS)-model. This quick, probabilistic model simulates and evaluates the impact of many flood risk reduction strategies while accounting for interdependencies amongst measures. The simulation includes hydraulic calculations, damage calculations, and the effects of measures for various return periods. The application and potential of this model is shown with a conceptual and simplified case study, based on the Houston-Galveston Bay area. The analyses demonstrate how the MODOS-model identifies trade-offs within the system and shows how flood risk, cost, and impact respond to flood management decisions. This improved understanding of the impact of design and planning choices can benefit the discussions in finding the optimal flood risk reduction strategy for coastal regions.;,citation_author=Erik C. Berchum;,citation_author=William Mobley;,citation_author=Sebastiaan N. Jonkman;,citation_author=Jos S. Timmermans;,citation_author=Jan H. Kwakkel;,citation_author=Samuel D. Brody;,citation_publication_date=2019;,citation_cover_date=2019;,citation_year=2019;,citation_fulltext_html_url=https://onlinelibrary.wiley.com/doi/abs/10.1111/jfr3.12506;,citation_issue=S2;,citation_doi=10.1111/jfr3.12506;,citation_issn=1753-318X;,citation_volume=12;,citation_journal_title=Journal of Flood Risk Management;">
<meta name="citation_reference" content="citation_title=Economic Decision Problems for Flood Prevention;,citation_author=D. Dantzig;,citation_publication_date=1956;,citation_cover_date=1956;,citation_year=1956;,citation_fulltext_html_url=https://doi.org/10.2307/1911632;,citation_issue=3;,citation_doi=10.2307/1911632;,citation_issn=0012-9682;,citation_volume=24;,citation_journal_title=Econometrica;">
<meta name="citation_reference" content="citation_title=Deep Uncertainty;,citation_author=Warren E. Walker;,citation_author=Robert J. Lempert;,citation_author=Jan H. Kwakkel;,citation_editor=Saul I. Gass;,citation_editor=Michael C. Fu;,citation_publication_date=2013;,citation_cover_date=2013;,citation_year=2013;,citation_fulltext_html_url=https://doi.org/10.1007/978-1-4419-1153-7_1140;,citation_doi=10.1007/978-1-4419-1153-7_1140;,citation_isbn=978-1-4419-1153-7;,citation_inbook_title=Encyclopedia of Operations Research and Management Science;">
<meta name="citation_reference" content="citation_title=A review of the Stern review on the economics of climate change;,citation_author=Martin L. Weitzman;,citation_publication_date=2007-09;,citation_cover_date=2007-09;,citation_year=2007;,citation_fulltext_html_url=https://www.aeaweb.org/articles?id=10.1257/jel.45.3.703;,citation_issue=3;,citation_doi=10.1257/jel.45.3.703;,citation_issn=0022-0515;,citation_volume=45;,citation_journal_title=Journal of Economic Literature;">
<meta name="citation_reference" content="citation_title=New insights into US flood vulnerability revealed from flood insurance big data;,citation_abstract=Improvements in modelling power and input data have vastly improved the precision of physical flood models, but translation into economic outputs requires depth–damage functions that are inadequately verified. In particular, flood damage is widely assumed to increase monotonically with water depth. Here, we assess flood vulnerability in the US using $&amp;amp;amp;gt;$2 million claims from the National Flood Insurance Program (NFIP). NFIP claims data are messy, but the size of the dataset provides powerful empirical tests of damage patterns and modelling approaches. We show that current depth–damage functions consist of disparate relationships that match poorly with observations. Observed flood losses are not monotonic functions of depth, but instead better follow a beta function, with bimodal distributions for different water depths. Uncertainty in flood losses has been called the main bottleneck in flood risk studies, an obstacle that may be remedied using large-scale empirical flood damage data.;,citation_author=Oliver E. J. Wing;,citation_author=Nicholas Pinter;,citation_author=Paul D. Bates;,citation_author=Carolyn Kousky;,citation_publication_date=2020-03-19;,citation_cover_date=2020-03-19;,citation_year=2020;,citation_fulltext_html_url=https://www.nature.com/articles/s41467-020-15264-2;,citation_issue=1, 1;,citation_doi=10.1038/s41467-020-15264-2;,citation_issn=2041-1723;,citation_volume=11;,citation_journal_title=Nature Communications;,citation_publisher=Nature Publishing Group;">
<meta name="citation_reference" content="citation_title=MAD water: Integrating modular, adaptive, and decentralized approaches for water security in the climate change era;,citation_abstract=Centralized water infrastructure has, over the last century, brought safe and reliable drinking water to much of the world. But climate change, combined with aging and underfunded infrastructure, is increasingly testing the limits of—and reversing gains made by—this approach. To address these growing strains and gaps, we must assess and advance alternatives to centralized water provision and sanitation. The water literature is rife with examples of systems that are neither centralized nor networked, yet meet water needs of local communities in important ways, including: informal and hybrid water systems, decentralized water provision, community-based water management, small drinking water systems, point-of-use treatment, small-scale water vendors, and packaged water. Our work builds on these literatures by proposing a convergence approach that can integrate and explore the benefits and challenges of modular, adaptive, and decentralized (“MAD”) water provision and sanitation, often foregrounding important advances in engineering technology. We further provide frameworks to evaluate justice, economic feasibility, governance, human health, and environmental sustainability as key parameters of MAD water system performance.;,citation_author=Amber Wutich;,citation_author=Patrick Thomson;,citation_author=Wendy Jepson;,citation_author=Justin Stoler;,citation_author=Alicia D. Cooperman;,citation_author=James Doss-Gollin;,citation_author=Anish Jantrania;,citation_author=Alex Mayer;,citation_author=Jami Nelson-Nuñez;,citation_author=W. Shane Walker;,citation_author=Paul Westerhoff;,citation_publication_date=2023-07-12;,citation_cover_date=2023-07-12;,citation_year=2023;,citation_fulltext_html_url=https://onlinelibrary.wiley.com/doi/abs/10.1002/wat2.1680;,citation_issue=n/a;,citation_doi=10.1002/wat2.1680;,citation_issn=2049-1948;,citation_volume=n/a;,citation_journal_title=WIREs Water;">
<meta name="citation_reference" content="citation_title=Neglecting uncertainties biases house-elevation decisions to manage riverine flood risks;,citation_abstract=Homeowners around the world elevate houses to manage flood risks. Deciding how high to elevate a house poses a nontrivial decision problem. The U.S. Federal Emergency Management Agency (FEMA) recommends elevating existing houses to the Base Flood Elevation (the elevation of the 100-year flood) plus a freeboard. This recommendation neglects many uncertainties. Here we analyze a case-study of riverine flood risk management using a multi-objective robust decision-making framework in the face of deep uncertainties. While the quantitative results are location-specific, the approach and overall insights are generalizable. We find strong interactions between the economic, engineering, and Earth science uncertainties, illustrating the need for expanding on previous integrated analyses to further understand the nature and strength of these connections. Considering deep uncertainties surrounding flood hazards, the discount rate, the house lifetime, and the fragility can increase the economically optimal house elevation to values well above FEMA’s recommendation.;,citation_author=Mahkameh Zarekarizi;,citation_author=Vivek Srikrishnan;,citation_author=Klaus Keller;,citation_publication_date=2020-10-26;,citation_cover_date=2020-10-26;,citation_year=2020;,citation_fulltext_html_url=http://www.nature.com/articles/s41467-020-19188-9;,citation_issue=1, 1;,citation_doi=10.1038/s41467-020-19188-9;,citation_issn=2041-1723;,citation_volume=11;,citation_journal_title=Nature Communications;,citation_publisher=Nature Publishing Group;">
</head>

<body class="nav-sidebar floating quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../notebooks/index.html"><strong>Computational Case Studies</strong></a></li><li class="breadcrumb-item"><a href="../notebooks/probability-stats-examples.html">Probability, inference, and computation: examples ✏️</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../">Climate Risk Assessment and Management</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome 🎯</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text"><strong>About this book</strong></span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/about/license.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">License</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/about/contributing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Contributing</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/about/citing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Citing</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/about/resources.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Further Reading</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/preface.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text"><strong>I: Foundations</strong></span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/fundamentals/climate-science.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Climate science 🚧</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/fundamentals/probability-stats.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Probability and inference ✏️</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/fundamentals/ml-nonparametric.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Machine learning and nonparametric methods 🚧</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/fundamentals/correlation-dimensionality.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Correlation and dimensionality 🚧</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/fundamentals/model-comparison.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Model validation and comparison 🚧</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true">
 <span class="menu-text"><strong>II: Hazard Assessment</strong></span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/hazard/extremes.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Extreme value theory ✏️</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/hazard/downscaling-bias-correction.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Downscaling and Bias Correction 🚧</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/hazard/generators.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Stochastic weather generators 🚧</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/hazard/physics-models.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Physics-based models and calibration 🚧</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/hazard/sampling.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Optimal sampling methods 🚧</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/hazard/sensitivity-analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Global sensitivity analysis 🚧</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true">
 <span class="menu-text"><strong>III: Risk Management</strong></span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/risk/exposure-vulnerability.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Exposure and Vulnerability 🚧</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/risk/expectations-cost-benefit.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Cost-Benefit Analysis and Net Present Value 🚧</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/risk/policy-search.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Policy Search &amp; Optimization 🚧</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/risk/risk-transfer.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Risk Transfer 🚧</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/risk/deep-uncertainty.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Deep Uncertainty and Model Structure 🚧</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/risk/robustness.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">Robustness 🚧</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/risk/adaptive.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Adaptive Planning and Flexibility 🚧</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/risk/social-science.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Working with People: Values, Participation, and Communication 🚧</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true">
 <span class="menu-text"><strong>Computational Case Studies</strong></span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../notebooks/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Overview</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../notebooks/probability-stats-examples.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Probability, inference, and computation: examples ✏️</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../notebooks/extremes-examples.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Extreme Value Theory Examples ✏️</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../notebooks/mosquitos.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Statistics Without the Agonizing Details ✏️</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><strong>References</strong></span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="true">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/appendices/software.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">Software Setup ✏️</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/appendices/julia.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">B</span>&nbsp; <span class="chapter-title">Julia Learning Resources ✏️</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/appendices/llm.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">C</span>&nbsp; <span class="chapter-title">Large Language Models (“AI”) ✏️</span></span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
<div class="quarto-sidebar-footer"><div class="sidebar-footer-item">
<p>This project is <strong>under construction.</strong> Each page is either: 🚧 Planning, ✏️ Draft, or complete.</p>
</div></div></nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#distribution-examples" id="toc-distribution-examples" class="nav-link active" data-scroll-target="#distribution-examples">Distribution examples</a></li>
  <li><a href="#joint-marginal-and-conditional-distributions" id="toc-joint-marginal-and-conditional-distributions" class="nav-link" data-scroll-target="#joint-marginal-and-conditional-distributions">Joint, marginal, and conditional distributions</a></li>
  <li><a href="#coin-flip-examples" id="toc-coin-flip-examples" class="nav-link" data-scroll-target="#coin-flip-examples">Coin flip examples</a>
  <ul class="collapse">
  <li><a href="#likelihood-visualization" id="toc-likelihood-visualization" class="nav-link" data-scroll-target="#likelihood-visualization">Likelihood visualization</a></li>
  <li><a href="#maximum-likelihood-estimation" id="toc-maximum-likelihood-estimation" class="nav-link" data-scroll-target="#maximum-likelihood-estimation">Maximum likelihood estimation</a></li>
  <li><a href="#bayesian-analysis" id="toc-bayesian-analysis" class="nav-link" data-scroll-target="#bayesian-analysis">Bayesian analysis</a></li>
  <li><a href="#beta-binomial-updating-in-action" id="toc-beta-binomial-updating-in-action" class="nav-link" data-scroll-target="#beta-binomial-updating-in-action">Beta-Binomial updating in action</a></li>
  </ul></li>
  <li><a href="#linear-regression-example" id="toc-linear-regression-example" class="nav-link" data-scroll-target="#linear-regression-example">Linear regression example</a>
  <ul class="collapse">
  <li><a href="#generating-synthetic-data" id="toc-generating-synthetic-data" class="nav-link" data-scroll-target="#generating-synthetic-data">Generating synthetic data</a></li>
  <li><a href="#curve-fitting-by-minimizing-prediction-error" id="toc-curve-fitting-by-minimizing-prediction-error" class="nav-link" data-scroll-target="#curve-fitting-by-minimizing-prediction-error">Curve fitting by minimizing prediction error</a></li>
  <li><a href="#maximum-likelihood-estimation-1" id="toc-maximum-likelihood-estimation-1" class="nav-link" data-scroll-target="#maximum-likelihood-estimation-1">Maximum likelihood estimation</a></li>
  <li><a href="#bayesian-inference" id="toc-bayesian-inference" class="nav-link" data-scroll-target="#bayesian-inference">Bayesian inference</a></li>
  <li><a href="#mle-with-turing" id="toc-mle-with-turing" class="nav-link" data-scroll-target="#mle-with-turing">MLE with Turing</a></li>
  <li><a href="#comparison-of-approaches" id="toc-comparison-of-approaches" class="nav-link" data-scroll-target="#comparison-of-approaches">Comparison of approaches</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../notebooks/index.html"><strong>Computational Case Studies</strong></a></li><li class="breadcrumb-item"><a href="../notebooks/probability-stats-examples.html">Probability, inference, and computation: examples ✏️</a></li></ol></nav>
<div class="quarto-title">
<h1 class="title">Probability, inference, and computation: examples ✏️</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>This notebook provides computational examples to accompany the <a href="../chapters/fundamentals/probability-stats.html">Probability and inference</a> chapter. See also: <a href="../notebooks/mosquitos.html">Moquito bites and beer consumption</a></p>
<div id="setup" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="annotated-cell-1"><pre class="sourceCode numberSource julia code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode julia"><button class="code-annotation-anchor" data-target-cell="annotated-cell-1" data-target-annotation="1">1</button><span id="annotated-cell-1-1" class="code-annotation-target"><a href="#annotated-cell-1-1"></a><span class="im">using</span> <span class="bu">CairoMakie</span></span>
<span id="annotated-cell-1-2"><a href="#annotated-cell-1-2"></a><span class="im">using</span> <span class="bu">Distributions</span></span>
<span id="annotated-cell-1-3"><a href="#annotated-cell-1-3"></a><span class="im">using</span> <span class="bu">LaTeXStrings</span></span>
<span id="annotated-cell-1-4"><a href="#annotated-cell-1-4"></a><span class="im">using</span> <span class="bu">Turing</span></span>
<span id="annotated-cell-1-5"><a href="#annotated-cell-1-5"></a><span class="im">using</span> <span class="bu">Printf</span></span>
<span id="annotated-cell-1-6"><a href="#annotated-cell-1-6"></a><span class="im">using</span> <span class="bu">Random</span></span>
<span id="annotated-cell-1-7"><a href="#annotated-cell-1-7"></a><span class="im">using</span> <span class="bu">StatsBase</span></span>
<span id="annotated-cell-1-8"><a href="#annotated-cell-1-8"></a><span class="im">using</span> <span class="bu">HypothesisTests</span></span>
<span id="annotated-cell-1-9"><a href="#annotated-cell-1-9"></a><span class="im">using</span> <span class="bu">Optim</span></span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-annotation">
<dl class="code-annotation-container-hidden code-annotation-container-grid">
<dt data-target-cell="annotated-cell-1" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-1" data-code-lines="1" data-code-annotation="1">We use <code>using</code> to import necessary packages</span>
</dd>
</dl>
</div>
</div>
<section id="distribution-examples" class="level2">
<h2 class="anchored" data-anchor-id="distribution-examples">Distribution examples</h2>
<p>Understanding probability distributions requires grasping three fundamental functions: the probability density function (PDF) or probability mass function (PMF), the cumulative distribution function (CDF), and the quantile function. These examples demonstrate how these functions work together using the normal distribution (continuous) and Poisson distribution (discrete).</p>
<p>The visualizations show both forward operations (finding probabilities from values using the CDF) and inverse operations (finding values from probabilities using quantiles). This computational perspective reinforces the theoretical relationships discussed in the main chapter.</p>
<p>We start by defining some helper functions.</p>
<div id="distribution-helpers" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="annotated-cell-2"><pre class="sourceCode numberSource julia code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode julia"><button class="code-annotation-anchor" data-target-cell="annotated-cell-2" data-target-annotation="1">1</button><span id="annotated-cell-2-1" class="code-annotation-target"><a href="#annotated-cell-2-1"></a><span class="kw">function</span> <span class="fu">add_pdf_area!</span>(ax, dist, a, b; color <span class="op">=</span> (<span class="op">:</span>orange, <span class="fl">0.4</span>), label <span class="op">=</span> <span class="cn">nothing</span>)</span>
<span id="annotated-cell-2-2"><a href="#annotated-cell-2-2"></a>    x_fill <span class="op">=</span> a<span class="op">:</span><span class="fl">0.01</span><span class="op">:</span>b</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-2" data-target-annotation="2">2</button><span id="annotated-cell-2-3" class="code-annotation-target"><a href="#annotated-cell-2-3"></a>    pdf_fill <span class="op">=</span> <span class="fu">pdf</span>.(dist, x_fill)</span>
<span id="annotated-cell-2-4"><a href="#annotated-cell-2-4"></a>    <span class="fu">band!</span>(ax, x_fill, <span class="fu">zeros</span>(<span class="fu">length</span>(x_fill)), pdf_fill, color <span class="op">=</span> color, label <span class="op">=</span> label)</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-2" data-target-annotation="3">3</button><span id="annotated-cell-2-5" class="code-annotation-target"><a href="#annotated-cell-2-5"></a>    prob <span class="op">=</span> <span class="fu">cdf</span>(dist, b) <span class="op">-</span> <span class="fu">cdf</span>(dist, a)</span>
<span id="annotated-cell-2-6"><a href="#annotated-cell-2-6"></a>    <span class="cf">return</span> prob</span>
<span id="annotated-cell-2-7"><a href="#annotated-cell-2-7"></a><span class="kw">end</span></span>
<span id="annotated-cell-2-8"><a href="#annotated-cell-2-8"></a></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-2" data-target-annotation="4">4</button><span id="annotated-cell-2-9" class="code-annotation-target"><a href="#annotated-cell-2-9"></a><span class="kw">function</span> <span class="fu">add_forward_cdf!</span>(ax, dist, x_point; color <span class="op">=</span> <span class="op">:</span>red, x_min <span class="op">=</span> <span class="op">-</span><span class="fl">4</span>)</span>
<span id="annotated-cell-2-10"><a href="#annotated-cell-2-10"></a>    y_point <span class="op">=</span> <span class="fu">cdf</span>(dist, x_point)</span>
<span id="annotated-cell-2-11"><a href="#annotated-cell-2-11"></a>    <span class="fu">scatter!</span>(ax, [x_point], [y_point], color <span class="op">=</span> color, markersize <span class="op">=</span> <span class="fl">8</span>)</span>
<span id="annotated-cell-2-12"><a href="#annotated-cell-2-12"></a>    <span class="fu">lines!</span>(ax, [x_point, x_point], [<span class="fl">0</span>, y_point], color <span class="op">=</span> color, linestyle <span class="op">=</span> <span class="op">:</span>dash)</span>
<span id="annotated-cell-2-13"><a href="#annotated-cell-2-13"></a>    <span class="fu">lines!</span>(ax, [x_min, x_point], [y_point, y_point], color <span class="op">=</span> color, linestyle <span class="op">=</span> <span class="op">:</span>dash)</span>
<span id="annotated-cell-2-14"><a href="#annotated-cell-2-14"></a>    <span class="cf">return</span> y_point</span>
<span id="annotated-cell-2-15"><a href="#annotated-cell-2-15"></a><span class="kw">end</span></span>
<span id="annotated-cell-2-16"><a href="#annotated-cell-2-16"></a></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-2" data-target-annotation="5">5</button><span id="annotated-cell-2-17" class="code-annotation-target"><a href="#annotated-cell-2-17"></a><span class="kw">function</span> <span class="fu">add_inverse_cdf!</span>(ax, dist, p_target; color <span class="op">=</span> <span class="op">:</span>green, x_min <span class="op">=</span> <span class="op">-</span><span class="fl">4</span>)</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-2" data-target-annotation="6">6</button><span id="annotated-cell-2-18" class="code-annotation-target"><a href="#annotated-cell-2-18"></a>    x_inv <span class="op">=</span> <span class="fu">quantile</span>(dist, p_target)</span>
<span id="annotated-cell-2-19"><a href="#annotated-cell-2-19"></a>    y_actual <span class="op">=</span> <span class="fu">cdf</span>(dist, x_inv)</span>
<span id="annotated-cell-2-20"><a href="#annotated-cell-2-20"></a>    <span class="fu">scatter!</span>(ax, [x_inv], [y_actual], color <span class="op">=</span> color, markersize <span class="op">=</span> <span class="fl">8</span>)</span>
<span id="annotated-cell-2-21"><a href="#annotated-cell-2-21"></a>    <span class="fu">lines!</span>(ax, [x_inv, x_inv], [<span class="fl">0</span>, y_actual], color <span class="op">=</span> color, linestyle <span class="op">=</span> <span class="op">:</span>dash)</span>
<span id="annotated-cell-2-22"><a href="#annotated-cell-2-22"></a>    <span class="fu">lines!</span>(ax, [x_min, x_inv], [p_target, p_target], color <span class="op">=</span> color, linestyle <span class="op">=</span> <span class="op">:</span>dash)</span>
<span id="annotated-cell-2-23"><a href="#annotated-cell-2-23"></a>    <span class="cf">return</span> x_inv, y_actual</span>
<span id="annotated-cell-2-24"><a href="#annotated-cell-2-24"></a><span class="kw">end</span></span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-annotation">
<dl class="code-annotation-container-hidden code-annotation-container-grid">
<dt data-target-cell="annotated-cell-2" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-2" data-code-lines="1" data-code-annotation="1">Creates function to shade area under <code>PDF</code> curve between bounds <span class="math inline">\(a\)</span> and <span class="math inline">\(b\)</span>. By convention a function that ends in <code>!</code> modifies its arguments (in this case, a plot axis).</span>
</dd>
<dt data-target-cell="annotated-cell-2" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-2" data-code-lines="3" data-code-annotation="2">Vectorized evaluation of <code>PDF</code> using broadcasting (the <code>.</code> operator)<br>
</span>
</dd>
<dt data-target-cell="annotated-cell-2" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-2" data-code-lines="5" data-code-annotation="3">Calculate probability using <code>CDF</code> difference: <span class="math inline">\(P(a \leq X \leq b) = F(b) - F(a)\)</span></span>
</dd>
<dt data-target-cell="annotated-cell-2" data-target-annotation="4">4</dt>
<dd>
<span data-code-cell="annotated-cell-2" data-code-lines="9" data-code-annotation="4">Forward <code>CDF</code> operation: given <span class="math inline">\(x\)</span>, find <span class="math inline">\(F(x) = P(X \leq x)\)</span></span>
</dd>
<dt data-target-cell="annotated-cell-2" data-target-annotation="5">5</dt>
<dd>
<span data-code-cell="annotated-cell-2" data-code-lines="17" data-code-annotation="5">Inverse <code>CDF</code> operation: given probability <span class="math inline">\(p\)</span>, find <span class="math inline">\(x\)</span> such that <span class="math inline">\(F(x) = p\)</span><br>
</span>
</dd>
<dt data-target-cell="annotated-cell-2" data-target-annotation="6">6</dt>
<dd>
<span data-code-cell="annotated-cell-2" data-code-lines="18" data-code-annotation="6">The quantile function is the inverse of the <code>CDF</code></span>
</dd>
</dl>
</div>
</div>
<p>Next, we use <a href="https://docs.makie.org/"><code>Makie.jl</code></a> to creat the plot.</p>
<div id="cell-normal-distribution-example" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="annotated-cell-3"><pre class="sourceCode numberSource julia code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode julia"><span id="annotated-cell-3-1"><a href="#annotated-cell-3-1"></a><span class="kw">function</span> <span class="fu">plot_normal_pdf_cdf</span>()</span>
<span id="annotated-cell-3-2"><a href="#annotated-cell-3-2"></a>    μ, σ <span class="op">=</span> <span class="fl">0.0</span>, <span class="fl">1.0</span></span>
<span id="annotated-cell-3-3"><a href="#annotated-cell-3-3"></a>    x_range <span class="op">=</span> <span class="op">-</span><span class="fl">4</span><span class="op">:</span><span class="fl">0.01</span><span class="op">:</span><span class="fl">4</span></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-3" data-target-annotation="1">1</button><span id="annotated-cell-3-4" class="code-annotation-target"><a href="#annotated-cell-3-4"></a>    normal_dist <span class="op">=</span> <span class="fu">Normal</span>(μ, σ)</span>
<span id="annotated-cell-3-5"><a href="#annotated-cell-3-5"></a></span>
<span id="annotated-cell-3-6"><a href="#annotated-cell-3-6"></a>    fig <span class="op">=</span> <span class="fu">Figure</span>(size <span class="op">=</span> (<span class="fl">900</span>, <span class="fl">400</span>))</span>
<span id="annotated-cell-3-7"><a href="#annotated-cell-3-7"></a></span>
<span id="annotated-cell-3-8"><a href="#annotated-cell-3-8"></a>    ax1 <span class="op">=</span> <span class="fu">Axis</span>(fig[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="annotated-cell-3-9"><a href="#annotated-cell-3-9"></a>        xlabel <span class="op">=</span> L<span class="st">"x"</span>,</span>
<span id="annotated-cell-3-10"><a href="#annotated-cell-3-10"></a>        ylabel <span class="op">=</span> L<span class="st">"</span><span class="sc">\t</span><span class="st">ext{Density } </span><span class="sc">$</span>p<span class="st">(x)</span><span class="sc">$</span><span class="st">",</span></span>
<span id="annotated-cell-3-11"><a href="#annotated-cell-3-11"></a><span class="st">        </span>title<span class="st"> = "</span><span class="fu">Normal</span>(<span class="fl">0</span>, <span class="fl">1</span>) PDF<span class="st">")</span></span>
<span id="annotated-cell-3-12"><a href="#annotated-cell-3-12"></a></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-3" data-target-annotation="2">2</button><span id="annotated-cell-3-13" class="code-annotation-target"><a href="#annotated-cell-3-13"></a>    <span class="fu">lines!</span>(ax1, x_range, <span class="fu">pdf</span>.(normal_dist, x_range),</span>
<span id="annotated-cell-3-14"><a href="#annotated-cell-3-14"></a>        color <span class="op">=</span> <span class="op">:</span>blue, linewidth <span class="op">=</span> <span class="fl">2</span>, label <span class="op">=</span> L<span class="st">"p(x)"</span>)</span>
<span id="annotated-cell-3-15"><a href="#annotated-cell-3-15"></a></span>
<span id="annotated-cell-3-16"><a href="#annotated-cell-3-16"></a>    prob_area <span class="op">=</span> <span class="fu">add_pdf_area!</span>(ax1, normal_dist, <span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span>,</span>
<span id="annotated-cell-3-17"><a href="#annotated-cell-3-17"></a>        label <span class="op">=</span> L<span class="st">"P(-1 \leq X \leq 1)"</span>)</span>
<span id="annotated-cell-3-18"><a href="#annotated-cell-3-18"></a></span>
<span id="annotated-cell-3-19"><a href="#annotated-cell-3-19"></a>    <span class="fu">text!</span>(ax1, <span class="op">-</span><span class="fl">0.6</span>, <span class="fl">0.125</span>,</span>
<span id="annotated-cell-3-20"><a href="#annotated-cell-3-20"></a>        text <span class="op">=</span> L<span class="st">"</span><span class="sc">\t</span><span class="st">ext{Area} = %</span><span class="sc">$</span>(<span class="fu">round</span>(prob_area, digits<span class="op">=</span><span class="fl">3</span>))<span class="st">"</span>,</span>
<span id="annotated-cell-3-21"><a href="#annotated-cell-3-21"></a>        fontsize <span class="op">=</span> <span class="fl">14</span>, color <span class="op">=</span> <span class="op">:</span>black)</span>
<span id="annotated-cell-3-22"><a href="#annotated-cell-3-22"></a></span>
<span id="annotated-cell-3-23"><a href="#annotated-cell-3-23"></a>    <span class="fu">axislegend</span>(ax1, position <span class="op">=</span> <span class="op">:</span>rt)</span>
<span id="annotated-cell-3-24"><a href="#annotated-cell-3-24"></a></span>
<span id="annotated-cell-3-25"><a href="#annotated-cell-3-25"></a>    ax2 <span class="op">=</span> <span class="fu">Axis</span>(fig[<span class="fl">1</span>, <span class="fl">2</span>],</span>
<span id="annotated-cell-3-26"><a href="#annotated-cell-3-26"></a>        xlabel <span class="op">=</span> L<span class="st">"x"</span>,</span>
<span id="annotated-cell-3-27"><a href="#annotated-cell-3-27"></a>        ylabel <span class="op">=</span> L<span class="st">"</span><span class="sc">\t</span><span class="st">ext{Probability } </span><span class="sc">$</span>F<span class="st">(x)</span><span class="sc">$</span><span class="st">",</span></span>
<span id="annotated-cell-3-28"><a href="#annotated-cell-3-28"></a><span class="st">        </span>title<span class="st"> = "</span>Normal CDF<span class="op">:</span> Forward and Inverse<span class="st">")</span></span>
<span id="annotated-cell-3-29"><a href="#annotated-cell-3-29"></a></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-3" data-target-annotation="3">3</button><span id="annotated-cell-3-30" class="code-annotation-target"><a href="#annotated-cell-3-30"></a>    <span class="fu">lines!</span>(ax2, x_range, <span class="fu">cdf</span>.(normal_dist, x_range),</span>
<span id="annotated-cell-3-31"><a href="#annotated-cell-3-31"></a>        color <span class="op">=</span> <span class="op">:</span>blue, linewidth <span class="op">=</span> <span class="fl">2</span>, label <span class="op">=</span> L<span class="st">"</span><span class="sc">$</span>F<span class="st">(x)</span><span class="sc">$</span><span class="st">")</span></span>
<span id="annotated-cell-3-32"><a href="#annotated-cell-3-32"></a></span>
<span id="annotated-cell-3-33"><a href="#annotated-cell-3-33"></a><span class="st">    </span>y_point<span class="st"> = add_forward_cdf!(ax2, normal_dist, 1.0)</span></span>
<span id="annotated-cell-3-34"><a href="#annotated-cell-3-34"></a>    <span class="fu">text!</span>(ax2, <span class="fl">1.2</span>, y_point <span class="op">-</span> <span class="fl">0.1</span>,</span>
<span id="annotated-cell-3-35"><a href="#annotated-cell-3-35"></a>        text <span class="op">=</span> L<span class="st">"F(1) = %</span><span class="sc">$</span>(<span class="fu">round</span>(y_point, digits<span class="op">=</span><span class="fl">3</span>))<span class="st">"</span>, color <span class="op">=</span> <span class="op">:</span>red)</span>
<span id="annotated-cell-3-36"><a href="#annotated-cell-3-36"></a></span>
<span id="annotated-cell-3-37"><a href="#annotated-cell-3-37"></a>    x_inv, _ <span class="op">=</span> <span class="fu">add_inverse_cdf!</span>(ax2, normal_dist, <span class="fl">0.25</span>)</span>
<span id="annotated-cell-3-38"><a href="#annotated-cell-3-38"></a>    <span class="fu">text!</span>(ax2, x_inv <span class="op">-</span> <span class="fl">0.8</span>, <span class="fl">0.35</span>,</span>
<span id="annotated-cell-3-39"><a href="#annotated-cell-3-39"></a>        text <span class="op">=</span> L<span class="st">"F^{-1}(0.25) = %</span><span class="sc">$</span>(<span class="fu">round</span>(x_inv, digits<span class="op">=</span><span class="fl">2</span>))<span class="st">"</span>, color <span class="op">=</span> <span class="op">:</span>green)</span>
<span id="annotated-cell-3-40"><a href="#annotated-cell-3-40"></a></span>
<span id="annotated-cell-3-41"><a href="#annotated-cell-3-41"></a>    <span class="fu">axislegend</span>(ax2, position <span class="op">=</span> <span class="op">:</span>rb)</span>
<span id="annotated-cell-3-42"><a href="#annotated-cell-3-42"></a>    <span class="cf">return</span> fig</span>
<span id="annotated-cell-3-43"><a href="#annotated-cell-3-43"></a><span class="kw">end</span></span>
<span id="annotated-cell-3-44"><a href="#annotated-cell-3-44"></a></span>
<span id="annotated-cell-3-45"><a href="#annotated-cell-3-45"></a>fig_normal <span class="op">=</span> <span class="fu">plot_normal_pdf_cdf</span>()</span>
<span id="annotated-cell-3-46"><a href="#annotated-cell-3-46"></a>fig_normal</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-annotation">
<dl class="code-annotation-container-hidden code-annotation-container-grid">
<dt data-target-cell="annotated-cell-3" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-3" data-code-lines="4" data-code-annotation="1">Create <code>Normal(0,1)</code> distribution object from <code>Distributions.jl</code></span>
</dd>
<dt data-target-cell="annotated-cell-3" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-3" data-code-lines="13" data-code-annotation="2">Plot <code>PDF</code> using vectorized evaluation over x_range</span>
</dd>
<dt data-target-cell="annotated-cell-3" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-3" data-code-lines="30" data-code-annotation="3">Plot <code>CDF</code> showing cumulative probabilities</span>
</dd>
</dl>
</div>
<div class="cell-output cell-output-display" data-execution_count="1">
<div>
<figure class="figure">
<p><img src="probability-stats-examples_files/figure-html/normal-distribution-example-output-1.svg" id="normal-distribution-example" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>The normal distribution demonstrates these concepts for continuous variables, where probabilities correspond to areas under smooth curves. Next, we examine the Poisson distribution for discrete random variables, where probabilities correspond to point masses and the <code>CDF</code> becomes a step function.</p>
<p>We define some more helper functions for discrete distributions.</p>
<div id="discrete-helpers" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="annotated-cell-4"><pre class="sourceCode numberSource julia code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode julia"><button class="code-annotation-anchor" data-target-cell="annotated-cell-4" data-target-annotation="1">1</button><span id="annotated-cell-4-1" class="code-annotation-target"><a href="#annotated-cell-4-1"></a><span class="kw">function</span> <span class="fu">plot_pmf_stems!</span>(ax, dist, x_range; color <span class="op">=</span> <span class="op">:</span>blue, linewidth <span class="op">=</span> <span class="fl">3</span>, markersize <span class="op">=</span> <span class="fl">8</span>)</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-4" data-target-annotation="2">2</button><span id="annotated-cell-4-2" class="code-annotation-target"><a href="#annotated-cell-4-2"></a>    pmf_vals <span class="op">=</span> <span class="fu">pdf</span>.(dist, x_range)</span>
<span id="annotated-cell-4-3"><a href="#annotated-cell-4-3"></a>    <span class="cf">for</span> (i, x) <span class="kw">in</span> <span class="fu">enumerate</span>(x_range)</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-4" data-target-annotation="3">3</button><span id="annotated-cell-4-4" class="code-annotation-target"><a href="#annotated-cell-4-4"></a>        <span class="fu">lines!</span>(ax, [x, x], [<span class="fl">0</span>, pmf_vals[i]], color <span class="op">=</span> color, linewidth <span class="op">=</span> linewidth)</span>
<span id="annotated-cell-4-5"><a href="#annotated-cell-4-5"></a>        <span class="fu">scatter!</span>(ax, [x], [pmf_vals[i]], color <span class="op">=</span> color, markersize <span class="op">=</span> markersize)</span>
<span id="annotated-cell-4-6"><a href="#annotated-cell-4-6"></a>    <span class="cf">end</span></span>
<span id="annotated-cell-4-7"><a href="#annotated-cell-4-7"></a>    <span class="cf">return</span> pmf_vals</span>
<span id="annotated-cell-4-8"><a href="#annotated-cell-4-8"></a><span class="kw">end</span></span>
<span id="annotated-cell-4-9"><a href="#annotated-cell-4-9"></a></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-4" data-target-annotation="4">4</button><span id="annotated-cell-4-10" class="code-annotation-target"><a href="#annotated-cell-4-10"></a><span class="kw">function</span> <span class="fu">highlight_pmf_mass!</span>(ax, dist, x_range; color <span class="op">=</span> <span class="op">:</span>orange)</span>
<span id="annotated-cell-4-11"><a href="#annotated-cell-4-11"></a>    pmf_vals <span class="op">=</span> <span class="fu">pdf</span>.(dist, x_range)</span>
<span id="annotated-cell-4-12"><a href="#annotated-cell-4-12"></a>    <span class="cf">for</span> (i, x) <span class="kw">in</span> <span class="fu">enumerate</span>(x_range)</span>
<span id="annotated-cell-4-13"><a href="#annotated-cell-4-13"></a>        <span class="fu">lines!</span>(ax, [x, x], [<span class="fl">0</span>, pmf_vals[i]], color <span class="op">=</span> color, linewidth <span class="op">=</span> <span class="fl">5</span>)</span>
<span id="annotated-cell-4-14"><a href="#annotated-cell-4-14"></a>        <span class="fu">scatter!</span>(ax, [x], [pmf_vals[i]], color <span class="op">=</span> color, markersize <span class="op">=</span> <span class="fl">10</span>)</span>
<span id="annotated-cell-4-15"><a href="#annotated-cell-4-15"></a>    <span class="cf">end</span></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-4" data-target-annotation="5">5</button><span id="annotated-cell-4-16" class="code-annotation-target"><a href="#annotated-cell-4-16"></a>    <span class="cf">return</span> <span class="fu">sum</span>(pmf_vals)</span>
<span id="annotated-cell-4-17"><a href="#annotated-cell-4-17"></a><span class="kw">end</span></span>
<span id="annotated-cell-4-18"><a href="#annotated-cell-4-18"></a></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-4" data-target-annotation="6">6</button><span id="annotated-cell-4-19" class="code-annotation-target"><a href="#annotated-cell-4-19"></a><span class="kw">function</span> <span class="fu">plot_discrete_cdf!</span>(ax, dist, x_range; color <span class="op">=</span> <span class="op">:</span>blue, linewidth <span class="op">=</span> <span class="fl">2</span>, markersize <span class="op">=</span> <span class="fl">6</span>)</span>
<span id="annotated-cell-4-20"><a href="#annotated-cell-4-20"></a>    cdf_vals <span class="op">=</span> <span class="fu">cdf</span>.(dist, x_range)</span>
<span id="annotated-cell-4-21"><a href="#annotated-cell-4-21"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span>(<span class="fu">length</span>(x_range)<span class="op">-</span><span class="fl">1</span>)</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-4" data-target-annotation="7">7</button><span id="annotated-cell-4-22" class="code-annotation-target"><a href="#annotated-cell-4-22"></a>        <span class="fu">lines!</span>(ax, [x_range[i], x_range[i<span class="op">+</span><span class="fl">1</span>]], [cdf_vals[i], cdf_vals[i]],</span>
<span id="annotated-cell-4-23"><a href="#annotated-cell-4-23"></a>            color <span class="op">=</span> color, linewidth <span class="op">=</span> linewidth)</span>
<span id="annotated-cell-4-24"><a href="#annotated-cell-4-24"></a>    <span class="cf">end</span></span>
<span id="annotated-cell-4-25"><a href="#annotated-cell-4-25"></a>    <span class="fu">scatter!</span>(ax, x_range, cdf_vals, color <span class="op">=</span> color, markersize <span class="op">=</span> markersize)</span>
<span id="annotated-cell-4-26"><a href="#annotated-cell-4-26"></a>    <span class="cf">return</span> cdf_vals</span>
<span id="annotated-cell-4-27"><a href="#annotated-cell-4-27"></a><span class="kw">end</span></span>
<span id="annotated-cell-4-28"><a href="#annotated-cell-4-28"></a></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-4" data-target-annotation="8">8</button><span id="annotated-cell-4-29" class="code-annotation-target"><a href="#annotated-cell-4-29"></a><span class="kw">function</span> <span class="fu">add_discrete_forward_cdf!</span>(ax, dist, x_point; color <span class="op">=</span> <span class="op">:</span>red, x_min <span class="op">=</span> <span class="fl">0</span>)</span>
<span id="annotated-cell-4-30"><a href="#annotated-cell-4-30"></a>    y_point <span class="op">=</span> <span class="fu">cdf</span>(dist, x_point)</span>
<span id="annotated-cell-4-31"><a href="#annotated-cell-4-31"></a>    <span class="fu">scatter!</span>(ax, [x_point], [y_point], color <span class="op">=</span> color, markersize <span class="op">=</span> <span class="fl">10</span>)</span>
<span id="annotated-cell-4-32"><a href="#annotated-cell-4-32"></a>    <span class="fu">lines!</span>(ax, [x_point, x_point], [<span class="fl">0</span>, y_point], color <span class="op">=</span> color, linestyle <span class="op">=</span> <span class="op">:</span>dash)</span>
<span id="annotated-cell-4-33"><a href="#annotated-cell-4-33"></a>    <span class="fu">lines!</span>(ax, [x_min, x_point], [y_point, y_point], color <span class="op">=</span> color, linestyle <span class="op">=</span> <span class="op">:</span>dash)</span>
<span id="annotated-cell-4-34"><a href="#annotated-cell-4-34"></a>    <span class="cf">return</span> y_point</span>
<span id="annotated-cell-4-35"><a href="#annotated-cell-4-35"></a><span class="kw">end</span></span>
<span id="annotated-cell-4-36"><a href="#annotated-cell-4-36"></a></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-4" data-target-annotation="9">9</button><span id="annotated-cell-4-37" class="code-annotation-target"><a href="#annotated-cell-4-37"></a><span class="kw">function</span> <span class="fu">add_discrete_inverse_cdf!</span>(ax, dist, p_target; color <span class="op">=</span> <span class="op">:</span>green, x_min <span class="op">=</span> <span class="fl">0</span>)</span>
<span id="annotated-cell-4-38"><a href="#annotated-cell-4-38"></a>    x_inv <span class="op">=</span> <span class="fu">quantile</span>(dist, p_target)</span>
<span id="annotated-cell-4-39"><a href="#annotated-cell-4-39"></a>    y_actual <span class="op">=</span> <span class="fu">cdf</span>(dist, x_inv)</span>
<span id="annotated-cell-4-40"><a href="#annotated-cell-4-40"></a>    <span class="fu">scatter!</span>(ax, [x_inv], [y_actual], color <span class="op">=</span> color, markersize <span class="op">=</span> <span class="fl">10</span>)</span>
<span id="annotated-cell-4-41"><a href="#annotated-cell-4-41"></a>    <span class="fu">lines!</span>(ax, [x_min, x_inv], [p_target, p_target], color <span class="op">=</span> color, linestyle <span class="op">=</span> <span class="op">:</span>dash)</span>
<span id="annotated-cell-4-42"><a href="#annotated-cell-4-42"></a>    <span class="cf">return</span> x_inv, y_actual</span>
<span id="annotated-cell-4-43"><a href="#annotated-cell-4-43"></a><span class="kw">end</span></span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-annotation">
<dl class="code-annotation-container-hidden code-annotation-container-grid">
<dt data-target-cell="annotated-cell-4" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-4" data-code-lines="1" data-code-annotation="1">Plots discrete PMF as stems (vertical lines) with points at the top</span>
</dd>
<dt data-target-cell="annotated-cell-4" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-4" data-code-lines="2" data-code-annotation="2">For discrete distributions, <code>pdf()</code> returns probability mass <span class="math inline">\(P(X = x)\)</span></span>
</dd>
<dt data-target-cell="annotated-cell-4" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-4" data-code-lines="4" data-code-annotation="3">Draw vertical line from 0 to probability mass at each <span class="math inline">\(x\)</span> value</span>
</dd>
<dt data-target-cell="annotated-cell-4" data-target-annotation="4">4</dt>
<dd>
<span data-code-cell="annotated-cell-4" data-code-lines="10" data-code-annotation="4">Highlights specific probability masses in different color</span>
</dd>
<dt data-target-cell="annotated-cell-4" data-target-annotation="5">5</dt>
<dd>
<span data-code-cell="annotated-cell-4" data-code-lines="16" data-code-annotation="5">Sum individual masses to get cumulative probability <span class="math inline">\(P(X \leq 2)\)</span></span>
</dd>
<dt data-target-cell="annotated-cell-4" data-target-annotation="6">6</dt>
<dd>
<span data-code-cell="annotated-cell-4" data-code-lines="19" data-code-annotation="6">Creates step function visualization for discrete <code>CDF</code></span>
</dd>
<dt data-target-cell="annotated-cell-4" data-target-annotation="7">7</dt>
<dd>
<span data-code-cell="annotated-cell-4" data-code-lines="22" data-code-annotation="7">Horizontal line segments show constant <code>CDF</code> values between integers</span>
</dd>
<dt data-target-cell="annotated-cell-4" data-target-annotation="8">8</dt>
<dd>
<span data-code-cell="annotated-cell-4" data-code-lines="29" data-code-annotation="8">Same forward operation logic but adapted for discrete distributions</span>
</dd>
<dt data-target-cell="annotated-cell-4" data-target-annotation="9">9</dt>
<dd>
<span data-code-cell="annotated-cell-4" data-code-lines="37" data-code-annotation="9">Inverse operation finds the smallest integer <span class="math inline">\(x\)</span> such that <span class="math inline">\(F(x) \geq p\)</span></span>
</dd>
</dl>
</div>
</div>
<p>Now we create the plot for the Poisson distribution.</p>
<div id="cell-poisson-distribution-example" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="annotated-cell-5"><pre class="sourceCode numberSource julia code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode julia"><span id="annotated-cell-5-1"><a href="#annotated-cell-5-1"></a><span class="kw">function</span> <span class="fu">plot_poisson_pmf_cdf</span>()</span>
<span id="annotated-cell-5-2"><a href="#annotated-cell-5-2"></a>    λ <span class="op">=</span> <span class="fl">3.0</span></span>
<span id="annotated-cell-5-3"><a href="#annotated-cell-5-3"></a>    x_range <span class="op">=</span> <span class="fl">0</span><span class="op">:</span><span class="fl">10</span></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-5" data-target-annotation="1">1</button><span id="annotated-cell-5-4" class="code-annotation-target"><a href="#annotated-cell-5-4"></a>    poisson_dist <span class="op">=</span> <span class="fu">Poisson</span>(λ)</span>
<span id="annotated-cell-5-5"><a href="#annotated-cell-5-5"></a></span>
<span id="annotated-cell-5-6"><a href="#annotated-cell-5-6"></a>    fig <span class="op">=</span> <span class="fu">Figure</span>(size <span class="op">=</span> (<span class="fl">900</span>, <span class="fl">400</span>))</span>
<span id="annotated-cell-5-7"><a href="#annotated-cell-5-7"></a></span>
<span id="annotated-cell-5-8"><a href="#annotated-cell-5-8"></a>    ax1 <span class="op">=</span> <span class="fu">Axis</span>(fig[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="annotated-cell-5-9"><a href="#annotated-cell-5-9"></a>        xlabel <span class="op">=</span> L<span class="st">"x"</span>,</span>
<span id="annotated-cell-5-10"><a href="#annotated-cell-5-10"></a>        ylabel <span class="op">=</span> L<span class="st">"P(X = x)"</span>,</span>
<span id="annotated-cell-5-11"><a href="#annotated-cell-5-11"></a>        title <span class="op">=</span> L<span class="st">"</span><span class="sc">\t</span><span class="st">ext{Poisson}(3) </span><span class="sc">\t</span><span class="st">ext{ PMF}"</span>,</span>
<span id="annotated-cell-5-12"><a href="#annotated-cell-5-12"></a>        xticks <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>)</span>
<span id="annotated-cell-5-13"><a href="#annotated-cell-5-13"></a></span>
<span id="annotated-cell-5-14"><a href="#annotated-cell-5-14"></a>    <span class="fu">plot_pmf_stems!</span>(ax1, poisson_dist, x_range)</span>
<span id="annotated-cell-5-15"><a href="#annotated-cell-5-15"></a></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-5" data-target-annotation="2">2</button><span id="annotated-cell-5-16" class="code-annotation-target"><a href="#annotated-cell-5-16"></a>    prob_mass <span class="op">=</span> <span class="fu">highlight_pmf_mass!</span>(ax1, poisson_dist, <span class="fl">0</span><span class="op">:</span><span class="fl">2</span>)</span>
<span id="annotated-cell-5-17"><a href="#annotated-cell-5-17"></a></span>
<span id="annotated-cell-5-18"><a href="#annotated-cell-5-18"></a>    <span class="fu">text!</span>(ax1, <span class="fl">6</span>, <span class="fl">0.15</span>,</span>
<span id="annotated-cell-5-19"><a href="#annotated-cell-5-19"></a>        text <span class="op">=</span> L<span class="st">"P(X \leq 2) = %</span><span class="sc">$</span>(<span class="fu">round</span>(prob_mass, digits<span class="op">=</span><span class="fl">3</span>))<span class="st">"</span>,</span>
<span id="annotated-cell-5-20"><a href="#annotated-cell-5-20"></a>        fontsize <span class="op">=</span> <span class="fl">14</span>, color <span class="op">=</span> <span class="op">:</span>black)</span>
<span id="annotated-cell-5-21"><a href="#annotated-cell-5-21"></a></span>
<span id="annotated-cell-5-22"><a href="#annotated-cell-5-22"></a>    ax2 <span class="op">=</span> <span class="fu">Axis</span>(fig[<span class="fl">1</span>, <span class="fl">2</span>],</span>
<span id="annotated-cell-5-23"><a href="#annotated-cell-5-23"></a>        xlabel <span class="op">=</span> L<span class="st">"x"</span>,</span>
<span id="annotated-cell-5-24"><a href="#annotated-cell-5-24"></a>        ylabel <span class="op">=</span> L<span class="st">"</span><span class="sc">\t</span><span class="st">ext{Probability } F(x)"</span>,</span>
<span id="annotated-cell-5-25"><a href="#annotated-cell-5-25"></a>        title <span class="op">=</span> L<span class="st">"</span><span class="sc">\t</span><span class="st">ext{Poisson CDF: Forward and Inverse}"</span>,</span>
<span id="annotated-cell-5-26"><a href="#annotated-cell-5-26"></a>        xticks <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>)</span>
<span id="annotated-cell-5-27"><a href="#annotated-cell-5-27"></a></span>
<span id="annotated-cell-5-28"><a href="#annotated-cell-5-28"></a>    <span class="fu">plot_discrete_cdf!</span>(ax2, poisson_dist, x_range)</span>
<span id="annotated-cell-5-29"><a href="#annotated-cell-5-29"></a></span>
<span id="annotated-cell-5-30"><a href="#annotated-cell-5-30"></a>    y_point <span class="op">=</span> <span class="fu">add_discrete_forward_cdf!</span>(ax2, poisson_dist, <span class="fl">4</span>)</span>
<span id="annotated-cell-5-31"><a href="#annotated-cell-5-31"></a>    <span class="fu">text!</span>(ax2, <span class="fl">4.2</span>, y_point <span class="op">-</span> <span class="fl">0.1</span>,</span>
<span id="annotated-cell-5-32"><a href="#annotated-cell-5-32"></a>        text <span class="op">=</span> L<span class="st">"F(4) = %</span><span class="sc">$</span>(<span class="fu">round</span>(y_point, digits<span class="op">=</span><span class="fl">3</span>))<span class="st">"</span>, color <span class="op">=</span> <span class="op">:</span>red)</span>
<span id="annotated-cell-5-33"><a href="#annotated-cell-5-33"></a></span>
<span id="annotated-cell-5-34"><a href="#annotated-cell-5-34"></a>    x_inv, _ <span class="op">=</span> <span class="fu">add_discrete_inverse_cdf!</span>(ax2, poisson_dist, <span class="fl">0.4</span>)</span>
<span id="annotated-cell-5-35"><a href="#annotated-cell-5-35"></a>    <span class="fu">text!</span>(ax2, x_inv <span class="op">-</span> <span class="fl">1.5</span>, <span class="fl">0.5</span>,</span>
<span id="annotated-cell-5-36"><a href="#annotated-cell-5-36"></a>        text <span class="op">=</span> L<span class="st">"F^{-1}(0.4) = %</span><span class="sc">$</span>(<span class="fu">Int</span>(x_inv))<span class="st">"</span>, color <span class="op">=</span> <span class="op">:</span>green)</span>
<span id="annotated-cell-5-37"><a href="#annotated-cell-5-37"></a></span>
<span id="annotated-cell-5-38"><a href="#annotated-cell-5-38"></a>    <span class="cf">return</span> fig</span>
<span id="annotated-cell-5-39"><a href="#annotated-cell-5-39"></a><span class="kw">end</span></span>
<span id="annotated-cell-5-40"><a href="#annotated-cell-5-40"></a></span>
<span id="annotated-cell-5-41"><a href="#annotated-cell-5-41"></a>fig_poisson <span class="op">=</span> <span class="fu">plot_poisson_pmf_cdf</span>()</span>
<span id="annotated-cell-5-42"><a href="#annotated-cell-5-42"></a>fig_poisson</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-annotation">
<dl class="code-annotation-container-hidden code-annotation-container-grid">
<dt data-target-cell="annotated-cell-5" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-5" data-code-lines="4" data-code-annotation="1">Create Poisson distribution with rate parameter <span class="math inline">\(\lambda = 3\)</span></span>
</dd>
<dt data-target-cell="annotated-cell-5" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-5" data-code-lines="16" data-code-annotation="2">Calculate and highlight <span class="math inline">\(P(X \leq 2)\)</span> by summing individual probability masses</span>
</dd>
</dl>
</div>
<div class="cell-output cell-output-display" data-execution_count="1">
<div>
<figure class="figure">
<p><img src="probability-stats-examples_files/figure-html/poisson-distribution-example-output-1.svg" id="poisson-distribution-example" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="joint-marginal-and-conditional-distributions" class="level2">
<h2 class="anchored" data-anchor-id="joint-marginal-and-conditional-distributions">Joint, marginal, and conditional distributions</h2>
<p>This example demonstrates the relationships between joint, marginal, and conditional distributions using a bivariate normal distribution. The visualization shows how these concepts connect: marginal distributions are the “shadows” of the joint distribution, while conditional distributions are “slices” at specific values.</p>
<div id="cell-joint-marginal-conditional-example" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="annotated-cell-6"><pre class="sourceCode numberSource julia code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode julia"><span id="annotated-cell-6-1"><a href="#annotated-cell-6-1"></a><span class="kw">function</span> <span class="fu">plot_joint_marginal_conditional</span>()</span>
<span id="annotated-cell-6-2"><a href="#annotated-cell-6-2"></a>    <span class="co"># Parameters for bivariate normal</span></span>
<span id="annotated-cell-6-3"><a href="#annotated-cell-6-3"></a>    μ₁, μ₂ <span class="op">=</span> <span class="fl">2.0</span>, <span class="fl">1.0</span></span>
<span id="annotated-cell-6-4"><a href="#annotated-cell-6-4"></a>    σ₁, σ₂ <span class="op">=</span> <span class="fl">1.0</span>, <span class="fl">0.8</span></span>
<span id="annotated-cell-6-5"><a href="#annotated-cell-6-5"></a>    ρ <span class="op">=</span> <span class="fl">0.6</span>  <span class="co"># correlation coefficient</span></span>
<span id="annotated-cell-6-6"><a href="#annotated-cell-6-6"></a></span>
<span id="annotated-cell-6-7"><a href="#annotated-cell-6-7"></a>    <span class="co"># Create bivariate normal distribution</span></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-6" data-target-annotation="1">1</button><span id="annotated-cell-6-8" class="code-annotation-target"><a href="#annotated-cell-6-8"></a>    Σ <span class="op">=</span> [σ₁<span class="op">^</span><span class="fl">2</span> ρ<span class="op">*</span>σ₁<span class="op">*</span>σ₂; ρ<span class="op">*</span>σ₁<span class="op">*</span>σ₂ σ₂<span class="op">^</span><span class="fl">2</span>]</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-6" data-target-annotation="2">2</button><span id="annotated-cell-6-9" class="code-annotation-target"><a href="#annotated-cell-6-9"></a>    mvn <span class="op">=</span> <span class="fu">MvNormal</span>([μ₁, μ₂], Σ)</span>
<span id="annotated-cell-6-10"><a href="#annotated-cell-6-10"></a></span>
<span id="annotated-cell-6-11"><a href="#annotated-cell-6-11"></a>    <span class="co"># Generate samples</span></span>
<span id="annotated-cell-6-12"><a href="#annotated-cell-6-12"></a>    <span class="bu">Random</span>.<span class="fu">seed!</span>(<span class="fl">123</span>)</span>
<span id="annotated-cell-6-13"><a href="#annotated-cell-6-13"></a>    n_samples <span class="op">=</span> <span class="fl">1000</span></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-6" data-target-annotation="3">3</button><span id="annotated-cell-6-14" class="code-annotation-target"><a href="#annotated-cell-6-14"></a>    samples <span class="op">=</span> <span class="fu">rand</span>(mvn, n_samples)</span>
<span id="annotated-cell-6-15"><a href="#annotated-cell-6-15"></a>    x_samples <span class="op">=</span> samples[<span class="fl">1</span>, <span class="op">:</span>]</span>
<span id="annotated-cell-6-16"><a href="#annotated-cell-6-16"></a>    y_samples <span class="op">=</span> samples[<span class="fl">2</span>, <span class="op">:</span>]</span>
<span id="annotated-cell-6-17"><a href="#annotated-cell-6-17"></a></span>
<span id="annotated-cell-6-18"><a href="#annotated-cell-6-18"></a>    <span class="co"># Create figure with custom layout</span></span>
<span id="annotated-cell-6-19"><a href="#annotated-cell-6-19"></a>    fig <span class="op">=</span> <span class="fu">Figure</span>(size <span class="op">=</span> (<span class="fl">1000</span>, <span class="fl">800</span>))</span>
<span id="annotated-cell-6-20"><a href="#annotated-cell-6-20"></a></span>
<span id="annotated-cell-6-21"><a href="#annotated-cell-6-21"></a>    <span class="co"># Main scatter plot (bottom left)</span></span>
<span id="annotated-cell-6-22"><a href="#annotated-cell-6-22"></a>    ax_main <span class="op">=</span> <span class="fu">Axis</span>(fig[<span class="fl">2</span>, <span class="fl">1</span>],</span>
<span id="annotated-cell-6-23"><a href="#annotated-cell-6-23"></a>        xlabel <span class="op">=</span> L<span class="st">"X"</span>,</span>
<span id="annotated-cell-6-24"><a href="#annotated-cell-6-24"></a>        ylabel <span class="op">=</span> L<span class="st">"Y"</span>,</span>
<span id="annotated-cell-6-25"><a href="#annotated-cell-6-25"></a>        title <span class="op">=</span> <span class="st">"Joint Distribution"</span>)</span>
<span id="annotated-cell-6-26"><a href="#annotated-cell-6-26"></a></span>
<span id="annotated-cell-6-27"><a href="#annotated-cell-6-27"></a>    <span class="co"># Scatter plot of samples</span></span>
<span id="annotated-cell-6-28"><a href="#annotated-cell-6-28"></a>    <span class="fu">scatter!</span>(ax_main, x_samples, y_samples,</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-6" data-target-annotation="4">4</button><span id="annotated-cell-6-29" class="code-annotation-target"><a href="#annotated-cell-6-29"></a>        color <span class="op">=</span> (<span class="op">:</span>blue, <span class="fl">0.4</span>), markersize <span class="op">=</span> <span class="fl">4</span>)</span>
<span id="annotated-cell-6-30"><a href="#annotated-cell-6-30"></a></span>
<span id="annotated-cell-6-31"><a href="#annotated-cell-6-31"></a>    <span class="co"># Add conditional distribution line (vertical slice at x=2.5)</span></span>
<span id="annotated-cell-6-32"><a href="#annotated-cell-6-32"></a>    x_condition <span class="op">=</span> <span class="fl">2.5</span></span>
<span id="annotated-cell-6-33"><a href="#annotated-cell-6-33"></a>    <span class="fu">vlines!</span>(ax_main, [x_condition], color <span class="op">=</span> <span class="op">:</span>red, linewidth <span class="op">=</span> <span class="fl">3</span>,</span>
<span id="annotated-cell-6-34"><a href="#annotated-cell-6-34"></a>        linestyle <span class="op">=</span> <span class="op">:</span>dash, label <span class="op">=</span> L<span class="st">"</span><span class="sc">$</span>X<span class="st"> = %</span><span class="sc">$</span>(x_condition)<span class="sc">$</span><span class="st">")</span></span>
<span id="annotated-cell-6-35"><a href="#annotated-cell-6-35"></a></span>
<span id="annotated-cell-6-36"><a href="#annotated-cell-6-36"></a><span class="st">    # </span>Marginal<span class="st"> histogram for X (top)</span></span>
<span id="annotated-cell-6-37"><a href="#annotated-cell-6-37"></a>    ax_top <span class="op">=</span> <span class="fu">Axis</span>(fig[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="annotated-cell-6-38"><a href="#annotated-cell-6-38"></a>        ylabel <span class="op">=</span> <span class="st">"Density"</span>,</span>
<span id="annotated-cell-6-39"><a href="#annotated-cell-6-39"></a>        title <span class="op">=</span> L<span class="st">"</span><span class="sc">\t</span><span class="st">ext{Marginal Distribution of }X"</span>)</span>
<span id="annotated-cell-6-40"><a href="#annotated-cell-6-40"></a></span>
<span id="annotated-cell-6-41"><a href="#annotated-cell-6-41"></a>    <span class="fu">hist!</span>(ax_top, x_samples, bins <span class="op">=</span> <span class="fl">30</span>, normalization <span class="op">=</span> <span class="op">:</span>pdf,</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-6" data-target-annotation="5">5</button><span id="annotated-cell-6-42" class="code-annotation-target"><a href="#annotated-cell-6-42"></a>        color <span class="op">=</span> (<span class="op">:</span>green, <span class="fl">0.6</span>))</span>
<span id="annotated-cell-6-43"><a href="#annotated-cell-6-43"></a></span>
<span id="annotated-cell-6-44"><a href="#annotated-cell-6-44"></a>    <span class="co"># Plot true marginal density</span></span>
<span id="annotated-cell-6-45"><a href="#annotated-cell-6-45"></a>    x_range <span class="op">=</span> <span class="fu">range</span>(<span class="op">-</span><span class="fl">1</span>, <span class="fl">5</span>, length <span class="op">=</span> <span class="fl">100</span>)</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-6" data-target-annotation="6">6</button><span id="annotated-cell-6-46" class="code-annotation-target"><a href="#annotated-cell-6-46"></a>    marginal_x <span class="op">=</span> <span class="fu">Normal</span>(μ₁, σ₁)</span>
<span id="annotated-cell-6-47"><a href="#annotated-cell-6-47"></a>    <span class="fu">lines!</span>(ax_top, x_range, <span class="fu">pdf</span>.(marginal_x, x_range),</span>
<span id="annotated-cell-6-48"><a href="#annotated-cell-6-48"></a>        color <span class="op">=</span> <span class="op">:</span>green, linewidth <span class="op">=</span> <span class="fl">3</span>, label <span class="op">=</span> <span class="st">"True marginal"</span>)</span>
<span id="annotated-cell-6-49"><a href="#annotated-cell-6-49"></a></span>
<span id="annotated-cell-6-50"><a href="#annotated-cell-6-50"></a>    <span class="co"># Add vertical line at conditioning value</span></span>
<span id="annotated-cell-6-51"><a href="#annotated-cell-6-51"></a>    <span class="fu">vlines!</span>(ax_top, [x_condition], color <span class="op">=</span> <span class="op">:</span>red, linewidth <span class="op">=</span> <span class="fl">2</span>, linestyle <span class="op">=</span> <span class="op">:</span>dash)</span>
<span id="annotated-cell-6-52"><a href="#annotated-cell-6-52"></a></span>
<span id="annotated-cell-6-53"><a href="#annotated-cell-6-53"></a>    <span class="co"># Marginal histogram for Y (right)</span></span>
<span id="annotated-cell-6-54"><a href="#annotated-cell-6-54"></a>    ax_right <span class="op">=</span> <span class="fu">Axis</span>(fig[<span class="fl">2</span>, <span class="fl">2</span>],</span>
<span id="annotated-cell-6-55"><a href="#annotated-cell-6-55"></a>        xlabel <span class="op">=</span> <span class="st">"Density"</span>,</span>
<span id="annotated-cell-6-56"><a href="#annotated-cell-6-56"></a>        title <span class="op">=</span> L<span class="st">"Marginal Distribution of </span><span class="sc">$</span>Y<span class="sc">$</span><span class="st">")</span></span>
<span id="annotated-cell-6-57"><a href="#annotated-cell-6-57"></a></span>
<span id="annotated-cell-6-58"><a href="#annotated-cell-6-58"></a><span class="st">    # </span>Rotate<span class="st"> histogram for Y (plot horizontally)</span></span>
<span id="annotated-cell-6-59"><a href="#annotated-cell-6-59"></a>    <span class="fu">hist!</span>(ax_right, y_samples, bins <span class="op">=</span> <span class="fl">30</span>, normalization <span class="op">=</span> <span class="op">:</span>pdf,</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-6" data-target-annotation="7">7</button><span id="annotated-cell-6-60" class="code-annotation-target"><a href="#annotated-cell-6-60"></a>        color <span class="op">=</span> (<span class="op">:</span>orange, <span class="fl">0.6</span>), direction <span class="op">=</span> <span class="op">:</span>x)</span>
<span id="annotated-cell-6-61"><a href="#annotated-cell-6-61"></a></span>
<span id="annotated-cell-6-62"><a href="#annotated-cell-6-62"></a>    <span class="co"># Plot true marginal density</span></span>
<span id="annotated-cell-6-63"><a href="#annotated-cell-6-63"></a>    y_range <span class="op">=</span> <span class="fu">range</span>(<span class="op">-</span><span class="fl">2</span>, <span class="fl">4</span>, length <span class="op">=</span> <span class="fl">100</span>)</span>
<span id="annotated-cell-6-64"><a href="#annotated-cell-6-64"></a>    marginal_y <span class="op">=</span> <span class="fu">Normal</span>(μ₂, σ₂)</span>
<span id="annotated-cell-6-65"><a href="#annotated-cell-6-65"></a>    <span class="fu">lines!</span>(ax_right, <span class="fu">pdf</span>.(marginal_y, y_range), y_range,</span>
<span id="annotated-cell-6-66"><a href="#annotated-cell-6-66"></a>        color <span class="op">=</span> <span class="op">:</span>orange, linewidth <span class="op">=</span> <span class="fl">3</span>, label <span class="op">=</span> <span class="st">"True marginal"</span>)</span>
<span id="annotated-cell-6-67"><a href="#annotated-cell-6-67"></a></span>
<span id="annotated-cell-6-68"><a href="#annotated-cell-6-68"></a>    <span class="co"># Conditional distribution plot (top right)</span></span>
<span id="annotated-cell-6-69"><a href="#annotated-cell-6-69"></a>    ax_cond <span class="op">=</span> <span class="fu">Axis</span>(fig[<span class="fl">1</span>, <span class="fl">2</span>],</span>
<span id="annotated-cell-6-70"><a href="#annotated-cell-6-70"></a>        xlabel <span class="op">=</span> L<span class="st">"Y"</span>,</span>
<span id="annotated-cell-6-71"><a href="#annotated-cell-6-71"></a>        ylabel <span class="op">=</span> <span class="st">"Conditional Density"</span>,</span>
<span id="annotated-cell-6-72"><a href="#annotated-cell-6-72"></a>        title <span class="op">=</span> L<span class="st">"Conditional: </span><span class="sc">$</span>p<span class="st">(Y \mid X = %</span><span class="sc">$</span>(x_condition)<span class="st">)</span><span class="sc">$</span><span class="st">")</span></span>
<span id="annotated-cell-6-73"><a href="#annotated-cell-6-73"></a></span>
<span id="annotated-cell-6-74"><a href="#annotated-cell-6-74"></a><span class="st">    # </span>Calculate<span class="st"> conditional distribution parameters</span></span>
<span id="annotated-cell-6-75"><a href="#annotated-cell-6-75"></a>    <span class="co"># For bivariate normal: Y|X ~ N(μ₂ + ρ(σ₂/σ₁)(x - μ₁), σ₂²(1 - ρ²))</span></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-6" data-target-annotation="8">8</button><span id="annotated-cell-6-76" class="code-annotation-target"><a href="#annotated-cell-6-76"></a>    μ_conditional <span class="op">=</span> μ₂ <span class="op">+</span> ρ <span class="op">*</span> (σ₂ <span class="op">/</span> σ₁) <span class="op">*</span> (x_condition <span class="op">-</span> μ₁)</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-6" data-target-annotation="9">9</button><span id="annotated-cell-6-77" class="code-annotation-target"><a href="#annotated-cell-6-77"></a>    σ_conditional <span class="op">=</span> σ₂ <span class="op">*</span> <span class="fu">sqrt</span>(<span class="fl">1</span> <span class="op">-</span> ρ<span class="op">^</span><span class="fl">2</span>)</span>
<span id="annotated-cell-6-78"><a href="#annotated-cell-6-78"></a>    conditional_dist <span class="op">=</span> <span class="fu">Normal</span>(μ_conditional, σ_conditional)</span>
<span id="annotated-cell-6-79"><a href="#annotated-cell-6-79"></a></span>
<span id="annotated-cell-6-80"><a href="#annotated-cell-6-80"></a>    <span class="co"># Plot conditional distribution</span></span>
<span id="annotated-cell-6-81"><a href="#annotated-cell-6-81"></a>    <span class="fu">lines!</span>(ax_cond, y_range, <span class="fu">pdf</span>.(conditional_dist, y_range),</span>
<span id="annotated-cell-6-82"><a href="#annotated-cell-6-82"></a>        color <span class="op">=</span> <span class="op">:</span>red, linewidth <span class="op">=</span> <span class="fl">3</span>, label <span class="op">=</span> L<span class="st">"</span><span class="sc">$</span>p<span class="st">(y | X = %</span><span class="sc">$</span>(x_condition)<span class="st">)</span><span class="sc">$</span><span class="st">")</span></span>
<span id="annotated-cell-6-83"><a href="#annotated-cell-6-83"></a></span>
<span id="annotated-cell-6-84"><a href="#annotated-cell-6-84"></a><span class="st">    # </span>Highlight<span class="st"> samples near the conditioning value</span></span>
<span id="annotated-cell-6-85"><a href="#annotated-cell-6-85"></a>    tolerance <span class="op">=</span> <span class="fl">0.2</span></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-6" data-target-annotation="10">10</button><span id="annotated-cell-6-86" class="code-annotation-target"><a href="#annotated-cell-6-86"></a>    near_condition <span class="op">=</span> <span class="fu">abs</span>.(x_samples <span class="op">.-</span> x_condition) <span class="op">.&lt;</span> tolerance</span>
<span id="annotated-cell-6-87"><a href="#annotated-cell-6-87"></a>    y_near <span class="op">=</span> y_samples[near_condition]</span>
<span id="annotated-cell-6-88"><a href="#annotated-cell-6-88"></a></span>
<span id="annotated-cell-6-89"><a href="#annotated-cell-6-89"></a>    <span class="fu">hist!</span>(ax_cond, y_near, bins <span class="op">=</span> <span class="fl">15</span>, normalization <span class="op">=</span> <span class="op">:</span>pdf,</span>
<span id="annotated-cell-6-90"><a href="#annotated-cell-6-90"></a>        color <span class="op">=</span> (<span class="op">:</span>red, <span class="fl">0.4</span>), label <span class="op">=</span> L<span class="st">"</span><span class="sc">\t</span><span class="st">ext{Samples of </span><span class="sc">$</span>y<span class="st">$ given </span><span class="sc">$</span>X<span class="st"> </span><span class="sc">\a</span><span class="st">pprox %</span><span class="sc">$</span>(x_condition)<span class="sc">$</span><span class="st">}")</span></span>
<span id="annotated-cell-6-91"><a href="#annotated-cell-6-91"></a></span>
<span id="annotated-cell-6-92"><a href="#annotated-cell-6-92"></a><span class="st">    # </span>Link<span class="st"> axes appropriately</span></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-6" data-target-annotation="11">11</button><span id="annotated-cell-6-93" class="code-annotation-target"><a href="#annotated-cell-6-93"></a>    <span class="fu">linkxaxes!</span>(ax_main, ax_top)</span>
<span id="annotated-cell-6-94"><a href="#annotated-cell-6-94"></a>    <span class="fu">linkyaxes!</span>(ax_main, ax_right)</span>
<span id="annotated-cell-6-95"><a href="#annotated-cell-6-95"></a></span>
<span id="annotated-cell-6-96"><a href="#annotated-cell-6-96"></a>    <span class="co"># Hide x-axis labels for top plot</span></span>
<span id="annotated-cell-6-97"><a href="#annotated-cell-6-97"></a>    <span class="fu">hidexdecorations!</span>(ax_top, grid <span class="op">=</span> <span class="cn">false</span>)</span>
<span id="annotated-cell-6-98"><a href="#annotated-cell-6-98"></a>    <span class="co"># Hide y-axis labels for right plot</span></span>
<span id="annotated-cell-6-99"><a href="#annotated-cell-6-99"></a>    <span class="fu">hideydecorations!</span>(ax_right, grid <span class="op">=</span> <span class="cn">false</span>)</span>
<span id="annotated-cell-6-100"><a href="#annotated-cell-6-100"></a></span>
<span id="annotated-cell-6-101"><a href="#annotated-cell-6-101"></a>    <span class="co"># Add legends</span></span>
<span id="annotated-cell-6-102"><a href="#annotated-cell-6-102"></a>    <span class="fu">axislegend</span>(ax_main, position <span class="op">=</span> <span class="op">:</span>rt)</span>
<span id="annotated-cell-6-103"><a href="#annotated-cell-6-103"></a>    <span class="fu">axislegend</span>(ax_cond, position <span class="op">=</span> <span class="op">:</span>rt)</span>
<span id="annotated-cell-6-104"><a href="#annotated-cell-6-104"></a></span>
<span id="annotated-cell-6-105"><a href="#annotated-cell-6-105"></a>    <span class="cf">return</span> fig</span>
<span id="annotated-cell-6-106"><a href="#annotated-cell-6-106"></a><span class="kw">end</span></span>
<span id="annotated-cell-6-107"><a href="#annotated-cell-6-107"></a></span>
<span id="annotated-cell-6-108"><a href="#annotated-cell-6-108"></a>fig_joint <span class="op">=</span> <span class="fu">plot_joint_marginal_conditional</span>()</span>
<span id="annotated-cell-6-109"><a href="#annotated-cell-6-109"></a>fig_joint</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-annotation">
<dl class="code-annotation-container-hidden code-annotation-container-grid">
<dt data-target-cell="annotated-cell-6" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-6" data-code-lines="8" data-code-annotation="1">Covariance matrix for bivariate normal with correlation <span class="math inline">\(\rho\)</span></span>
</dd>
<dt data-target-cell="annotated-cell-6" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-6" data-code-lines="9" data-code-annotation="2">Create multivariate normal distribution using <code>MvNormal</code></span>
</dd>
<dt data-target-cell="annotated-cell-6" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-6" data-code-lines="14" data-code-annotation="3">Generate random samples from the joint distribution</span>
</dd>
<dt data-target-cell="annotated-cell-6" data-target-annotation="4">4</dt>
<dd>
<span data-code-cell="annotated-cell-6" data-code-lines="29" data-code-annotation="4">Scatter plot shows the joint distribution through sample points</span>
</dd>
<dt data-target-cell="annotated-cell-6" data-target-annotation="5">5</dt>
<dd>
<span data-code-cell="annotated-cell-6" data-code-lines="42" data-code-annotation="5">Marginal histogram for <span class="math inline">\(X\)</span> (top panel) shows distribution regardless of <span class="math inline">\(Y\)</span></span>
</dd>
<dt data-target-cell="annotated-cell-6" data-target-annotation="6">6</dt>
<dd>
<span data-code-cell="annotated-cell-6" data-code-lines="46" data-code-annotation="6">True marginal distribution is <span class="math inline">\(X \sim N(\mu_1, \sigma_1^2)\)</span></span>
</dd>
<dt data-target-cell="annotated-cell-6" data-target-annotation="7">7</dt>
<dd>
<span data-code-cell="annotated-cell-6" data-code-lines="60" data-code-annotation="7">Rotated histogram shows marginal distribution of <span class="math inline">\(Y\)</span></span>
</dd>
<dt data-target-cell="annotated-cell-6" data-target-annotation="8">8</dt>
<dd>
<span data-code-cell="annotated-cell-6" data-code-lines="76" data-code-annotation="8">Conditional mean: <span class="math inline">\(\mathbb{E}[Y \mid X] = \mu_2 + \rho(\sigma_2/\sigma_1)(x - \mu_1)\)</span></span>
</dd>
<dt data-target-cell="annotated-cell-6" data-target-annotation="9">9</dt>
<dd>
<span data-code-cell="annotated-cell-6" data-code-lines="77" data-code-annotation="9">Conditional variance: <span class="math inline">\(\text{Var}(Y \mid X) = \sigma_2^2(1 - \rho^2)\)</span></span>
</dd>
<dt data-target-cell="annotated-cell-6" data-target-annotation="10">10</dt>
<dd>
<span data-code-cell="annotated-cell-6" data-code-lines="86" data-code-annotation="10">Filter samples close to the conditioning value <span class="math inline">\(X = 2.5\)</span></span>
</dd>
<dt data-target-cell="annotated-cell-6" data-target-annotation="11">11</dt>
<dd>
<span data-code-cell="annotated-cell-6" data-code-lines="93" data-code-annotation="11">Link axes so zooming works across related panels</span>
</dd>
</dl>
</div>
<div class="cell-output cell-output-display" data-execution_count="1">
<div>
<figure class="figure">
<p><img src="probability-stats-examples_files/figure-html/joint-marginal-conditional-example-output-1.svg" id="joint-marginal-conditional-example" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="coin-flip-examples" class="level2">
<h2 class="anchored" data-anchor-id="coin-flip-examples">Coin flip examples</h2>
<p>Having explored the foundational concepts of probability distributions, we now turn to statistical inference—the process of learning about unknown parameters from observed data.</p>
<p>The coin flip example provides a pedagogically clean introduction to statistical inference. We observe a series of coin flips and want to infer the probability <span class="math inline">\(\theta\)</span> of heads. This simple setup allows us to demonstrate maximum likelihood estimation, Bayesian inference, and the relationship between analytical and computational approaches.</p>
<section id="likelihood-visualization" class="level3">
<h3 class="anchored" data-anchor-id="likelihood-visualization">Likelihood visualization</h3>
<p>The likelihood function shows how plausible different values of <span class="math inline">\(\theta\)</span> are given our observed data. For the <code>Binomial</code> model, this creates a clear peak at the observed proportion of heads.</p>
<div id="cell-coin-flip-likelihood" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="annotated-cell-7"><pre class="sourceCode numberSource julia code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode julia"><span id="annotated-cell-7-1"><a href="#annotated-cell-7-1"></a><span class="kw">function</span> <span class="fu">plot_binomial_likelihood</span>(y, n)</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-7" data-target-annotation="1">1</button><span id="annotated-cell-7-2" class="code-annotation-target"><a href="#annotated-cell-7-2"></a>    θ_grid <span class="op">=</span> <span class="fl">0.01</span><span class="op">:</span><span class="fl">0.01</span><span class="op">:</span><span class="fl">0.99</span></span>
<span id="annotated-cell-7-3"><a href="#annotated-cell-7-3"></a></span>
<span id="annotated-cell-7-4"><a href="#annotated-cell-7-4"></a>    <span class="co"># Using Distributions.jl</span></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-7" data-target-annotation="2">2</button><span id="annotated-cell-7-5" class="code-annotation-target"><a href="#annotated-cell-7-5"></a>    likelihood_vals <span class="op">=</span> [<span class="fu">pdf</span>(<span class="fu">Binomial</span>(n, θ), y) for θ <span class="kw">in</span> θ_grid]</span>
<span id="annotated-cell-7-6"><a href="#annotated-cell-7-6"></a></span>
<span id="annotated-cell-7-7"><a href="#annotated-cell-7-7"></a>    fig <span class="op">=</span> <span class="fu">Figure</span>(size <span class="op">=</span> (<span class="fl">800</span>, <span class="fl">500</span>))</span>
<span id="annotated-cell-7-8"><a href="#annotated-cell-7-8"></a>    ax <span class="op">=</span> <span class="fu">Axis</span>(fig[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="annotated-cell-7-9"><a href="#annotated-cell-7-9"></a>        xlabel <span class="op">=</span> L<span class="st">"</span><span class="sc">$</span><span class="st">\</span>theta<span class="st">$ (probability of heads)"</span>,</span>
<span id="annotated-cell-7-10"><a href="#annotated-cell-7-10"></a>        ylabel <span class="op">=</span> <span class="st">"Likelihood"</span>,</span>
<span id="annotated-cell-7-11"><a href="#annotated-cell-7-11"></a>        title <span class="op">=</span> <span class="st">"Likelihood for </span><span class="sc">$</span>(y)<span class="st"> heads in </span><span class="sc">$</span>(n)<span class="st"> flips"</span>)</span>
<span id="annotated-cell-7-12"><a href="#annotated-cell-7-12"></a></span>
<span id="annotated-cell-7-13"><a href="#annotated-cell-7-13"></a>    <span class="fu">lines!</span>(ax, θ_grid, likelihood_vals, label <span class="op">=</span> L<span class="st">"</span><span class="sc">$</span>P<span class="st">(y \mid </span><span class="sc">\t</span><span class="st">heta)</span><span class="sc">$</span><span class="st">", </span>linewidth<span class="st"> = 2, color = :blue)</span></span>
<span id="annotated-cell-7-14"><a href="#annotated-cell-7-14"></a></span>
<span id="annotated-cell-7-15"><a href="#annotated-cell-7-15"></a>    <span class="co"># Mark the MLE</span></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-7" data-target-annotation="3">3</button><span id="annotated-cell-7-16" class="code-annotation-target"><a href="#annotated-cell-7-16"></a>    mle <span class="op">=</span> y <span class="op">/</span> n</span>
<span id="annotated-cell-7-17"><a href="#annotated-cell-7-17"></a>    <span class="fu">vlines!</span>(ax, [mle], color <span class="op">=</span> <span class="op">:</span>black, linestyle <span class="op">=</span> <span class="op">:</span>dot, linewidth <span class="op">=</span> <span class="fl">2</span>, label <span class="op">=</span> <span class="st">"MLE = </span><span class="sc">$</span>(<span class="fu">round</span>(mle, digits<span class="op">=</span><span class="fl">2</span>))<span class="st">"</span>)</span>
<span id="annotated-cell-7-18"><a href="#annotated-cell-7-18"></a></span>
<span id="annotated-cell-7-19"><a href="#annotated-cell-7-19"></a>    <span class="fu">axislegend</span>(ax, position <span class="op">=</span> <span class="op">:</span>lt)</span>
<span id="annotated-cell-7-20"><a href="#annotated-cell-7-20"></a>    <span class="cf">return</span> fig</span>
<span id="annotated-cell-7-21"><a href="#annotated-cell-7-21"></a><span class="kw">end</span></span>
<span id="annotated-cell-7-22"><a href="#annotated-cell-7-22"></a></span>
<span id="annotated-cell-7-23"><a href="#annotated-cell-7-23"></a><span class="co"># Example with 7 heads in 10 flips</span></span>
<span id="annotated-cell-7-24"><a href="#annotated-cell-7-24"></a>fig_likelihood <span class="op">=</span> <span class="fu">plot_binomial_likelihood</span>(<span class="fl">7</span>, <span class="fl">10</span>)</span>
<span id="annotated-cell-7-25"><a href="#annotated-cell-7-25"></a>fig_likelihood</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-annotation">
<dl class="code-annotation-container-hidden code-annotation-container-grid">
<dt data-target-cell="annotated-cell-7" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-7" data-code-lines="2" data-code-annotation="1">Create grid of <span class="math inline">\(\theta\)</span> values from 0.01 to 0.99 to evaluate likelihood</span>
</dd>
<dt data-target-cell="annotated-cell-7" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-7" data-code-lines="5" data-code-annotation="2">Evaluate binomial likelihood <span class="math inline">\(L(\theta) = P(y|\theta,n)\)</span> for each <span class="math inline">\(\theta\)</span> value</span>
</dd>
<dt data-target-cell="annotated-cell-7" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-7" data-code-lines="16" data-code-annotation="3">Maximum likelihood estimate is simply the observed proportion <span class="math inline">\(\frac{y}{n}\)</span></span>
</dd>
</dl>
</div>
<div class="cell-output cell-output-display" data-execution_count="1">
<div>
<figure class="figure">
<p><img src="probability-stats-examples_files/figure-html/coin-flip-likelihood-output-1.svg" id="coin-flip-likelihood" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>The likelihood peaks at <span class="math inline">\(\theta = 0.7\)</span> when we observe 7 heads out of 10 flips. This demonstrates that the maximum likelihood estimate is simply the observed proportion.</p>
</section>
<section id="maximum-likelihood-estimation" class="level3">
<h3 class="anchored" data-anchor-id="maximum-likelihood-estimation">Maximum likelihood estimation</h3>
<p>Maximum likelihood estimation finds the parameter value that makes the observed data most probable. For the coin flip problem, we can solve this analytically by taking the derivative of the log-likelihood and setting it to zero. The following shows how the <code>MLE</code> varies with different numbers of observed heads:</p>
<div id="cell-coin-flip-mle" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="annotated-cell-8"><pre class="sourceCode numberSource julia code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode julia"><span id="annotated-cell-8-1"><a href="#annotated-cell-8-1"></a><span class="kw">function</span> <span class="fu">demonstrate_mle_estimation</span>(y_values, n)</span>
<span id="annotated-cell-8-2"><a href="#annotated-cell-8-2"></a>    θ_grid <span class="op">=</span> <span class="fl">0.01</span><span class="op">:</span><span class="fl">0.01</span><span class="op">:</span><span class="fl">0.99</span></span>
<span id="annotated-cell-8-3"><a href="#annotated-cell-8-3"></a></span>
<span id="annotated-cell-8-4"><a href="#annotated-cell-8-4"></a>    fig <span class="op">=</span> <span class="fu">Figure</span>(size <span class="op">=</span> (<span class="fl">800</span>, <span class="fl">600</span>))</span>
<span id="annotated-cell-8-5"><a href="#annotated-cell-8-5"></a></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-8" data-target-annotation="1">1</button><span id="annotated-cell-8-6" class="code-annotation-target"><a href="#annotated-cell-8-6"></a>    <span class="cf">for</span> (i, y) <span class="kw">in</span> <span class="fu">enumerate</span>(y_values)</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-8" data-target-annotation="2">2</button><span id="annotated-cell-8-7" class="code-annotation-target"><a href="#annotated-cell-8-7"></a>        ax <span class="op">=</span> <span class="fu">Axis</span>(fig[<span class="fu">div</span>(i <span class="op">-</span> <span class="fl">1</span>, <span class="fl">2</span>)<span class="op">+</span><span class="fl">1</span>, ((i<span class="op">-</span><span class="fl">1</span>)<span class="op">%</span><span class="fl">2</span>)<span class="op">+</span><span class="fl">1</span>],</span>
<span id="annotated-cell-8-8"><a href="#annotated-cell-8-8"></a>            xlabel <span class="op">=</span> L<span class="st">"</span><span class="sc">\t</span><span class="st">heta"</span>,</span>
<span id="annotated-cell-8-9"><a href="#annotated-cell-8-9"></a>            ylabel <span class="op">=</span> <span class="st">"Log-likelihood"</span>,</span>
<span id="annotated-cell-8-10"><a href="#annotated-cell-8-10"></a>            title <span class="op">=</span> <span class="st">"</span><span class="sc">$</span>(y)<span class="st">/</span><span class="sc">$</span>(n)<span class="st"> heads"</span>)</span>
<span id="annotated-cell-8-11"><a href="#annotated-cell-8-11"></a></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-8" data-target-annotation="3">3</button><span id="annotated-cell-8-12" class="code-annotation-target"><a href="#annotated-cell-8-12"></a>        log_likelihood <span class="op">=</span> [y <span class="op">*</span> <span class="fu">log</span>(θ) <span class="op">+</span> (n <span class="op">-</span> y) <span class="op">*</span> <span class="fu">log</span>(<span class="fl">1</span> <span class="op">-</span> θ) for θ <span class="kw">in</span> θ_grid]</span>
<span id="annotated-cell-8-13"><a href="#annotated-cell-8-13"></a></span>
<span id="annotated-cell-8-14"><a href="#annotated-cell-8-14"></a>        <span class="fu">lines!</span>(ax, θ_grid, log_likelihood, linewidth <span class="op">=</span> <span class="fl">2</span>, color <span class="op">=</span> <span class="op">:</span>blue)</span>
<span id="annotated-cell-8-15"><a href="#annotated-cell-8-15"></a></span>
<span id="annotated-cell-8-16"><a href="#annotated-cell-8-16"></a>        mle <span class="op">=</span> y <span class="op">/</span> n</span>
<span id="annotated-cell-8-17"><a href="#annotated-cell-8-17"></a>        <span class="fu">vlines!</span>(ax, [mle], color <span class="op">=</span> <span class="op">:</span>red, linestyle <span class="op">=</span> <span class="op">:</span>dash, linewidth <span class="op">=</span> <span class="fl">2</span>,</span>
<span id="annotated-cell-8-18"><a href="#annotated-cell-8-18"></a>            label <span class="op">=</span> <span class="st">"MLE"</span>)</span>
<span id="annotated-cell-8-19"><a href="#annotated-cell-8-19"></a></span>
<span id="annotated-cell-8-20"><a href="#annotated-cell-8-20"></a>        <span class="cf">if</span> i <span class="op">==</span> <span class="fl">1</span></span>
<span id="annotated-cell-8-21"><a href="#annotated-cell-8-21"></a>            <span class="fu">axislegend</span>(ax, position <span class="op">=</span> <span class="op">:</span>cb)</span>
<span id="annotated-cell-8-22"><a href="#annotated-cell-8-22"></a>        <span class="cf">end</span></span>
<span id="annotated-cell-8-23"><a href="#annotated-cell-8-23"></a>    <span class="cf">end</span></span>
<span id="annotated-cell-8-24"><a href="#annotated-cell-8-24"></a></span>
<span id="annotated-cell-8-25"><a href="#annotated-cell-8-25"></a>    <span class="cf">return</span> fig</span>
<span id="annotated-cell-8-26"><a href="#annotated-cell-8-26"></a><span class="kw">end</span></span>
<span id="annotated-cell-8-27"><a href="#annotated-cell-8-27"></a></span>
<span id="annotated-cell-8-28"><a href="#annotated-cell-8-28"></a><span class="co"># Show MLE for different outcomes</span></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-8" data-target-annotation="4">4</button><span id="annotated-cell-8-29" class="code-annotation-target"><a href="#annotated-cell-8-29"></a>fig_mle <span class="op">=</span> <span class="fu">demonstrate_mle_estimation</span>([<span class="fl">3</span>, <span class="fl">5</span>, <span class="fl">7</span>, <span class="fl">9</span>], <span class="fl">10</span>)</span>
<span id="annotated-cell-8-30"><a href="#annotated-cell-8-30"></a>fig_mle</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-annotation">
<dl class="code-annotation-container-hidden code-annotation-container-grid">
<dt data-target-cell="annotated-cell-8" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-8" data-code-lines="6" data-code-annotation="1">Loop through different numbers of observed heads to show MLE behavior</span>
</dd>
<dt data-target-cell="annotated-cell-8" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-8" data-code-lines="7" data-code-annotation="2">Create <span class="math inline">\(2 \times 2\)</span> grid of subplots using integer division and modulo</span>
</dd>
<dt data-target-cell="annotated-cell-8" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-8" data-code-lines="12" data-code-annotation="3">Analytical log-likelihood: <span class="math inline">\(\ell(\theta) = y \ln(\theta) + (n-y) \ln(1-\theta)\)</span></span>
</dd>
<dt data-target-cell="annotated-cell-8" data-target-annotation="4">4</dt>
<dd>
<span data-code-cell="annotated-cell-8" data-code-lines="29" data-code-annotation="4">Demonstrate <code>MLE</code> for 3, 5, 7, and 9 heads out of 10 flips</span>
</dd>
</dl>
</div>
<div class="cell-output cell-output-display" data-execution_count="1">
<div>
<figure class="figure">
<p><img src="probability-stats-examples_files/figure-html/coin-flip-mle-output-1.svg" id="coin-flip-mle" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Each plot shows the log-likelihood function as a smooth curve with its maximum clearly indicated. Notice that the <code>MLE</code> is always at the observed proportion <span class="math inline">\(\frac{y}{n}\)</span>, confirming our analytical derivation.</p>
</section>
<section id="bayesian-analysis" class="level3">
<h3 class="anchored" data-anchor-id="bayesian-analysis">Bayesian analysis</h3>
<p>Bayesian inference treats the parameter <span class="math inline">\(\theta\)</span> as a random variable with its own probability distribution. We start with a prior belief about <span class="math inline">\(\theta\)</span>, observe data, and update our beliefs to get the posterior distribution.</p>
<p>For the coin flip problem with a <code>Beta</code> prior, we can compute the posterior analytically using conjugacy. The posterior parameters are simply the prior parameters plus the observed successes and failures. This example compares the exact analytical solution with <code>MCMC</code> sampling.</p>
<p>First we define the <code>Turing.jl</code> model and the comparison function. See the official <a href="https://turinglang.org/"><code>Turing.jl</code> documentation and tutorials</a>.</p>
<div id="18" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="annotated-cell-9"><pre class="sourceCode numberSource julia code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode julia"><span id="annotated-cell-9-1"><a href="#annotated-cell-9-1"></a><span class="co"># Turing model for coin flipping</span></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-9" data-target-annotation="1">1</button><span id="annotated-cell-9-2" class="code-annotation-target"><a href="#annotated-cell-9-2"></a><span class="pp">@model</span> <span class="kw">function</span> <span class="fu">coin_flip_model</span>(y, n, prior_α, prior_β)</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-9" data-target-annotation="2">2</button><span id="annotated-cell-9-3" class="code-annotation-target"><a href="#annotated-cell-9-3"></a>    θ <span class="op">~</span> <span class="fu">Beta</span>(prior_α, prior_β)</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-9" data-target-annotation="3">3</button><span id="annotated-cell-9-4" class="code-annotation-target"><a href="#annotated-cell-9-4"></a>    y <span class="op">~</span> <span class="fu">Binomial</span>(n, θ)</span>
<span id="annotated-cell-9-5"><a href="#annotated-cell-9-5"></a>    <span class="cf">return</span> θ</span>
<span id="annotated-cell-9-6"><a href="#annotated-cell-9-6"></a><span class="kw">end</span></span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-annotation">
<dl class="code-annotation-container-hidden code-annotation-container-grid">
<dt data-target-cell="annotated-cell-9" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-9" data-code-lines="2" data-code-annotation="1"><code>Turing.jl</code> probabilistic programming model specification using <code>@model</code> macro</span>
</dd>
<dt data-target-cell="annotated-cell-9" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-9" data-code-lines="3" data-code-annotation="2">Prior distribution: <span class="math inline">\(\theta \sim \text{Beta}(\alpha, \beta)\)</span></span>
</dd>
<dt data-target-cell="annotated-cell-9" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-9" data-code-lines="4" data-code-annotation="3">Likelihood: <span class="math inline">\(y \sim \text{Binomial}(n, \theta)\)</span> connects data to parameter</span>
</dd>
</dl>
</div>
<div class="cell-output cell-output-display" data-execution_count="1">
<pre><code>coin_flip_model (generic function with 2 methods)</code></pre>
</div>
</div>
<p>Next we define a function to help us create the data and fits</p>
<div id="20" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="annotated-cell-10"><pre class="sourceCode numberSource julia code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode julia"><button class="code-annotation-anchor" data-target-cell="annotated-cell-10" data-target-annotation="4">4</button><span id="annotated-cell-10-1" class="code-annotation-target"><a href="#annotated-cell-10-1"></a><span class="kw">function</span> <span class="fu">compare_analytical_mcmc</span>(y, n; prior_α <span class="op">=</span> <span class="fl">1</span>, prior_β <span class="op">=</span> <span class="fl">1</span>, n_samples <span class="op">=</span> <span class="fl">5000</span>)</span>
<span id="annotated-cell-10-2"><a href="#annotated-cell-10-2"></a>    <span class="co"># Analytical posterior</span></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-10" data-target-annotation="5">5</button><span id="annotated-cell-10-3" class="code-annotation-target"><a href="#annotated-cell-10-3"></a>    post_α <span class="op">=</span> prior_α <span class="op">+</span> y</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-10" data-target-annotation="6">6</button><span id="annotated-cell-10-4" class="code-annotation-target"><a href="#annotated-cell-10-4"></a>    post_β <span class="op">=</span> prior_β <span class="op">+</span> n <span class="op">-</span> y</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-10" data-target-annotation="7">7</button><span id="annotated-cell-10-5" class="code-annotation-target"><a href="#annotated-cell-10-5"></a>    analytical_posterior <span class="op">=</span> <span class="fu">Beta</span>(post_α, post_β)</span>
<span id="annotated-cell-10-6"><a href="#annotated-cell-10-6"></a></span>
<span id="annotated-cell-10-7"><a href="#annotated-cell-10-7"></a>    <span class="co"># MCMC sampling</span></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-10" data-target-annotation="8">8</button><span id="annotated-cell-10-8" class="code-annotation-target"><a href="#annotated-cell-10-8"></a>    model <span class="op">=</span> <span class="fu">coin_flip_model</span>(y, n, prior_α, prior_β)</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-10" data-target-annotation="9">9</button><span id="annotated-cell-10-9" class="code-annotation-target"><a href="#annotated-cell-10-9"></a>    chain <span class="op">=</span> <span class="fu">sample</span>(model, <span class="fu">NUTS</span>(), n_samples, verbose <span class="op">=</span> <span class="cn">false</span>, progress <span class="op">=</span> <span class="cn">false</span>)</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-10" data-target-annotation="10">10</button><span id="annotated-cell-10-10" class="code-annotation-target"><a href="#annotated-cell-10-10"></a>    mcmc_samples <span class="op">=</span> <span class="fu">Array</span>(chain[<span class="op">:</span>θ])</span>
<span id="annotated-cell-10-11"><a href="#annotated-cell-10-11"></a></span>
<span id="annotated-cell-10-12"><a href="#annotated-cell-10-12"></a>    <span class="co"># Return data for plotting instead of creating plots</span></span>
<span id="annotated-cell-10-13"><a href="#annotated-cell-10-13"></a>    <span class="cf">return</span> (analytical_posterior <span class="op">=</span> analytical_posterior, mcmc_samples <span class="op">=</span> mcmc_samples, n_samples <span class="op">=</span> n_samples)</span>
<span id="annotated-cell-10-14"><a href="#annotated-cell-10-14"></a><span class="kw">end</span></span>
<span id="annotated-cell-10-15"><a href="#annotated-cell-10-15"></a></span>
<span id="annotated-cell-10-16"><a href="#annotated-cell-10-16"></a><span class="co"># Generate data by calling the same function twice  </span></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-10" data-target-annotation="11">11</button><span id="annotated-cell-10-17" class="code-annotation-target"><a href="#annotated-cell-10-17"></a>data_low <span class="op">=</span> <span class="fu">compare_analytical_mcmc</span>(<span class="fl">7</span>, <span class="fl">10</span>, n_samples <span class="op">=</span> <span class="fl">500</span>)</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-10" data-target-annotation="12">12</button><span id="annotated-cell-10-18" class="code-annotation-target"><a href="#annotated-cell-10-18"></a>data_high <span class="op">=</span> <span class="fu">compare_analytical_mcmc</span>(<span class="fl">7</span>, <span class="fl">10</span>, n_samples <span class="op">=</span> <span class="fl">50_000</span>)</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-annotation">
<dl class="code-annotation-container-hidden code-annotation-container-grid">
<dt data-target-cell="annotated-cell-10" data-target-annotation="4">4</dt>
<dd>
<span data-code-cell="annotated-cell-10" data-code-lines="1" data-code-annotation="4">Function comparing analytical vs MCMC approaches</span>
</dd>
<dt data-target-cell="annotated-cell-10" data-target-annotation="5">5</dt>
<dd>
<span data-code-cell="annotated-cell-10" data-code-lines="3" data-code-annotation="5">Posterior <span class="math inline">\(\alpha = \text{prior } \alpha + \text{successes}\)</span> (conjugate update)</span>
</dd>
<dt data-target-cell="annotated-cell-10" data-target-annotation="6">6</dt>
<dd>
<span data-code-cell="annotated-cell-10" data-code-lines="4" data-code-annotation="6">Posterior <span class="math inline">\(\beta = \text{prior } \beta + \text{failures}\)</span> (conjugate update)</span>
</dd>
<dt data-target-cell="annotated-cell-10" data-target-annotation="7">7</dt>
<dd>
<span data-code-cell="annotated-cell-10" data-code-lines="5" data-code-annotation="7">Exact posterior using <code>Beta</code>-Binomial conjugacy</span>
</dd>
<dt data-target-cell="annotated-cell-10" data-target-annotation="8">8</dt>
<dd>
<span data-code-cell="annotated-cell-10" data-code-lines="8" data-code-annotation="8">Create Turing model instance with data and priors</span>
</dd>
<dt data-target-cell="annotated-cell-10" data-target-annotation="9">9</dt>
<dd>
<span data-code-cell="annotated-cell-10" data-code-lines="9" data-code-annotation="9">MCMC sampling using No-U-Turn Sampler (<code>NUTS</code>)</span>
</dd>
<dt data-target-cell="annotated-cell-10" data-target-annotation="10">10</dt>
<dd>
<span data-code-cell="annotated-cell-10" data-code-lines="10" data-code-annotation="10">Extract <span class="math inline">\(\theta\)</span> samples from MCMC chain using <code>Array(chain[:θ])</code></span>
</dd>
<dt data-target-cell="annotated-cell-10" data-target-annotation="11">11</dt>
<dd>
<span data-code-cell="annotated-cell-10" data-code-lines="17" data-code-annotation="11">Low sample count (500) to show MCMC variability</span>
</dd>
<dt data-target-cell="annotated-cell-10" data-target-annotation="12">12</dt>
<dd>
<span data-code-cell="annotated-cell-10" data-code-lines="18" data-code-annotation="12">High sample count (50,000) to show MCMC convergence</span>
</dd>
</dl>
</div>
<div class="cell-output cell-output-stdout">
<div class="ansi-escaped-output">
<pre><span class="ansi-cyan-fg ansi-bold">┌ </span><span class="ansi-cyan-fg ansi-bold">Info: </span>Found initial step size
<span class="ansi-cyan-fg ansi-bold">└ </span>  ϵ = 1.6
<span class="ansi-cyan-fg ansi-bold">┌ </span><span class="ansi-cyan-fg ansi-bold">Info: </span>Found initial step size
<span class="ansi-cyan-fg ansi-bold">└ </span>  ϵ = 1.6
</pre>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="1">
<pre><code>(analytical_posterior = Distributions.Beta{Float64}(α=8.0, β=4.0), mcmc_samples = [0.8214313405535059; 0.646445076138614; … ; 0.5588684710867051; 0.6492558369061647;;], n_samples = 50000)</code></pre>
</div>
</div>
<p>And then we plot</p>
<div id="cell-coin-flip-bayesian" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="annotated-cell-11"><pre class="sourceCode numberSource julia code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode julia"><span id="annotated-cell-11-1"><a href="#annotated-cell-11-1"></a><span class="co"># Combine plots</span></span>
<span id="annotated-cell-11-2"><a href="#annotated-cell-11-2"></a>fig_comparison <span class="op">=</span> <span class="fu">Figure</span>(size <span class="op">=</span> (<span class="fl">1200</span>, <span class="fl">500</span>))</span>
<span id="annotated-cell-11-3"><a href="#annotated-cell-11-3"></a>ax1 <span class="op">=</span> <span class="fu">Axis</span>(fig_comparison[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="annotated-cell-11-4"><a href="#annotated-cell-11-4"></a>    xlabel <span class="op">=</span> L<span class="st">"</span><span class="sc">$</span><span class="st">\</span>theta<span class="st">$ (probability of heads)"</span>,</span>
<span id="annotated-cell-11-5"><a href="#annotated-cell-11-5"></a>    ylabel <span class="op">=</span> <span class="st">"Density"</span>,</span>
<span id="annotated-cell-11-6"><a href="#annotated-cell-11-6"></a>    title <span class="op">=</span> <span class="st">"MCMC with 500 samples"</span>)</span>
<span id="annotated-cell-11-7"><a href="#annotated-cell-11-7"></a>ax2 <span class="op">=</span> <span class="fu">Axis</span>(fig_comparison[<span class="fl">1</span>, <span class="fl">2</span>],</span>
<span id="annotated-cell-11-8"><a href="#annotated-cell-11-8"></a>    xlabel <span class="op">=</span> L<span class="st">"</span><span class="sc">$</span><span class="st">\</span>theta<span class="st">$ (probability of heads)"</span>,</span>
<span id="annotated-cell-11-9"><a href="#annotated-cell-11-9"></a>    title <span class="op">=</span> <span class="st">"MCMC with 50,000 samples"</span>)</span>
<span id="annotated-cell-11-10"><a href="#annotated-cell-11-10"></a></span>
<span id="annotated-cell-11-11"><a href="#annotated-cell-11-11"></a><span class="co"># Link y-axes for comparison</span></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-11" data-target-annotation="13">13</button><span id="annotated-cell-11-12" class="code-annotation-target"><a href="#annotated-cell-11-12"></a><span class="fu">linkaxes!</span>(ax1, ax2)</span>
<span id="annotated-cell-11-13"><a href="#annotated-cell-11-13"></a></span>
<span id="annotated-cell-11-14"><a href="#annotated-cell-11-14"></a><span class="co"># Plot using the same plotting logic for both panels</span></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-11" data-target-annotation="14">14</button><span id="annotated-cell-11-15" class="code-annotation-target"><a href="#annotated-cell-11-15"></a><span class="cf">for</span> (ax, data) <span class="kw">in</span> [(ax1, data_low), (ax2, data_high)]</span>
<span id="annotated-cell-11-16"><a href="#annotated-cell-11-16"></a>    <span class="co"># Plot analytical posterior</span></span>
<span id="annotated-cell-11-17"><a href="#annotated-cell-11-17"></a>    θ_grid <span class="op">=</span> <span class="fl">0.01</span><span class="op">:</span><span class="fl">0.01</span><span class="op">:</span><span class="fl">0.99</span></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-11" data-target-annotation="15">15</button><span id="annotated-cell-11-18" class="code-annotation-target"><a href="#annotated-cell-11-18"></a>    analytical_density <span class="op">=</span> <span class="fu">pdf</span>.(data.analytical_posterior, θ_grid)</span>
<span id="annotated-cell-11-19"><a href="#annotated-cell-11-19"></a>    <span class="fu">lines!</span>(ax, θ_grid, analytical_density, label <span class="op">=</span> <span class="st">"Analytical"</span>,</span>
<span id="annotated-cell-11-20"><a href="#annotated-cell-11-20"></a>        linewidth <span class="op">=</span> <span class="fl">3</span>, color <span class="op">=</span> <span class="op">:</span>blue)</span>
<span id="annotated-cell-11-21"><a href="#annotated-cell-11-21"></a></span>
<span id="annotated-cell-11-22"><a href="#annotated-cell-11-22"></a>    <span class="co"># Plot MCMC histogram</span></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-11" data-target-annotation="16">16</button><span id="annotated-cell-11-23" class="code-annotation-target"><a href="#annotated-cell-11-23"></a>    <span class="fu">hist!</span>(ax, <span class="fu">vec</span>(data.mcmc_samples), bins <span class="op">=</span> <span class="fl">40</span>, normalization <span class="op">=</span> <span class="op">:</span>pdf,</span>
<span id="annotated-cell-11-24"><a href="#annotated-cell-11-24"></a>        color <span class="op">=</span> (<span class="op">:</span>red, <span class="fl">0.6</span>), label <span class="op">=</span> <span class="st">"MCMC samples"</span>)</span>
<span id="annotated-cell-11-25"><a href="#annotated-cell-11-25"></a></span>
<span id="annotated-cell-11-26"><a href="#annotated-cell-11-26"></a>    <span class="co"># Add summary statistics</span></span>
<span id="annotated-cell-11-27"><a href="#annotated-cell-11-27"></a>    analytical_mean <span class="op">=</span> <span class="fu">mean</span>(data.analytical_posterior)</span>
<span id="annotated-cell-11-28"><a href="#annotated-cell-11-28"></a>    mcmc_mean <span class="op">=</span> <span class="fu">mean</span>(data.mcmc_samples)</span>
<span id="annotated-cell-11-29"><a href="#annotated-cell-11-29"></a></span>
<span id="annotated-cell-11-30"><a href="#annotated-cell-11-30"></a>    <span class="co"># Vertical lines for means</span></span>
<span id="annotated-cell-11-31"><a href="#annotated-cell-11-31"></a>    <span class="fu">vlines!</span>(ax, [analytical_mean], color <span class="op">=</span> <span class="op">:</span>blue, linestyle <span class="op">=</span> <span class="op">:</span>dash, alpha <span class="op">=</span> <span class="fl">0.8</span>)</span>
<span id="annotated-cell-11-32"><a href="#annotated-cell-11-32"></a>    <span class="fu">vlines!</span>(ax, [mcmc_mean], color <span class="op">=</span> <span class="op">:</span>red, linestyle <span class="op">=</span> <span class="op">:</span>dot, alpha <span class="op">=</span> <span class="fl">0.8</span>)</span>
<span id="annotated-cell-11-33"><a href="#annotated-cell-11-33"></a></span>
<span id="annotated-cell-11-34"><a href="#annotated-cell-11-34"></a>    <span class="co"># Add vertical line at 0.3 to show tail cutoff</span></span>
<span id="annotated-cell-11-35"><a href="#annotated-cell-11-35"></a>    <span class="fu">vlines!</span>(ax, [<span class="fl">0.3</span>], color <span class="op">=</span> <span class="op">:</span>gray, linestyle <span class="op">=</span> <span class="op">:</span>solid, alpha <span class="op">=</span> <span class="fl">0.6</span>)</span>
<span id="annotated-cell-11-36"><a href="#annotated-cell-11-36"></a></span>
<span id="annotated-cell-11-37"><a href="#annotated-cell-11-37"></a>    <span class="co"># Add text annotations with statistics  </span></span>
<span id="annotated-cell-11-38"><a href="#annotated-cell-11-38"></a>    <span class="fu">text!</span>(ax, <span class="fl">0.35</span>, <span class="fu">maximum</span>(analytical_density) <span class="op">*</span> <span class="fl">0.95</span>,</span>
<span id="annotated-cell-11-39"><a href="#annotated-cell-11-39"></a>        text <span class="op">=</span> L<span class="st">"</span><span class="sc">\t</span><span class="st">ext{Analytical: Mean} = %</span><span class="sc">$</span>(<span class="fu">round</span>(analytical_mean, digits<span class="op">=</span><span class="fl">3</span>))<span class="st">"</span>,</span>
<span id="annotated-cell-11-40"><a href="#annotated-cell-11-40"></a>        fontsize <span class="op">=</span> <span class="fl">10</span>, color <span class="op">=</span> <span class="op">:</span>blue)</span>
<span id="annotated-cell-11-41"><a href="#annotated-cell-11-41"></a>    <span class="fu">text!</span>(ax, <span class="fl">0.35</span>, <span class="fu">maximum</span>(analytical_density) <span class="op">*</span> <span class="fl">0.85</span>,</span>
<span id="annotated-cell-11-42"><a href="#annotated-cell-11-42"></a>        text <span class="op">=</span> L<span class="st">"</span><span class="sc">\t</span><span class="st">ext{MCMC: Mean} = %</span><span class="sc">$</span>(<span class="fu">round</span>(mcmc_mean, digits<span class="op">=</span><span class="fl">3</span>))<span class="st">"</span>,</span>
<span id="annotated-cell-11-43"><a href="#annotated-cell-11-43"></a>        fontsize <span class="op">=</span> <span class="fl">10</span>, color <span class="op">=</span> <span class="op">:</span>red)</span>
<span id="annotated-cell-11-44"><a href="#annotated-cell-11-44"></a></span>
<span id="annotated-cell-11-45"><a href="#annotated-cell-11-45"></a>    <span class="co"># Calculate P(theta &lt; 0.25) - tail probability</span></span>
<span id="annotated-cell-11-46"><a href="#annotated-cell-11-46"></a>    tail <span class="op">=</span> <span class="fl">0.25</span></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-11" data-target-annotation="17">17</button><span id="annotated-cell-11-47" class="code-annotation-target"><a href="#annotated-cell-11-47"></a>    analytical_tail_prob <span class="op">=</span> <span class="fu">cdf</span>(data.analytical_posterior, tail)</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-11" data-target-annotation="18">18</button><span id="annotated-cell-11-48" class="code-annotation-target"><a href="#annotated-cell-11-48"></a>    analytical_tail_prob <span class="op">=</span> <span class="pp">@sprintf</span>(<span class="st">"%.2E"</span>, analytical_tail_prob)</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-11" data-target-annotation="19">19</button><span id="annotated-cell-11-49" class="code-annotation-target"><a href="#annotated-cell-11-49"></a>    mcmc_tail_prob <span class="op">=</span> <span class="fu">mean</span>(data.mcmc_samples <span class="op">.&lt;</span> tail)</span>
<span id="annotated-cell-11-50"><a href="#annotated-cell-11-50"></a>    mcmc_tail_prob <span class="op">=</span> <span class="pp">@sprintf</span>(<span class="st">"%.2E"</span>, mcmc_tail_prob)</span>
<span id="annotated-cell-11-51"><a href="#annotated-cell-11-51"></a>    <span class="fu">text!</span>(ax, <span class="fl">0.05</span>, <span class="fu">maximum</span>(analytical_density) <span class="op">*</span> <span class="fl">0.55</span>,</span>
<span id="annotated-cell-11-52"><a href="#annotated-cell-11-52"></a>        text <span class="op">=</span> L<span class="st">"</span><span class="sc">\t</span><span class="st">ext{Analytical}: </span><span class="sc">$</span>P<span class="st">(</span><span class="sc">\t</span><span class="st">heta &lt; %</span><span class="sc">$</span>(tail)<span class="st">)$ = %</span><span class="sc">$</span>(analytical_tail_prob)<span class="st">"</span>,</span>
<span id="annotated-cell-11-53"><a href="#annotated-cell-11-53"></a>        fontsize <span class="op">=</span> <span class="fl">10</span>, color <span class="op">=</span> <span class="op">:</span>blue)</span>
<span id="annotated-cell-11-54"><a href="#annotated-cell-11-54"></a>    <span class="fu">text!</span>(ax, <span class="fl">0.05</span>, <span class="fu">maximum</span>(analytical_density) <span class="op">*</span> <span class="fl">0.45</span>,</span>
<span id="annotated-cell-11-55"><a href="#annotated-cell-11-55"></a>        text <span class="op">=</span> L<span class="st">"</span><span class="sc">\t</span><span class="st">ext{MCMC}: </span><span class="sc">$</span>P<span class="st">(</span><span class="sc">\t</span><span class="st">heta &lt; %</span><span class="sc">$</span>(tail)<span class="st">)$ = %</span><span class="sc">$</span>(mcmc_tail_prob)<span class="st">"</span>,</span>
<span id="annotated-cell-11-56"><a href="#annotated-cell-11-56"></a>        fontsize <span class="op">=</span> <span class="fl">10</span>, color <span class="op">=</span> <span class="op">:</span>red)</span>
<span id="annotated-cell-11-57"><a href="#annotated-cell-11-57"></a></span>
<span id="annotated-cell-11-58"><a href="#annotated-cell-11-58"></a>    <span class="fu">axislegend</span>(ax, position <span class="op">=</span> <span class="op">:</span>rt)</span>
<span id="annotated-cell-11-59"><a href="#annotated-cell-11-59"></a><span class="cf">end</span></span>
<span id="annotated-cell-11-60"><a href="#annotated-cell-11-60"></a></span>
<span id="annotated-cell-11-61"><a href="#annotated-cell-11-61"></a>fig_comparison</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-annotation">
<dl class="code-annotation-container-hidden code-annotation-container-grid">
<dt data-target-cell="annotated-cell-11" data-target-annotation="13">13</dt>
<dd>
<span data-code-cell="annotated-cell-11" data-code-lines="12" data-code-annotation="13">Link axes for direct comparison between panels</span>
</dd>
<dt data-target-cell="annotated-cell-11" data-target-annotation="14">14</dt>
<dd>
<span data-code-cell="annotated-cell-11" data-code-lines="15" data-code-annotation="14">Loop over both low and high sample datasets</span>
</dd>
<dt data-target-cell="annotated-cell-11" data-target-annotation="15">15</dt>
<dd>
<span data-code-cell="annotated-cell-11" data-code-lines="18" data-code-annotation="15">Exact analytical posterior density</span>
</dd>
<dt data-target-cell="annotated-cell-11" data-target-annotation="16">16</dt>
<dd>
<span data-code-cell="annotated-cell-11" data-code-lines="23" data-code-annotation="16">Normalized histogram of <code>MCMC</code> samples</span>
</dd>
<dt data-target-cell="annotated-cell-11" data-target-annotation="17">17</dt>
<dd>
<span data-code-cell="annotated-cell-11" data-code-lines="47" data-code-annotation="17">Exact tail probability using <code>CDF</code></span>
</dd>
<dt data-target-cell="annotated-cell-11" data-target-annotation="18">18</dt>
<dd>
<span data-code-cell="annotated-cell-11" data-code-lines="48" data-code-annotation="18">Format tail probability in scientific notation</span>
</dd>
<dt data-target-cell="annotated-cell-11" data-target-annotation="19">19</dt>
<dd>
<span data-code-cell="annotated-cell-11" data-code-lines="49" data-code-annotation="19">Estimate tail probability from <code>MCMC</code> samples</span>
</dd>
</dl>
</div>
<div class="cell-output cell-output-display" data-execution_count="1">
<div>
<figure class="figure">
<p><img src="probability-stats-examples_files/figure-html/coin-flip-bayesian-output-1.svg" id="coin-flip-bayesian" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>The results show excellent agreement between analytical and <code>MCMC</code> approaches. With sufficient samples, <code>MCMC</code> accurately approximates the true posterior distribution. The comparison also illustrates how <code>MCMC</code> precision varies across the distribution—central quantities like the mean are estimated more accurately than tail probabilities.</p>
</section>
<section id="beta-binomial-updating-in-action" class="level3">
<h3 class="anchored" data-anchor-id="beta-binomial-updating-in-action">Beta-Binomial updating in action</h3>
<p>This example demonstrates how Bayesian updating works step by step. We start with a Beta(3,3) prior and observe a sequence of coin flips from a coin with true probability θ = 0.55. Each subplot shows how the posterior distribution evolves as new evidence arrives.</p>
<div id="cell-beta-binomial-updating" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource julia number-lines code-with-copy"><code class="sourceCode julia"><span id="cb3-1"><a href="#cb3-1"></a><span class="kw">function</span> <span class="fu">demonstrate_bayesian_updating</span>()</span>
<span id="cb3-2"><a href="#cb3-2"></a>    <span class="co"># True parameter and prior</span></span>
<span id="cb3-3"><a href="#cb3-3"></a>    true_θ <span class="op">=</span> <span class="fl">0.55</span></span>
<span id="cb3-4"><a href="#cb3-4"></a>    prior_α, prior_β <span class="op">=</span> <span class="fl">3</span>, <span class="fl">3</span></span>
<span id="cb3-5"><a href="#cb3-5"></a></span>
<span id="cb3-6"><a href="#cb3-6"></a>    <span class="co"># Generate sequence of coin flips</span></span>
<span id="cb3-7"><a href="#cb3-7"></a>    <span class="bu">Random</span>.<span class="fu">seed!</span>(<span class="fl">42</span>)</span>
<span id="cb3-8"><a href="#cb3-8"></a>    n_total <span class="op">=</span> <span class="fl">12</span></span>
<span id="cb3-9"><a href="#cb3-9"></a>    flips <span class="op">=</span> <span class="fu">rand</span>(n_total) <span class="op">.&lt;</span> true_θ  <span class="co"># true = heads, false = tails</span></span>
<span id="cb3-10"><a href="#cb3-10"></a></span>
<span id="cb3-11"><a href="#cb3-11"></a>    <span class="co"># Convert to H/T string for display</span></span>
<span id="cb3-12"><a href="#cb3-12"></a>    flip_chars <span class="op">=</span> [f ? <span class="st">"H"</span> <span class="op">:</span> <span class="st">"T"</span> for f <span class="kw">in</span> flips]</span>
<span id="cb3-13"><a href="#cb3-13"></a></span>
<span id="cb3-14"><a href="#cb3-14"></a>    <span class="co"># Stages to show (number of flips observed)</span></span>
<span id="cb3-15"><a href="#cb3-15"></a>    stages <span class="op">=</span> [<span class="fl">1</span>, <span class="fl">3</span>, <span class="fl">5</span>, <span class="fl">8</span>, <span class="fl">12</span>]</span>
<span id="cb3-16"><a href="#cb3-16"></a></span>
<span id="cb3-17"><a href="#cb3-17"></a>    <span class="co"># Create figure with subplots</span></span>
<span id="cb3-18"><a href="#cb3-18"></a>    fig <span class="op">=</span> <span class="fu">Figure</span>(size <span class="op">=</span> (<span class="fl">1000</span>, <span class="fl">600</span>))</span>
<span id="cb3-19"><a href="#cb3-19"></a></span>
<span id="cb3-20"><a href="#cb3-20"></a>    θ_range <span class="op">=</span> <span class="fl">0</span><span class="op">:</span><span class="fl">0.01</span><span class="op">:</span><span class="fl">1</span></span>
<span id="cb3-21"><a href="#cb3-21"></a></span>
<span id="cb3-22"><a href="#cb3-22"></a>    <span class="cf">for</span> (i, stage) <span class="kw">in</span> <span class="fu">enumerate</span>(stages)</span>
<span id="cb3-23"><a href="#cb3-23"></a>        <span class="co"># Calculate posterior parameters</span></span>
<span id="cb3-24"><a href="#cb3-24"></a>        y_obs <span class="op">=</span> <span class="fu">sum</span>(flips[<span class="fl">1</span><span class="op">:</span>stage])  <span class="co"># heads observed so far</span></span>
<span id="cb3-25"><a href="#cb3-25"></a>        n_obs <span class="op">=</span> stage                <span class="co"># total flips observed</span></span>
<span id="cb3-26"><a href="#cb3-26"></a></span>
<span id="cb3-27"><a href="#cb3-27"></a>        post_α <span class="op">=</span> prior_α <span class="op">+</span> y_obs</span>
<span id="cb3-28"><a href="#cb3-28"></a>        post_β <span class="op">=</span> prior_β <span class="op">+</span> n_obs <span class="op">-</span> y_obs</span>
<span id="cb3-29"><a href="#cb3-29"></a></span>
<span id="cb3-30"><a href="#cb3-30"></a>        <span class="co"># Create subplot</span></span>
<span id="cb3-31"><a href="#cb3-31"></a>        row <span class="op">=</span> <span class="fu">div</span>(i <span class="op">-</span> <span class="fl">1</span>, <span class="fl">3</span>) <span class="op">+</span> <span class="fl">1</span></span>
<span id="cb3-32"><a href="#cb3-32"></a>        col <span class="op">=</span> ((i <span class="op">-</span> <span class="fl">1</span>) <span class="op">%</span> <span class="fl">3</span>) <span class="op">+</span> <span class="fl">1</span></span>
<span id="cb3-33"><a href="#cb3-33"></a>        ax <span class="op">=</span> <span class="fu">Axis</span>(fig[row, col],</span>
<span id="cb3-34"><a href="#cb3-34"></a>            xlabel <span class="op">=</span> <span class="st">"θ (probability of heads)"</span>,</span>
<span id="cb3-35"><a href="#cb3-35"></a>            ylabel <span class="op">=</span> <span class="st">"Density"</span>,</span>
<span id="cb3-36"><a href="#cb3-36"></a>            title <span class="op">=</span> <span class="st">"</span><span class="sc">$</span>(<span class="fu">join</span>(flip_chars[<span class="fl">1</span><span class="op">:</span>stage]))<span class="st"> (</span><span class="sc">$</span>(y_obs)<span class="st">/</span><span class="sc">$</span>(n_obs)<span class="st"> heads)"</span>)</span>
<span id="cb3-37"><a href="#cb3-37"></a></span>
<span id="cb3-38"><a href="#cb3-38"></a>        <span class="co"># Plot prior (only for first subplot)</span></span>
<span id="cb3-39"><a href="#cb3-39"></a>        <span class="cf">if</span> i <span class="op">==</span> <span class="fl">1</span></span>
<span id="cb3-40"><a href="#cb3-40"></a>            prior_density <span class="op">=</span> <span class="fu">pdf</span>.(<span class="fu">Beta</span>(prior_α, prior_β), θ_range)</span>
<span id="cb3-41"><a href="#cb3-41"></a>            <span class="fu">lines!</span>(ax, θ_range, prior_density,</span>
<span id="cb3-42"><a href="#cb3-42"></a>                color <span class="op">=</span> <span class="op">:</span>gray, linewidth <span class="op">=</span> <span class="fl">2</span>, linestyle <span class="op">=</span> <span class="op">:</span>dash, label <span class="op">=</span> <span class="st">"Prior Beta(3,3)"</span>)</span>
<span id="cb3-43"><a href="#cb3-43"></a>        <span class="cf">end</span></span>
<span id="cb3-44"><a href="#cb3-44"></a></span>
<span id="cb3-45"><a href="#cb3-45"></a>        <span class="co"># Plot posterior</span></span>
<span id="cb3-46"><a href="#cb3-46"></a>        posterior <span class="op">=</span> <span class="fu">Beta</span>(post_α, post_β)</span>
<span id="cb3-47"><a href="#cb3-47"></a>        posterior_density <span class="op">=</span> <span class="fu">pdf</span>.(posterior, θ_range)</span>
<span id="cb3-48"><a href="#cb3-48"></a>        <span class="fu">lines!</span>(ax, θ_range, posterior_density,</span>
<span id="cb3-49"><a href="#cb3-49"></a>            color <span class="op">=</span> <span class="op">:</span>blue, linewidth <span class="op">=</span> <span class="fl">3</span>, label <span class="op">=</span> <span class="st">"Posterior"</span>)</span>
<span id="cb3-50"><a href="#cb3-50"></a></span>
<span id="cb3-51"><a href="#cb3-51"></a>        <span class="co"># Add vertical lines</span></span>
<span id="cb3-52"><a href="#cb3-52"></a>        <span class="fu">vlines!</span>(ax, [true_θ], color <span class="op">=</span> <span class="op">:</span>green, linewidth <span class="op">=</span> <span class="fl">2</span>, linestyle <span class="op">=</span> <span class="op">:</span>dot,</span>
<span id="cb3-53"><a href="#cb3-53"></a>            label <span class="op">=</span> i <span class="op">==</span> <span class="fl">1</span> ? <span class="st">"True θ = </span><span class="sc">$</span>(true_θ)<span class="st">"</span> <span class="op">:</span> <span class="st">""</span>)</span>
<span id="cb3-54"><a href="#cb3-54"></a></span>
<span id="cb3-55"><a href="#cb3-55"></a>        <span class="co"># MLE estimate</span></span>
<span id="cb3-56"><a href="#cb3-56"></a>        mle_θ <span class="op">=</span> stage <span class="op">==</span> <span class="fl">0</span> ? <span class="fl">0.5</span> <span class="op">:</span> y_obs <span class="op">/</span> n_obs</span>
<span id="cb3-57"><a href="#cb3-57"></a>        <span class="fu">vlines!</span>(ax, [mle_θ], color <span class="op">=</span> <span class="op">:</span>red, linewidth <span class="op">=</span> <span class="fl">2</span>, linestyle <span class="op">=</span> <span class="op">:</span>dashdot,</span>
<span id="cb3-58"><a href="#cb3-58"></a>            label <span class="op">=</span> i <span class="op">==</span> <span class="fl">1</span> ? <span class="st">"MLE"</span> <span class="op">:</span> <span class="st">""</span>)</span>
<span id="cb3-59"><a href="#cb3-59"></a></span>
<span id="cb3-60"><a href="#cb3-60"></a>        <span class="co"># Posterior mean</span></span>
<span id="cb3-61"><a href="#cb3-61"></a>        post_mean <span class="op">=</span> post_α <span class="op">/</span> (post_α <span class="op">+</span> post_β)</span>
<span id="cb3-62"><a href="#cb3-62"></a>        <span class="fu">vlines!</span>(ax, [post_mean], color <span class="op">=</span> <span class="op">:</span>blue, linewidth <span class="op">=</span> <span class="fl">2</span>, alpha <span class="op">=</span> <span class="fl">0.7</span>,</span>
<span id="cb3-63"><a href="#cb3-63"></a>            label <span class="op">=</span> i <span class="op">==</span> <span class="fl">1</span> ? <span class="st">"Posterior mean"</span> <span class="op">:</span> <span class="st">""</span>)</span>
<span id="cb3-64"><a href="#cb3-64"></a></span>
<span id="cb3-65"><a href="#cb3-65"></a>        <span class="co"># Add text with posterior parameters</span></span>
<span id="cb3-66"><a href="#cb3-66"></a>        <span class="fu">text!</span>(ax, <span class="fl">0.05</span>, <span class="fu">maximum</span>(posterior_density) <span class="op">*</span> <span class="fl">0.9</span>,</span>
<span id="cb3-67"><a href="#cb3-67"></a>            text <span class="op">=</span> <span class="st">"Beta(</span><span class="sc">$</span>(post_α)<span class="st">, </span><span class="sc">$</span>(post_β)<span class="st">)"</span>,</span>
<span id="cb3-68"><a href="#cb3-68"></a>            fontsize <span class="op">=</span> <span class="fl">11</span>, color <span class="op">=</span> <span class="op">:</span>blue)</span>
<span id="cb3-69"><a href="#cb3-69"></a></span>
<span id="cb3-70"><a href="#cb3-70"></a>        <span class="co"># Add 95% credible interval</span></span>
<span id="cb3-71"><a href="#cb3-71"></a>        ci_lower <span class="op">=</span> <span class="fu">quantile</span>(posterior, <span class="fl">0.025</span>)</span>
<span id="cb3-72"><a href="#cb3-72"></a>        ci_upper <span class="op">=</span> <span class="fu">quantile</span>(posterior, <span class="fl">0.975</span>)</span>
<span id="cb3-73"><a href="#cb3-73"></a>        <span class="fu">text!</span>(ax, <span class="fl">0.05</span>, <span class="fu">maximum</span>(posterior_density) <span class="op">*</span> <span class="fl">0.75</span>,</span>
<span id="cb3-74"><a href="#cb3-74"></a>            text <span class="op">=</span> <span class="st">"95% CI: [</span><span class="sc">$</span>(<span class="fu">round</span>(ci_lower, digits<span class="op">=</span><span class="fl">2</span>))<span class="st">, </span><span class="sc">$</span>(<span class="fu">round</span>(ci_upper, digits<span class="op">=</span><span class="fl">2</span>))<span class="st">]"</span>,</span>
<span id="cb3-75"><a href="#cb3-75"></a>            fontsize <span class="op">=</span> <span class="fl">10</span>, color <span class="op">=</span> <span class="op">:</span>black)</span>
<span id="cb3-76"><a href="#cb3-76"></a></span>
<span id="cb3-77"><a href="#cb3-77"></a>        <span class="co"># Set consistent x-axis limits</span></span>
<span id="cb3-78"><a href="#cb3-78"></a>        <span class="fu">xlims!</span>(ax, <span class="fl">0</span>, <span class="fl">1</span>)</span>
<span id="cb3-79"><a href="#cb3-79"></a>    <span class="cf">end</span></span>
<span id="cb3-80"><a href="#cb3-80"></a></span>
<span id="cb3-81"><a href="#cb3-81"></a>    <span class="co"># Add legend in the sixth panel (bottom right)</span></span>
<span id="cb3-82"><a href="#cb3-82"></a>    ax_legend <span class="op">=</span> <span class="fu">Axis</span>(fig[<span class="fl">2</span>, <span class="fl">3</span>],</span>
<span id="cb3-83"><a href="#cb3-83"></a>        xlabel <span class="op">=</span> <span class="st">"θ (probability of heads)"</span>,</span>
<span id="cb3-84"><a href="#cb3-84"></a>        ylabel <span class="op">=</span> <span class="st">"Density"</span>)</span>
<span id="cb3-85"><a href="#cb3-85"></a></span>
<span id="cb3-86"><a href="#cb3-86"></a>    <span class="co"># Create dummy plots for legend</span></span>
<span id="cb3-87"><a href="#cb3-87"></a>    <span class="fu">lines!</span>(ax_legend, [<span class="fl">0</span>, <span class="fl">0</span>], [<span class="fl">0</span>, <span class="fl">0</span>], color <span class="op">=</span> <span class="op">:</span>gray, linewidth <span class="op">=</span> <span class="fl">2</span>, linestyle <span class="op">=</span> <span class="op">:</span>dash, label <span class="op">=</span> <span class="st">"Prior Beta(3,3)"</span>)</span>
<span id="cb3-88"><a href="#cb3-88"></a>    <span class="fu">lines!</span>(ax_legend, [<span class="fl">0</span>, <span class="fl">0</span>], [<span class="fl">0</span>, <span class="fl">0</span>], color <span class="op">=</span> <span class="op">:</span>blue, linewidth <span class="op">=</span> <span class="fl">3</span>, label <span class="op">=</span> <span class="st">"Posterior"</span>)</span>
<span id="cb3-89"><a href="#cb3-89"></a>    <span class="fu">vlines!</span>(ax_legend, [<span class="fl">0</span>], color <span class="op">=</span> <span class="op">:</span>green, linewidth <span class="op">=</span> <span class="fl">2</span>, linestyle <span class="op">=</span> <span class="op">:</span>dot, label <span class="op">=</span> <span class="st">"True θ = </span><span class="sc">$</span>(true_θ)<span class="st">"</span>)</span>
<span id="cb3-90"><a href="#cb3-90"></a>    <span class="fu">vlines!</span>(ax_legend, [<span class="fl">0</span>], color <span class="op">=</span> <span class="op">:</span>red, linewidth <span class="op">=</span> <span class="fl">2</span>, linestyle <span class="op">=</span> <span class="op">:</span>dashdot, label <span class="op">=</span> <span class="st">"MLE"</span>)</span>
<span id="cb3-91"><a href="#cb3-91"></a>    <span class="fu">vlines!</span>(ax_legend, [<span class="fl">0</span>], color <span class="op">=</span> <span class="op">:</span>blue, linewidth <span class="op">=</span> <span class="fl">2</span>, alpha <span class="op">=</span> <span class="fl">0.7</span>, label <span class="op">=</span> <span class="st">"Posterior mean"</span>)</span>
<span id="cb3-92"><a href="#cb3-92"></a></span>
<span id="cb3-93"><a href="#cb3-93"></a>    <span class="fu">axislegend</span>(ax_legend, position <span class="op">=</span> <span class="op">:</span>cc, framevisible <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb3-94"><a href="#cb3-94"></a>    <span class="fu">hidespines!</span>(ax_legend)</span>
<span id="cb3-95"><a href="#cb3-95"></a>    <span class="fu">hidedecorations!</span>(ax_legend)</span>
<span id="cb3-96"><a href="#cb3-96"></a></span>
<span id="cb3-97"><a href="#cb3-97"></a>    <span class="co"># Add overall title</span></span>
<span id="cb3-98"><a href="#cb3-98"></a>    <span class="fu">Label</span>(fig[<span class="fl">0</span>, <span class="op">:</span>], <span class="st">"Bayesian Updating: How Posterior Evolves with New Evidence"</span>,</span>
<span id="cb3-99"><a href="#cb3-99"></a>        fontsize <span class="op">=</span> <span class="fl">16</span>, font <span class="op">=</span> <span class="st">"TeX Gyre Heros Bold"</span>)</span>
<span id="cb3-100"><a href="#cb3-100"></a></span>
<span id="cb3-101"><a href="#cb3-101"></a>    <span class="cf">return</span> fig</span>
<span id="cb3-102"><a href="#cb3-102"></a><span class="kw">end</span></span>
<span id="cb3-103"><a href="#cb3-103"></a></span>
<span id="cb3-104"><a href="#cb3-104"></a>fig_updating <span class="op">=</span> <span class="fu">demonstrate_bayesian_updating</span>()</span>
<span id="cb3-105"><a href="#cb3-105"></a>fig_updating</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="1">
<div>
<figure class="figure">
<p><img src="probability-stats-examples_files/figure-html/beta-binomial-updating-output-1.svg" id="beta-binomial-updating" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>This visualization shows several key insights:</p>
<ol type="1">
<li><strong>Prior influence</strong>: Initially, the Beta(3,3) prior (slightly peaked at 0.5) influences our beliefs</li>
<li><strong>Gradual learning</strong>: Each new observation updates the posterior, gradually moving toward the true value</li>
<li><strong>Uncertainty reduction</strong>: The posterior becomes more concentrated (narrower) as we observe more data</li>
<li><strong>Convergence</strong>: The posterior mean and MLE estimate gradually converge toward the true θ = 0.55</li>
<li><strong>Credible intervals</strong>: The 95% credible intervals become narrower, showing reduced uncertainty</li>
</ol>
</section>
</section>
<section id="linear-regression-example" class="level2">
<h2 class="anchored" data-anchor-id="linear-regression-example">Linear regression example</h2>
<p>The coin flip example demonstrated inference for a single parameter using a simple likelihood function. Real-world problems typically involve multiple parameters and more complex relationships between variables. Linear regression provides our first example of multivariate statistical inference while maintaining analytical tractability.</p>
<p>This example demonstrates the progression from curve fitting to probabilistic inference. We’ll examine the same problem through three lenses: minimizing prediction error, maximum likelihood estimation, and Bayesian inference. All three approaches yield the same parameter estimates for the linear model with normal errors, but provide different ways of quantifying uncertainty.</p>
<section id="generating-synthetic-data" class="level3">
<h3 class="anchored" data-anchor-id="generating-synthetic-data">Generating synthetic data</h3>
<p>We start by creating synthetic data from a known linear relationship with added noise. This allows us to verify that our inference methods recover the true parameters.</p>
<div id="linear-regression-data" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="annotated-cell-13"><pre class="sourceCode numberSource julia code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode julia"><span id="annotated-cell-13-1"><a href="#annotated-cell-13-1"></a><span class="co"># Set parameters for data generation</span></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-13" data-target-annotation="1">1</button><span id="annotated-cell-13-2" class="code-annotation-target"><a href="#annotated-cell-13-2"></a><span class="bu">Random</span>.<span class="fu">seed!</span>(<span class="fl">42</span>)</span>
<span id="annotated-cell-13-3"><a href="#annotated-cell-13-3"></a>n_obs <span class="op">=</span> <span class="fl">30</span></span>
<span id="annotated-cell-13-4"><a href="#annotated-cell-13-4"></a>true_intercept <span class="op">=</span> <span class="fl">2.0</span></span>
<span id="annotated-cell-13-5"><a href="#annotated-cell-13-5"></a>true_slope <span class="op">=</span> <span class="fl">1.5</span></span>
<span id="annotated-cell-13-6"><a href="#annotated-cell-13-6"></a>true_sigma <span class="op">=</span> <span class="fl">1.5</span></span>
<span id="annotated-cell-13-7"><a href="#annotated-cell-13-7"></a></span>
<span id="annotated-cell-13-8"><a href="#annotated-cell-13-8"></a><span class="co"># Generate predictor variables</span></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-13" data-target-annotation="2">2</button><span id="annotated-cell-13-9" class="code-annotation-target"><a href="#annotated-cell-13-9"></a>x_reg <span class="op">=</span> <span class="fu">sort</span>(<span class="fu">rand</span>(<span class="fu">Uniform</span>(<span class="fl">0</span>, <span class="fl">5</span>), n_obs))</span>
<span id="annotated-cell-13-10"><a href="#annotated-cell-13-10"></a></span>
<span id="annotated-cell-13-11"><a href="#annotated-cell-13-11"></a><span class="co"># Generate response with linear relationship plus noise</span></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-13" data-target-annotation="3">3</button><span id="annotated-cell-13-12" class="code-annotation-target"><a href="#annotated-cell-13-12"></a>y_reg <span class="op">=</span> true_intercept <span class="op">.+</span> true_slope <span class="op">.*</span> x_reg <span class="op">.+</span> <span class="fu">rand</span>(<span class="fu">Normal</span>(<span class="fl">0</span>, true_sigma), n_obs)</span>
<span id="annotated-cell-13-13"><a href="#annotated-cell-13-13"></a></span>
<span id="annotated-cell-13-14"><a href="#annotated-cell-13-14"></a><span class="co"># Store true parameter values for comparison</span></span>
<span id="annotated-cell-13-15"><a href="#annotated-cell-13-15"></a>true_params <span class="op">=</span> (intercept <span class="op">=</span> true_intercept, slope <span class="op">=</span> true_slope, sigma <span class="op">=</span> true_sigma)</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-annotation">
<dl class="code-annotation-container-hidden code-annotation-container-grid">
<dt data-target-cell="annotated-cell-13" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-13" data-code-lines="2" data-code-annotation="1">Set random seed for reproducible results</span>
</dd>
<dt data-target-cell="annotated-cell-13" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-13" data-code-lines="9" data-code-annotation="2">Generate sorted predictor values from uniform distribution</span>
</dd>
<dt data-target-cell="annotated-cell-13" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-13" data-code-lines="12" data-code-annotation="3">Generate response following <span class="math inline">\(y_i = \alpha + \beta x_i + \epsilon_i\)</span> where <span class="math inline">\(\epsilon_i \sim N(0, \sigma^2)\)</span></span>
</dd>
</dl>
</div>
</div>
<p>Let’s visualize our synthetic dataset:</p>
<div id="cell-linear-regression-data-plot" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode numberSource julia number-lines code-with-copy"><code class="sourceCode julia"><span id="cb4-1"><a href="#cb4-1"></a><span class="kw">function</span> <span class="fu">plot_synthetic_data</span>(x_data, y_data, true_params)</span>
<span id="cb4-2"><a href="#cb4-2"></a>    fig <span class="op">=</span> <span class="fu">Figure</span>(size <span class="op">=</span> (<span class="fl">700</span>, <span class="fl">500</span>))</span>
<span id="cb4-3"><a href="#cb4-3"></a>    ax <span class="op">=</span> <span class="fu">Axis</span>(fig[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb4-4"><a href="#cb4-4"></a>        xlabel <span class="op">=</span> <span class="st">"x"</span>,</span>
<span id="cb4-5"><a href="#cb4-5"></a>        ylabel <span class="op">=</span> <span class="st">"y"</span>,</span>
<span id="cb4-6"><a href="#cb4-6"></a>        title <span class="op">=</span> <span class="st">"Synthetic Linear Regression Dataset"</span>)</span>
<span id="cb4-7"><a href="#cb4-7"></a></span>
<span id="cb4-8"><a href="#cb4-8"></a>    <span class="co"># Plot data points</span></span>
<span id="cb4-9"><a href="#cb4-9"></a>    <span class="fu">scatter!</span>(ax, x_data, y_data, color <span class="op">=</span> <span class="op">:</span>blue, markersize <span class="op">=</span> <span class="fl">8</span>, label <span class="op">=</span> <span class="st">"Observed data"</span>)</span>
<span id="cb4-10"><a href="#cb4-10"></a></span>
<span id="cb4-11"><a href="#cb4-11"></a>    <span class="co"># Plot true relationship</span></span>
<span id="cb4-12"><a href="#cb4-12"></a>    x_line <span class="op">=</span> <span class="fu">range</span>(<span class="fu">minimum</span>(x_data), <span class="fu">maximum</span>(x_data), length <span class="op">=</span> <span class="fl">100</span>)</span>
<span id="cb4-13"><a href="#cb4-13"></a>    y_true <span class="op">=</span> true_params.intercept <span class="op">.+</span> true_params.slope <span class="op">.*</span> x_line</span>
<span id="cb4-14"><a href="#cb4-14"></a>    <span class="fu">lines!</span>(ax, x_line, y_true, color <span class="op">=</span> <span class="op">:</span>red, linewidth <span class="op">=</span> <span class="fl">3</span>,</span>
<span id="cb4-15"><a href="#cb4-15"></a>        label <span class="op">=</span> L<span class="st">"</span><span class="sc">\t</span><span class="st">ext{True relationship: } </span><span class="sc">$</span>y<span class="st"> = %</span><span class="sc">$</span>(true_params.intercept)<span class="st"> + %</span><span class="sc">$</span>(true_params.slope)<span class="st"> x</span><span class="sc">$</span><span class="st">")</span></span>
<span id="cb4-16"><a href="#cb4-16"></a></span>
<span id="cb4-17"><a href="#cb4-17"></a><span class="st">    # </span>Add<span class="st"> noise band to show true sigma</span></span>
<span id="cb4-18"><a href="#cb4-18"></a>    y_upper <span class="op">=</span> y_true <span class="op">.+</span> <span class="fl">2</span> <span class="op">*</span> true_params.sigma</span>
<span id="cb4-19"><a href="#cb4-19"></a>    y_lower <span class="op">=</span> y_true <span class="op">.-</span> <span class="fl">2</span> <span class="op">*</span> true_params.sigma</span>
<span id="cb4-20"><a href="#cb4-20"></a>    <span class="fu">band!</span>(ax, x_line, y_lower, y_upper, color <span class="op">=</span> (<span class="op">:</span>red, <span class="fl">0.2</span>),</span>
<span id="cb4-21"><a href="#cb4-21"></a>        label <span class="op">=</span> L<span class="st">"</span><span class="sc">$</span><span class="st">\</span>pm<span class="st"> 2 \sigma</span><span class="sc">$</span><span class="st">")</span></span>
<span id="cb4-22"><a href="#cb4-22"></a></span>
<span id="cb4-23"><a href="#cb4-23"></a><span class="st">    </span>axislegend<span class="st">(ax, position = :lt)</span></span>
<span id="cb4-24"><a href="#cb4-24"></a>    <span class="cf">return</span> fig</span>
<span id="cb4-25"><a href="#cb4-25"></a><span class="kw">end</span></span>
<span id="cb4-26"><a href="#cb4-26"></a></span>
<span id="cb4-27"><a href="#cb4-27"></a>fig_data <span class="op">=</span> <span class="fu">plot_synthetic_data</span>(x_reg, y_reg, true_params)</span>
<span id="cb4-28"><a href="#cb4-28"></a>fig_data</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="1">
<div>
<figure class="figure">
<p><img src="probability-stats-examples_files/figure-html/linear-regression-data-plot-output-1.svg" id="linear-regression-data-plot" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>This visualization shows our synthetic dataset with the true linear relationship and noise level. The scattered points represent our “observations,” while the red line shows the true underlying relationship we’re trying to recover. The shaded band indicates the expected range of observations given the noise level (<span class="math inline">\(\pm 2\sigma\)</span>).</p>
</section>
<section id="curve-fitting-by-minimizing-prediction-error" class="level3">
<h3 class="anchored" data-anchor-id="curve-fitting-by-minimizing-prediction-error">Curve fitting by minimizing prediction error</h3>
<p>Without probability theory, we can view regression as finding the line that best fits the data. This requires choosing a loss function to measure fit quality. The most common choice is mean squared error (<code>MSE</code>), which penalizes large prediction errors more than small ones. This is more of a “machine learning” perspective on the problem, and is one to which we return in <a href="../chapters/fundamentals/ml-nonparametric.html">the ML chapter</a>.</p>
<div id="cell-linear-regression-curve-fitting" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="annotated-cell-15"><pre class="sourceCode numberSource julia code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode julia"><button class="code-annotation-anchor" data-target-cell="annotated-cell-15" data-target-annotation="1">1</button><span id="annotated-cell-15-1" class="code-annotation-target"><a href="#annotated-cell-15-1"></a><span class="kw">function</span> <span class="fu">fit_least_squares</span>(x_data, y_data)</span>
<span id="annotated-cell-15-2"><a href="#annotated-cell-15-2"></a>    n <span class="op">=</span> <span class="fu">length</span>(x_data)</span>
<span id="annotated-cell-15-3"><a href="#annotated-cell-15-3"></a>    x_bar <span class="op">=</span> <span class="fu">mean</span>(x_data)</span>
<span id="annotated-cell-15-4"><a href="#annotated-cell-15-4"></a>    y_bar <span class="op">=</span> <span class="fu">mean</span>(y_data)</span>
<span id="annotated-cell-15-5"><a href="#annotated-cell-15-5"></a></span>
<span id="annotated-cell-15-6"><a href="#annotated-cell-15-6"></a>    <span class="co"># Analytical least squares formulas</span></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-15" data-target-annotation="2">2</button><span id="annotated-cell-15-7" class="code-annotation-target"><a href="#annotated-cell-15-7"></a>    slope <span class="op">=</span> <span class="fu">sum</span>((x_data <span class="op">.-</span> x_bar) <span class="op">.*</span> (y_data <span class="op">.-</span> y_bar)) <span class="op">/</span> <span class="fu">sum</span>((x_data <span class="op">.-</span> x_bar) <span class="op">.^</span> <span class="fl">2</span>)</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-15" data-target-annotation="3">3</button><span id="annotated-cell-15-8" class="code-annotation-target"><a href="#annotated-cell-15-8"></a>    intercept <span class="op">=</span> y_bar <span class="op">-</span> slope <span class="op">*</span> x_bar</span>
<span id="annotated-cell-15-9"><a href="#annotated-cell-15-9"></a></span>
<span id="annotated-cell-15-10"><a href="#annotated-cell-15-10"></a>    <span class="co"># Calculate residuals and MSE</span></span>
<span id="annotated-cell-15-11"><a href="#annotated-cell-15-11"></a>    y_pred <span class="op">=</span> intercept <span class="op">.+</span> slope <span class="op">.*</span> x_data</span>
<span id="annotated-cell-15-12"><a href="#annotated-cell-15-12"></a>    residuals <span class="op">=</span> y_data <span class="op">.-</span> y_pred</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-15" data-target-annotation="4">4</button><span id="annotated-cell-15-13" class="code-annotation-target"><a href="#annotated-cell-15-13"></a>    mse <span class="op">=</span> <span class="fu">mean</span>(residuals <span class="op">.^</span> <span class="fl">2</span>)</span>
<span id="annotated-cell-15-14"><a href="#annotated-cell-15-14"></a></span>
<span id="annotated-cell-15-15"><a href="#annotated-cell-15-15"></a>    <span class="cf">return</span> (intercept <span class="op">=</span> intercept, slope <span class="op">=</span> slope, mse <span class="op">=</span> mse, residuals <span class="op">=</span> residuals, predictions <span class="op">=</span> y_pred)</span>
<span id="annotated-cell-15-16"><a href="#annotated-cell-15-16"></a><span class="kw">end</span></span>
<span id="annotated-cell-15-17"><a href="#annotated-cell-15-17"></a></span>
<span id="annotated-cell-15-18"><a href="#annotated-cell-15-18"></a><span class="kw">function</span> <span class="fu">plot_least_squares_fit</span>(x_data, y_data, fit_results)</span>
<span id="annotated-cell-15-19"><a href="#annotated-cell-15-19"></a>    fig <span class="op">=</span> <span class="fu">Figure</span>(size <span class="op">=</span> (<span class="fl">800</span>, <span class="fl">500</span>))</span>
<span id="annotated-cell-15-20"><a href="#annotated-cell-15-20"></a>    ax <span class="op">=</span> <span class="fu">Axis</span>(fig[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="annotated-cell-15-21"><a href="#annotated-cell-15-21"></a>        xlabel <span class="op">=</span> <span class="st">"x"</span>,</span>
<span id="annotated-cell-15-22"><a href="#annotated-cell-15-22"></a>        ylabel <span class="op">=</span> <span class="st">"y"</span>,</span>
<span id="annotated-cell-15-23"><a href="#annotated-cell-15-23"></a>        title <span class="op">=</span> <span class="st">"Least Squares Curve Fitting"</span>)</span>
<span id="annotated-cell-15-24"><a href="#annotated-cell-15-24"></a></span>
<span id="annotated-cell-15-25"><a href="#annotated-cell-15-25"></a>    <span class="co"># Plot data points</span></span>
<span id="annotated-cell-15-26"><a href="#annotated-cell-15-26"></a>    <span class="fu">scatter!</span>(ax, x_data, y_data, color <span class="op">=</span> <span class="op">:</span>blue, markersize <span class="op">=</span> <span class="fl">8</span>, label <span class="op">=</span> <span class="st">"Data"</span>)</span>
<span id="annotated-cell-15-27"><a href="#annotated-cell-15-27"></a></span>
<span id="annotated-cell-15-28"><a href="#annotated-cell-15-28"></a>    <span class="co"># Plot fitted line</span></span>
<span id="annotated-cell-15-29"><a href="#annotated-cell-15-29"></a>    x_line <span class="op">=</span> <span class="fu">range</span>(<span class="fu">minimum</span>(x_data), <span class="fu">maximum</span>(x_data), length <span class="op">=</span> <span class="fl">100</span>)</span>
<span id="annotated-cell-15-30"><a href="#annotated-cell-15-30"></a>    y_line <span class="op">=</span> fit_results.intercept <span class="op">.+</span> fit_results.slope <span class="op">.*</span> x_line</span>
<span id="annotated-cell-15-31"><a href="#annotated-cell-15-31"></a>    <span class="fu">lines!</span>(ax, x_line, y_line, color <span class="op">=</span> <span class="op">:</span>red, linewidth <span class="op">=</span> <span class="fl">2</span>,</span>
<span id="annotated-cell-15-32"><a href="#annotated-cell-15-32"></a>        label <span class="op">=</span> L<span class="st">"</span><span class="sc">\t</span><span class="st">ext{Fitted line:} </span><span class="sc">$</span>y<span class="st"> = %</span><span class="sc">$</span>(<span class="fu">round</span>(fit_results.intercept, digits<span class="op">=</span><span class="fl">2</span>))<span class="st"> + %</span><span class="sc">$</span>(<span class="fu">round</span>(fit_results.slope, digits<span class="op">=</span><span class="fl">2</span>))<span class="st"> x</span><span class="sc">$</span><span class="st">")</span></span>
<span id="annotated-cell-15-33"><a href="#annotated-cell-15-33"></a></span>
<span id="annotated-cell-15-34"><a href="#annotated-cell-15-34"></a><span class="st">    # </span>Show<span class="st"> residuals as vertical lines</span></span>
<span id="annotated-cell-15-35"><a href="#annotated-cell-15-35"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(x_data)</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-15" data-target-annotation="5">5</button><span id="annotated-cell-15-36" class="code-annotation-target"><a href="#annotated-cell-15-36"></a>        <span class="fu">lines!</span>(ax, [x_data[i], x_data[i]], [y_data[i], fit_results.predictions[i]],</span>
<span id="annotated-cell-15-37"><a href="#annotated-cell-15-37"></a>            color <span class="op">=</span> <span class="op">:</span>gray, linestyle <span class="op">=</span> <span class="op">:</span>dash, alpha <span class="op">=</span> <span class="fl">0.7</span>, linewidth <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="annotated-cell-15-38"><a href="#annotated-cell-15-38"></a>    <span class="cf">end</span></span>
<span id="annotated-cell-15-39"><a href="#annotated-cell-15-39"></a></span>
<span id="annotated-cell-15-40"><a href="#annotated-cell-15-40"></a>    <span class="co"># Add true line for comparison</span></span>
<span id="annotated-cell-15-41"><a href="#annotated-cell-15-41"></a>    y_true_line <span class="op">=</span> true_params.intercept <span class="op">.+</span> true_params.slope <span class="op">.*</span> x_line</span>
<span id="annotated-cell-15-42"><a href="#annotated-cell-15-42"></a>    <span class="fu">lines!</span>(ax, x_line, y_true_line, color <span class="op">=</span> <span class="op">:</span>green, linewidth <span class="op">=</span> <span class="fl">2</span>, linestyle <span class="op">=</span> <span class="op">:</span>dot,</span>
<span id="annotated-cell-15-43"><a href="#annotated-cell-15-43"></a>        label <span class="op">=</span> L<span class="st">"</span><span class="sc">\t</span><span class="st">ext{True line: } </span><span class="sc">$</span>y<span class="st"> = %</span><span class="sc">$</span>(true_params.intercept)<span class="st"> + %</span><span class="sc">$</span>(true_params.slope)<span class="st">x</span><span class="sc">$</span><span class="st">")</span></span>
<span id="annotated-cell-15-44"><a href="#annotated-cell-15-44"></a></span>
<span id="annotated-cell-15-45"><a href="#annotated-cell-15-45"></a><span class="st">    </span>axislegend<span class="st">(ax, position = :lt)</span></span>
<span id="annotated-cell-15-46"><a href="#annotated-cell-15-46"></a>    <span class="cf">return</span> fig</span>
<span id="annotated-cell-15-47"><a href="#annotated-cell-15-47"></a><span class="kw">end</span></span>
<span id="annotated-cell-15-48"><a href="#annotated-cell-15-48"></a></span>
<span id="annotated-cell-15-49"><a href="#annotated-cell-15-49"></a><span class="co"># Fit the model and create visualization</span></span>
<span id="annotated-cell-15-50"><a href="#annotated-cell-15-50"></a>ols_fit <span class="op">=</span> <span class="fu">fit_least_squares</span>(x_reg, y_reg)</span>
<span id="annotated-cell-15-51"><a href="#annotated-cell-15-51"></a>fig_curve_fit <span class="op">=</span> <span class="fu">plot_least_squares_fit</span>(x_reg, y_reg, ols_fit)</span>
<span id="annotated-cell-15-52"><a href="#annotated-cell-15-52"></a>fig_curve_fit</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-annotation">
<dl class="code-annotation-container-hidden code-annotation-container-grid">
<dt data-target-cell="annotated-cell-15" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-15" data-code-lines="1" data-code-annotation="1">Calculate least squares estimates using analytical formulas</span>
</dd>
<dt data-target-cell="annotated-cell-15" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-15" data-code-lines="7" data-code-annotation="2">Slope: <span class="math inline">\(\hat{\beta} = \frac{\sum (x_i - \bar{x})(y_i - \bar{y})}{\sum (x_i - \bar{x})^2}\)</span></span>
</dd>
<dt data-target-cell="annotated-cell-15" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-15" data-code-lines="8" data-code-annotation="3">Intercept: <span class="math inline">\(\hat{\alpha} = \bar{y} - \hat{\beta} \bar{x}\)</span> (fitted line passes through centroid)</span>
</dd>
<dt data-target-cell="annotated-cell-15" data-target-annotation="4">4</dt>
<dd>
<span data-code-cell="annotated-cell-15" data-code-lines="13" data-code-annotation="4">Mean squared error quantifies overall fit quality</span>
</dd>
<dt data-target-cell="annotated-cell-15" data-target-annotation="5">5</dt>
<dd>
<span data-code-cell="annotated-cell-15" data-code-lines="36" data-code-annotation="5">Residual lines show prediction errors for each observation</span>
</dd>
</dl>
</div>
<div class="cell-output cell-output-display" data-execution_count="1">
<div>
<figure class="figure">
<p><img src="probability-stats-examples_files/figure-html/linear-regression-curve-fitting-output-1.svg" id="linear-regression-curve-fitting" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>The least squares method finds parameter estimates close to the true values. The residual lines visualize the model’s prediction errors, with shorter lines indicating better fit.</p>
</section>
<section id="maximum-likelihood-estimation-1" class="level3">
<h3 class="anchored" data-anchor-id="maximum-likelihood-estimation-1">Maximum likelihood estimation</h3>
<p>The curve fitting approach works but provides no way to quantify uncertainty in our parameter estimates. By introducing a probabilistic model, we can use maximum likelihood estimation to both estimate parameters and assess their precision.</p>
<p>For linear regression, we assume the response follows a normal distribution around the linear predictor:</p>
<p><span class="math display">\[
y_i \sim N(\alpha + \beta x_i, \sigma^2)
\]</span></p>
<p>as discussed in the theory section. Using the <code>Optim</code> package, we can numerically maximize the likelihood function to find the <code>MLE</code> estimates for <span class="math inline">\(\alpha\)</span>, <span class="math inline">\(\beta\)</span>, and <span class="math inline">\(\sigma\)</span>. In <code>Optim</code>, all functions are minimized, so we minimize the negative log-likelihood instead.</p>
<div id="linear-regression-mle" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="annotated-cell-16"><pre class="sourceCode numberSource julia code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode julia"><button class="code-annotation-anchor" data-target-cell="annotated-cell-16" data-target-annotation="1">1</button><span id="annotated-cell-16-1" class="code-annotation-target"><a href="#annotated-cell-16-1"></a><span class="kw">function</span> <span class="fu">negative_log_likelihood</span>(params, x_data, y_data)</span>
<span id="annotated-cell-16-2"><a href="#annotated-cell-16-2"></a>    α, β, σ <span class="op">=</span> params</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-16" data-target-annotation="2">2</button><span id="annotated-cell-16-3" class="code-annotation-target"><a href="#annotated-cell-16-3"></a>    <span class="cf">if</span> σ <span class="op">&lt;=</span> <span class="fl">0</span></span>
<span id="annotated-cell-16-4"><a href="#annotated-cell-16-4"></a>        <span class="cf">return</span> <span class="cn">Inf</span></span>
<span id="annotated-cell-16-5"><a href="#annotated-cell-16-5"></a>    <span class="cf">end</span></span>
<span id="annotated-cell-16-6"><a href="#annotated-cell-16-6"></a></span>
<span id="annotated-cell-16-7"><a href="#annotated-cell-16-7"></a>    <span class="co"># Calculate predicted values</span></span>
<span id="annotated-cell-16-8"><a href="#annotated-cell-16-8"></a>    μ <span class="op">=</span> α <span class="op">.+</span> β <span class="op">.*</span> x_data</span>
<span id="annotated-cell-16-9"><a href="#annotated-cell-16-9"></a></span>
<span id="annotated-cell-16-10"><a href="#annotated-cell-16-10"></a>    <span class="co"># Sum of negative log-likelihoods for each observation</span></span>
<span id="annotated-cell-16-11"><a href="#annotated-cell-16-11"></a>    nll <span class="op">=</span> <span class="fl">0.0</span></span>
<span id="annotated-cell-16-12"><a href="#annotated-cell-16-12"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(y_data)</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-16" data-target-annotation="3">3</button><span id="annotated-cell-16-13" class="code-annotation-target"><a href="#annotated-cell-16-13"></a>        nll <span class="op">+=</span> <span class="fu">-logpdf</span>(<span class="fu">Normal</span>(μ[i], σ), y_data[i])</span>
<span id="annotated-cell-16-14"><a href="#annotated-cell-16-14"></a>    <span class="cf">end</span></span>
<span id="annotated-cell-16-15"><a href="#annotated-cell-16-15"></a>    <span class="cf">return</span> nll</span>
<span id="annotated-cell-16-16"><a href="#annotated-cell-16-16"></a><span class="kw">end</span></span>
<span id="annotated-cell-16-17"><a href="#annotated-cell-16-17"></a></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-16" data-target-annotation="4">4</button><span id="annotated-cell-16-18" class="code-annotation-target"><a href="#annotated-cell-16-18"></a><span class="kw">function</span> <span class="fu">find_mle_estimates</span>(x_data, y_data; initial_guess <span class="op">=</span> [<span class="fl">0.0</span>, <span class="fl">1.0</span>, <span class="fl">1.0</span>])</span>
<span id="annotated-cell-16-19"><a href="#annotated-cell-16-19"></a>    <span class="co"># Use optimization to find MLE</span></span>
<span id="annotated-cell-16-20"><a href="#annotated-cell-16-20"></a>    result <span class="op">=</span> <span class="fu">optimize</span>(params <span class="op">-&gt;</span> <span class="fu">negative_log_likelihood</span>(params, x_data, y_data),</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-16" data-target-annotation="5">5</button><span id="annotated-cell-16-21" class="code-annotation-target"><a href="#annotated-cell-16-21"></a>        initial_guess, <span class="fu">NelderMead</span>())</span>
<span id="annotated-cell-16-22"><a href="#annotated-cell-16-22"></a></span>
<span id="annotated-cell-16-23"><a href="#annotated-cell-16-23"></a>    <span class="cf">if</span> !Optim.<span class="fu">converged</span>(result)</span>
<span id="annotated-cell-16-24"><a href="#annotated-cell-16-24"></a>        <span class="pp">@warn</span> <span class="st">"MLE optimization did not converge"</span></span>
<span id="annotated-cell-16-25"><a href="#annotated-cell-16-25"></a>    <span class="cf">end</span></span>
<span id="annotated-cell-16-26"><a href="#annotated-cell-16-26"></a></span>
<span id="annotated-cell-16-27"><a href="#annotated-cell-16-27"></a>    mle_params <span class="op">=</span> Optim.<span class="fu">minimizer</span>(result)</span>
<span id="annotated-cell-16-28"><a href="#annotated-cell-16-28"></a>    <span class="cf">return</span> (intercept <span class="op">=</span> mle_params[<span class="fl">1</span>], slope <span class="op">=</span> mle_params[<span class="fl">2</span>], sigma <span class="op">=</span> mle_params[<span class="fl">3</span>])</span>
<span id="annotated-cell-16-29"><a href="#annotated-cell-16-29"></a><span class="kw">end</span></span>
<span id="annotated-cell-16-30"><a href="#annotated-cell-16-30"></a></span>
<span id="annotated-cell-16-31"><a href="#annotated-cell-16-31"></a><span class="co"># Find MLE estimates</span></span>
<span id="annotated-cell-16-32"><a href="#annotated-cell-16-32"></a>mle_estimates <span class="op">=</span> <span class="fu">find_mle_estimates</span>(x_reg, y_reg)</span>
<span id="annotated-cell-16-33"><a href="#annotated-cell-16-33"></a></span>
<span id="annotated-cell-16-34"><a href="#annotated-cell-16-34"></a><span class="fu">println</span>(<span class="st">"MLE Estimates:"</span>)</span>
<span id="annotated-cell-16-35"><a href="#annotated-cell-16-35"></a><span class="fu">println</span>(<span class="st">"  Intercept: </span><span class="sc">$</span>(<span class="fu">round</span>(mle_estimates.intercept, digits<span class="op">=</span><span class="fl">3</span>))<span class="st">"</span>)</span>
<span id="annotated-cell-16-36"><a href="#annotated-cell-16-36"></a><span class="fu">println</span>(<span class="st">"  Slope: </span><span class="sc">$</span>(<span class="fu">round</span>(mle_estimates.slope, digits<span class="op">=</span><span class="fl">3</span>))<span class="st">"</span>)</span>
<span id="annotated-cell-16-37"><a href="#annotated-cell-16-37"></a><span class="fu">println</span>(<span class="st">"  Sigma: </span><span class="sc">$</span>(<span class="fu">round</span>(mle_estimates.sigma, digits<span class="op">=</span><span class="fl">3</span>))<span class="st">"</span>)</span>
<span id="annotated-cell-16-38"><a href="#annotated-cell-16-38"></a><span class="fu">println</span>()</span>
<span id="annotated-cell-16-39"><a href="#annotated-cell-16-39"></a><span class="fu">println</span>(<span class="st">"True Parameters:"</span>)</span>
<span id="annotated-cell-16-40"><a href="#annotated-cell-16-40"></a><span class="fu">println</span>(<span class="st">"  Intercept: </span><span class="sc">$</span>(true_params.intercept)<span class="st">"</span>)</span>
<span id="annotated-cell-16-41"><a href="#annotated-cell-16-41"></a><span class="fu">println</span>(<span class="st">"  Slope: </span><span class="sc">$</span>(true_params.slope)<span class="st">"</span>)</span>
<span id="annotated-cell-16-42"><a href="#annotated-cell-16-42"></a><span class="fu">println</span>(<span class="st">"  Sigma: </span><span class="sc">$</span>(true_params.sigma)<span class="st">"</span>)</span>
<span id="annotated-cell-16-43"><a href="#annotated-cell-16-43"></a><span class="fu">println</span>()</span>
<span id="annotated-cell-16-44"><a href="#annotated-cell-16-44"></a><span class="fu">println</span>(<span class="st">"Least Squares Estimates (for comparison):"</span>)</span>
<span id="annotated-cell-16-45"><a href="#annotated-cell-16-45"></a><span class="fu">println</span>(<span class="st">"  Intercept: </span><span class="sc">$</span>(<span class="fu">round</span>(ols_fit.intercept, digits<span class="op">=</span><span class="fl">3</span>))<span class="st">"</span>)</span>
<span id="annotated-cell-16-46"><a href="#annotated-cell-16-46"></a><span class="fu">println</span>(<span class="st">"  Slope: </span><span class="sc">$</span>(<span class="fu">round</span>(ols_fit.slope, digits<span class="op">=</span><span class="fl">3</span>))<span class="st">"</span>)</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-annotation">
<dl class="code-annotation-container-hidden code-annotation-container-grid">
<dt data-target-cell="annotated-cell-16" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-16" data-code-lines="1" data-code-annotation="1">Negative log-likelihood function for optimization (we minimize this)</span>
</dd>
<dt data-target-cell="annotated-cell-16" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-16" data-code-lines="3" data-code-annotation="2">Enforce constraint that <span class="math inline">\(\sigma &gt; 0\)</span> by returning infinite cost for invalid values</span>
</dd>
<dt data-target-cell="annotated-cell-16" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-16" data-code-lines="13" data-code-annotation="3">Sum negative log-likelihoods: <span class="math inline">\(-\sum_i \log p(y_i | \alpha + \beta x_i, \sigma)\)</span></span>
</dd>
<dt data-target-cell="annotated-cell-16" data-target-annotation="4">4</dt>
<dd>
<span data-code-cell="annotated-cell-16" data-code-lines="18" data-code-annotation="4">Function to find <code>MLE</code> using numerical optimization</span>
</dd>
<dt data-target-cell="annotated-cell-16" data-target-annotation="5">5</dt>
<dd>
<span data-code-cell="annotated-cell-16" data-code-lines="21" data-code-annotation="5">Use Nelder-Mead simplex algorithm for unconstrained optimization</span>
</dd>
</dl>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>MLE Estimates:
  Intercept: 1.851
  Slope: 1.584
  Sigma: 1.492

True Parameters:
  Intercept: 2.0
  Slope: 1.5
  Sigma: 1.5

Least Squares Estimates (for comparison):
  Intercept: 1.851
  Slope: 1.584</code></pre>
</div>
</div>
<p>The <code>MLE</code> estimates for the intercept and slope match the least squares results exactly. This confirms the theoretical result that for linear regression with normal errors, maximum likelihood estimation is equivalent to minimizing mean squared error. The <code>MLE</code> approach additionally estimates <span class="math inline">\(\sigma\)</span>, quantifying the noise level in the data.</p>
</section>
<section id="bayesian-inference" class="level3">
<h3 class="anchored" data-anchor-id="bayesian-inference">Bayesian inference</h3>
<p>Bayesian inference treats all parameters as random variables with probability distributions. We specify prior beliefs, observe data, and compute posterior distributions that quantify parameter uncertainty.</p>
<div id="cell-linear-regression-bayesian" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="annotated-cell-17"><pre class="sourceCode numberSource julia code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode julia"><span id="annotated-cell-17-1"><a href="#annotated-cell-17-1"></a><span class="co"># Define Turing.jl model for Bayesian linear regression</span></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-17" data-target-annotation="1">1</button><span id="annotated-cell-17-2" class="code-annotation-target"><a href="#annotated-cell-17-2"></a><span class="pp">@model</span> <span class="kw">function</span> <span class="fu">linear_regression_model</span>(x, y)</span>
<span id="annotated-cell-17-3"><a href="#annotated-cell-17-3"></a>    n <span class="op">=</span> <span class="fu">length</span>(y)</span>
<span id="annotated-cell-17-4"><a href="#annotated-cell-17-4"></a></span>
<span id="annotated-cell-17-5"><a href="#annotated-cell-17-5"></a>    <span class="co"># Priors</span></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-17" data-target-annotation="2">2</button><span id="annotated-cell-17-6" class="code-annotation-target"><a href="#annotated-cell-17-6"></a>    α <span class="op">~</span> <span class="fu">Normal</span>(<span class="fl">0</span>, <span class="fl">5</span>)</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-17" data-target-annotation="3">3</button><span id="annotated-cell-17-7" class="code-annotation-target"><a href="#annotated-cell-17-7"></a>    β <span class="op">~</span> <span class="fu">Normal</span>(<span class="fl">0</span>, <span class="fl">5</span>)</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-17" data-target-annotation="4">4</button><span id="annotated-cell-17-8" class="code-annotation-target"><a href="#annotated-cell-17-8"></a>    σ <span class="op">~</span> <span class="fu">Exponential</span>(<span class="fl">1</span>)</span>
<span id="annotated-cell-17-9"><a href="#annotated-cell-17-9"></a></span>
<span id="annotated-cell-17-10"><a href="#annotated-cell-17-10"></a>    <span class="co"># Likelihood</span></span>
<span id="annotated-cell-17-11"><a href="#annotated-cell-17-11"></a>    μ <span class="op">=</span> α <span class="op">.+</span> β <span class="op">.*</span> x</span>
<span id="annotated-cell-17-12"><a href="#annotated-cell-17-12"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span>n</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-17" data-target-annotation="5">5</button><span id="annotated-cell-17-13" class="code-annotation-target"><a href="#annotated-cell-17-13"></a>        y[i] <span class="op">~</span> <span class="fu">Normal</span>(μ[i], σ)</span>
<span id="annotated-cell-17-14"><a href="#annotated-cell-17-14"></a>    <span class="cf">end</span></span>
<span id="annotated-cell-17-15"><a href="#annotated-cell-17-15"></a><span class="kw">end</span></span>
<span id="annotated-cell-17-16"><a href="#annotated-cell-17-16"></a></span>
<span id="annotated-cell-17-17"><a href="#annotated-cell-17-17"></a><span class="kw">function</span> <span class="fu">run_bayesian_regression</span>(x_data, y_data; n_samples <span class="op">=</span> <span class="fl">5000</span>)</span>
<span id="annotated-cell-17-18"><a href="#annotated-cell-17-18"></a>    <span class="co"># Create model instance</span></span>
<span id="annotated-cell-17-19"><a href="#annotated-cell-17-19"></a>    model <span class="op">=</span> <span class="fu">linear_regression_model</span>(x_data, y_data)</span>
<span id="annotated-cell-17-20"><a href="#annotated-cell-17-20"></a></span>
<span id="annotated-cell-17-21"><a href="#annotated-cell-17-21"></a>    <span class="co"># Sample from posterior using NUTS</span></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-17" data-target-annotation="6">6</button><span id="annotated-cell-17-22" class="code-annotation-target"><a href="#annotated-cell-17-22"></a>    chain <span class="op">=</span> <span class="fu">sample</span>(model, <span class="fu">NUTS</span>(), n_samples, verbose <span class="op">=</span> <span class="cn">false</span>, progress <span class="op">=</span> <span class="cn">false</span>)</span>
<span id="annotated-cell-17-23"><a href="#annotated-cell-17-23"></a></span>
<span id="annotated-cell-17-24"><a href="#annotated-cell-17-24"></a>    <span class="cf">return</span> chain</span>
<span id="annotated-cell-17-25"><a href="#annotated-cell-17-25"></a><span class="kw">end</span></span>
<span id="annotated-cell-17-26"><a href="#annotated-cell-17-26"></a></span>
<span id="annotated-cell-17-27"><a href="#annotated-cell-17-27"></a><span class="co"># Run Bayesian inference  </span></span>
<span id="annotated-cell-17-28"><a href="#annotated-cell-17-28"></a>bayesian_chain <span class="op">=</span> <span class="fu">run_bayesian_regression</span>(x_reg, y_reg)</span>
<span id="annotated-cell-17-29"><a href="#annotated-cell-17-29"></a></span>
<span id="annotated-cell-17-30"><a href="#annotated-cell-17-30"></a><span class="co"># Extract posterior samples</span></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-17" data-target-annotation="7">7</button><span id="annotated-cell-17-31" class="code-annotation-target"><a href="#annotated-cell-17-31"></a>α_samples <span class="op">=</span> <span class="fu">vec</span>(<span class="fu">Array</span>(bayesian_chain[<span class="op">:</span>α]))</span>
<span id="annotated-cell-17-32"><a href="#annotated-cell-17-32"></a>β_samples <span class="op">=</span> <span class="fu">vec</span>(<span class="fu">Array</span>(bayesian_chain[<span class="op">:</span>β]))</span>
<span id="annotated-cell-17-33"><a href="#annotated-cell-17-33"></a>σ_samples <span class="op">=</span> <span class="fu">vec</span>(<span class="fu">Array</span>(bayesian_chain[<span class="op">:</span>σ]))</span>
<span id="annotated-cell-17-34"><a href="#annotated-cell-17-34"></a></span>
<span id="annotated-cell-17-35"><a href="#annotated-cell-17-35"></a><span class="co"># Calculate posterior summaries</span></span>
<span id="annotated-cell-17-36"><a href="#annotated-cell-17-36"></a>bayesian_estimates <span class="op">=</span> (</span>
<span id="annotated-cell-17-37"><a href="#annotated-cell-17-37"></a>    intercept <span class="op">=</span> (mean <span class="op">=</span> <span class="fu">mean</span>(α_samples), std <span class="op">=</span> <span class="fu">std</span>(α_samples)),</span>
<span id="annotated-cell-17-38"><a href="#annotated-cell-17-38"></a>    slope <span class="op">=</span> (mean <span class="op">=</span> <span class="fu">mean</span>(β_samples), std <span class="op">=</span> <span class="fu">std</span>(β_samples)),</span>
<span id="annotated-cell-17-39"><a href="#annotated-cell-17-39"></a>    sigma <span class="op">=</span> (mean <span class="op">=</span> <span class="fu">mean</span>(σ_samples), std <span class="op">=</span> <span class="fu">std</span>(σ_samples)),</span>
<span id="annotated-cell-17-40"><a href="#annotated-cell-17-40"></a>)</span>
<span id="annotated-cell-17-41"><a href="#annotated-cell-17-41"></a></span>
<span id="annotated-cell-17-42"><a href="#annotated-cell-17-42"></a><span class="co"># Create visualization of posterior distributions and uncertainty</span></span>
<span id="annotated-cell-17-43"><a href="#annotated-cell-17-43"></a>fig <span class="op">=</span> <span class="fu">Figure</span>(size <span class="op">=</span> (<span class="fl">800</span>, <span class="fl">600</span>))</span>
<span id="annotated-cell-17-44"><a href="#annotated-cell-17-44"></a></span>
<span id="annotated-cell-17-45"><a href="#annotated-cell-17-45"></a><span class="co"># Plot 1: Posterior distributions</span></span>
<span id="annotated-cell-17-46"><a href="#annotated-cell-17-46"></a>ax1 <span class="op">=</span> <span class="fu">Axis</span>(fig[<span class="fl">1</span>, <span class="fl">1</span>], title <span class="op">=</span> <span class="st">"Posterior Distributions"</span>, xlabel <span class="op">=</span> <span class="st">"Parameter Value"</span>, ylabel <span class="op">=</span> <span class="st">"Density"</span>)</span>
<span id="annotated-cell-17-47"><a href="#annotated-cell-17-47"></a><span class="fu">hist!</span>(ax1, α_samples, bins <span class="op">=</span> <span class="fl">50</span>, color <span class="op">=</span> (<span class="op">:</span>blue, <span class="fl">0.7</span>), label <span class="op">=</span> <span class="st">"Intercept (α)"</span>)</span>
<span id="annotated-cell-17-48"><a href="#annotated-cell-17-48"></a><span class="fu">hist!</span>(ax1, β_samples, bins <span class="op">=</span> <span class="fl">50</span>, color <span class="op">=</span> (<span class="op">:</span>red, <span class="fl">0.7</span>), label <span class="op">=</span> <span class="st">"Slope (β)"</span>)</span>
<span id="annotated-cell-17-49"><a href="#annotated-cell-17-49"></a><span class="fu">axislegend</span>(ax1)</span>
<span id="annotated-cell-17-50"><a href="#annotated-cell-17-50"></a></span>
<span id="annotated-cell-17-51"><a href="#annotated-cell-17-51"></a><span class="co"># Plot 2: Data with uncertainty bands</span></span>
<span id="annotated-cell-17-52"><a href="#annotated-cell-17-52"></a>ax2 <span class="op">=</span> <span class="fu">Axis</span>(fig[<span class="fl">1</span>, <span class="fl">2</span>], title <span class="op">=</span> <span class="st">"Regression with Uncertainty"</span>, xlabel <span class="op">=</span> <span class="st">"x"</span>, ylabel <span class="op">=</span> <span class="st">"y"</span>)</span>
<span id="annotated-cell-17-53"><a href="#annotated-cell-17-53"></a></span>
<span id="annotated-cell-17-54"><a href="#annotated-cell-17-54"></a><span class="co"># Plot data points</span></span>
<span id="annotated-cell-17-55"><a href="#annotated-cell-17-55"></a><span class="fu">scatter!</span>(ax2, x_reg, y_reg, color <span class="op">=</span> <span class="op">:</span>black, markersize <span class="op">=</span> <span class="fl">8</span>, label <span class="op">=</span> <span class="st">"Data"</span>)</span>
<span id="annotated-cell-17-56"><a href="#annotated-cell-17-56"></a></span>
<span id="annotated-cell-17-57"><a href="#annotated-cell-17-57"></a><span class="co"># Generate prediction lines from posterior samples</span></span>
<span id="annotated-cell-17-58"><a href="#annotated-cell-17-58"></a>x_pred <span class="op">=</span> <span class="fu">range</span>(<span class="fu">minimum</span>(x_reg), <span class="fu">maximum</span>(x_reg), length <span class="op">=</span> <span class="fl">100</span>)</span>
<span id="annotated-cell-17-59"><a href="#annotated-cell-17-59"></a>n_lines <span class="op">=</span> <span class="fl">100</span>  <span class="co"># Number of posterior samples to show</span></span>
<span id="annotated-cell-17-60"><a href="#annotated-cell-17-60"></a>sample_indices <span class="op">=</span> <span class="fu">sample</span>(<span class="fl">1</span><span class="op">:</span><span class="fu">length</span>(α_samples), n_lines, replace <span class="op">=</span> <span class="cn">false</span>)</span>
<span id="annotated-cell-17-61"><a href="#annotated-cell-17-61"></a></span>
<span id="annotated-cell-17-62"><a href="#annotated-cell-17-62"></a><span class="cf">for</span> i <span class="kw">in</span> sample_indices[<span class="fl">1</span><span class="op">:</span><span class="fu">min</span>(<span class="fl">50</span>, <span class="fu">length</span>(sample_indices))]  <span class="co"># Show subset for clarity</span></span>
<span id="annotated-cell-17-63"><a href="#annotated-cell-17-63"></a>    y_pred <span class="op">=</span> α_samples[i] <span class="op">.+</span> β_samples[i] <span class="op">.*</span> x_pred</span>
<span id="annotated-cell-17-64"><a href="#annotated-cell-17-64"></a>    <span class="fu">lines!</span>(ax2, x_pred, y_pred, color <span class="op">=</span> (<span class="op">:</span>blue, <span class="fl">0.1</span>), linewidth <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="annotated-cell-17-65"><a href="#annotated-cell-17-65"></a><span class="cf">end</span></span>
<span id="annotated-cell-17-66"><a href="#annotated-cell-17-66"></a></span>
<span id="annotated-cell-17-67"><a href="#annotated-cell-17-67"></a><span class="co"># Add posterior mean line</span></span>
<span id="annotated-cell-17-68"><a href="#annotated-cell-17-68"></a>α_mean <span class="op">=</span> <span class="fu">mean</span>(α_samples)</span>
<span id="annotated-cell-17-69"><a href="#annotated-cell-17-69"></a>β_mean <span class="op">=</span> <span class="fu">mean</span>(β_samples)</span>
<span id="annotated-cell-17-70"><a href="#annotated-cell-17-70"></a>y_mean <span class="op">=</span> α_mean <span class="op">.+</span> β_mean <span class="op">.*</span> x_pred</span>
<span id="annotated-cell-17-71"><a href="#annotated-cell-17-71"></a><span class="fu">lines!</span>(ax2, x_pred, y_mean, color <span class="op">=</span> <span class="op">:</span>red, linewidth <span class="op">=</span> <span class="fl">3</span>, label <span class="op">=</span> <span class="st">"Posterior Mean"</span>)</span>
<span id="annotated-cell-17-72"><a href="#annotated-cell-17-72"></a></span>
<span id="annotated-cell-17-73"><a href="#annotated-cell-17-73"></a><span class="co"># Add MLE line for comparison (need to get MLE estimates first)</span></span>
<span id="annotated-cell-17-74"><a href="#annotated-cell-17-74"></a>mle_estimates <span class="op">=</span> <span class="fu">find_mle_estimates</span>(x_reg, y_reg)</span>
<span id="annotated-cell-17-75"><a href="#annotated-cell-17-75"></a><span class="fu">lines!</span>(ax2, x_pred, mle_estimates.intercept <span class="op">.+</span> mle_estimates.slope <span class="op">.*</span> x_pred, color <span class="op">=</span> <span class="op">:</span>orange, linewidth <span class="op">=</span> <span class="fl">3</span>, linestyle <span class="op">=</span> <span class="op">:</span>dash, label <span class="op">=</span> <span class="st">"MLE"</span>)</span>
<span id="annotated-cell-17-76"><a href="#annotated-cell-17-76"></a></span>
<span id="annotated-cell-17-77"><a href="#annotated-cell-17-77"></a><span class="fu">axislegend</span>(ax2, position <span class="op">=</span> <span class="op">:</span>lt)</span>
<span id="annotated-cell-17-78"><a href="#annotated-cell-17-78"></a></span>
<span id="annotated-cell-17-79"><a href="#annotated-cell-17-79"></a>fig</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-annotation">
<dl class="code-annotation-container-hidden code-annotation-container-grid">
<dt data-target-cell="annotated-cell-17" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-17" data-code-lines="2" data-code-annotation="1">Turing.jl probabilistic programming model using <code>@model</code> macro</span>
</dd>
<dt data-target-cell="annotated-cell-17" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-17" data-code-lines="6" data-code-annotation="2">Weakly informative prior for intercept: <span class="math inline">\(\alpha \sim N(0, 5^2)\)</span></span>
</dd>
<dt data-target-cell="annotated-cell-17" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-17" data-code-lines="7" data-code-annotation="3">Weakly informative prior for slope: <span class="math inline">\(\beta \sim N(0, 5^2)\)</span></span>
</dd>
<dt data-target-cell="annotated-cell-17" data-target-annotation="4">4</dt>
<dd>
<span data-code-cell="annotated-cell-17" data-code-lines="8" data-code-annotation="4">Exponential prior for noise standard deviation: <span class="math inline">\(\sigma \sim \text{Exp}(1)\)</span></span>
</dd>
<dt data-target-cell="annotated-cell-17" data-target-annotation="5">5</dt>
<dd>
<span data-code-cell="annotated-cell-17" data-code-lines="13" data-code-annotation="5">Likelihood specification: each <span class="math inline">\(y_i \sim N(\alpha + \beta x_i, \sigma)\)</span></span>
</dd>
<dt data-target-cell="annotated-cell-17" data-target-annotation="6">6</dt>
<dd>
<span data-code-cell="annotated-cell-17" data-code-lines="22" data-code-annotation="6">Sample from posterior using No-U-Turn Sampler (<code>NUTS</code>)</span>
</dd>
<dt data-target-cell="annotated-cell-17" data-target-annotation="7">7</dt>
<dd>
<span data-code-cell="annotated-cell-17" data-code-lines="31" data-code-annotation="7">Extract parameter samples from MCMC chain and flatten to vectors</span>
</dd>
</dl>
</div>
<div class="cell-output cell-output-stdout">
<div class="ansi-escaped-output">
<pre><span class="ansi-cyan-fg ansi-bold">┌ </span><span class="ansi-cyan-fg ansi-bold">Info: </span>Found initial step size
<span class="ansi-cyan-fg ansi-bold">└ </span>  ϵ = 0.0125
</pre>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="1">
<div>
<figure class="figure">
<p><img src="probability-stats-examples_files/figure-html/linear-regression-bayesian-output-2.svg" id="linear-regression-bayesian" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="mle-with-turing" class="level3">
<h3 class="anchored" data-anchor-id="mle-with-turing">MLE with Turing</h3>
<p>We can also use <code>Turing.jl</code> to find the <code>MLE</code> by optimizing the model instead of sampling. This demonstrates how the same probabilistic model can be used for both optimization-based and sampling-based inference.</p>
<div id="linear-regression-mle-turing" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="annotated-cell-18"><pre class="sourceCode numberSource julia code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode julia"><span id="annotated-cell-18-1"><a href="#annotated-cell-18-1"></a><span class="co"># Use the same Turing model but optimize instead of sample</span></span>
<span id="annotated-cell-18-2"><a href="#annotated-cell-18-2"></a><span class="kw">function</span> <span class="fu">find_turing_mle</span>(x_data, y_data)</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-18" data-target-annotation="1">1</button><span id="annotated-cell-18-3" class="code-annotation-target"><a href="#annotated-cell-18-3"></a>    model <span class="op">=</span> <span class="fu">linear_regression_model</span>(x_data, y_data)</span>
<span id="annotated-cell-18-4"><a href="#annotated-cell-18-4"></a></span>
<span id="annotated-cell-18-5"><a href="#annotated-cell-18-5"></a>    <span class="co"># Find MAP estimate using optimization</span></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-18" data-target-annotation="2">2</button><span id="annotated-cell-18-6" class="code-annotation-target"><a href="#annotated-cell-18-6"></a>    mle_result <span class="op">=</span> <span class="fu">optimize</span>(model, <span class="fu">MAP</span>())</span>
<span id="annotated-cell-18-7"><a href="#annotated-cell-18-7"></a></span>
<span id="annotated-cell-18-8"><a href="#annotated-cell-18-8"></a>    <span class="co"># Extract parameter estimates</span></span>
<span id="annotated-cell-18-9"><a href="#annotated-cell-18-9"></a>    estimates <span class="op">=</span> (</span>
<span id="annotated-cell-18-10"><a href="#annotated-cell-18-10"></a>        intercept <span class="op">=</span> mle_result.values[<span class="op">:</span>α],</span>
<span id="annotated-cell-18-11"><a href="#annotated-cell-18-11"></a>        slope <span class="op">=</span> mle_result.values[<span class="op">:</span>β],</span>
<span id="annotated-cell-18-12"><a href="#annotated-cell-18-12"></a>        sigma <span class="op">=</span> mle_result.values[<span class="op">:</span>σ],</span>
<span id="annotated-cell-18-13"><a href="#annotated-cell-18-13"></a>    )</span>
<span id="annotated-cell-18-14"><a href="#annotated-cell-18-14"></a></span>
<span id="annotated-cell-18-15"><a href="#annotated-cell-18-15"></a>    <span class="cf">return</span> estimates</span>
<span id="annotated-cell-18-16"><a href="#annotated-cell-18-16"></a><span class="kw">end</span></span>
<span id="annotated-cell-18-17"><a href="#annotated-cell-18-17"></a></span>
<span id="annotated-cell-18-18"><a href="#annotated-cell-18-18"></a><span class="co"># Find MLE using Turing optimization</span></span>
<span id="annotated-cell-18-19"><a href="#annotated-cell-18-19"></a>turing_mle <span class="op">=</span> <span class="fu">find_turing_mle</span>(x_reg, y_reg)</span>
<span id="annotated-cell-18-20"><a href="#annotated-cell-18-20"></a></span>
<span id="annotated-cell-18-21"><a href="#annotated-cell-18-21"></a><span class="fu">println</span>(<span class="st">"Turing MLE Estimates:"</span>)</span>
<span id="annotated-cell-18-22"><a href="#annotated-cell-18-22"></a><span class="fu">println</span>(<span class="st">"  Intercept: </span><span class="sc">$</span>(<span class="fu">round</span>(turing_mle.intercept, digits<span class="op">=</span><span class="fl">3</span>))<span class="st">"</span>)</span>
<span id="annotated-cell-18-23"><a href="#annotated-cell-18-23"></a><span class="fu">println</span>(<span class="st">"  Slope: </span><span class="sc">$</span>(<span class="fu">round</span>(turing_mle.slope, digits<span class="op">=</span><span class="fl">3</span>))<span class="st">"</span>)</span>
<span id="annotated-cell-18-24"><a href="#annotated-cell-18-24"></a><span class="fu">println</span>(<span class="st">"  Sigma: </span><span class="sc">$</span>(<span class="fu">round</span>(turing_mle.sigma, digits<span class="op">=</span><span class="fl">3</span>))<span class="st">"</span>)</span>
<span id="annotated-cell-18-25"><a href="#annotated-cell-18-25"></a><span class="fu">println</span>()</span>
<span id="annotated-cell-18-26"><a href="#annotated-cell-18-26"></a><span class="fu">println</span>(<span class="st">"Comparison with Manual MLE:"</span>)</span>
<span id="annotated-cell-18-27"><a href="#annotated-cell-18-27"></a><span class="fu">println</span>(<span class="st">"  Intercept difference: </span><span class="sc">$</span>(<span class="fu">round</span>(<span class="fu">abs</span>(turing_mle.intercept <span class="op">-</span> mle_estimates.intercept), digits<span class="op">=</span><span class="fl">6</span>))<span class="st">"</span>)</span>
<span id="annotated-cell-18-28"><a href="#annotated-cell-18-28"></a><span class="fu">println</span>(<span class="st">"  Slope difference: </span><span class="sc">$</span>(<span class="fu">round</span>(<span class="fu">abs</span>(turing_mle.slope <span class="op">-</span> mle_estimates.slope), digits<span class="op">=</span><span class="fl">6</span>))<span class="st">"</span>)</span>
<span id="annotated-cell-18-29"><a href="#annotated-cell-18-29"></a><span class="fu">println</span>(<span class="st">"  Sigma difference: </span><span class="sc">$</span>(<span class="fu">round</span>(<span class="fu">abs</span>(turing_mle.sigma <span class="op">-</span> mle_estimates.sigma), digits<span class="op">=</span><span class="fl">6</span>))<span class="st">"</span>)</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-annotation">
<dl class="code-annotation-container-hidden code-annotation-container-grid">
<dt data-target-cell="annotated-cell-18" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-18" data-code-lines="3" data-code-annotation="1">Use the same probabilistic model defined earlier for Bayesian inference</span>
</dd>
<dt data-target-cell="annotated-cell-18" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-18" data-code-lines="6" data-code-annotation="2">Optimize using <code>MAP()</code> (Maximum A Posteriori) which reduces to <code>MLE</code> with flat priors</span>
</dd>
</dl>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Turing MLE Estimates:
  Intercept: 1.838
  Slope: 1.587
  Sigma: 1.457

Comparison with Manual MLE:
  Intercept difference: 0.013192
  Slope difference: 0.00346
  Sigma difference: 0.03495</code></pre>
</div>
</div>
<p>This approach shows that <code>Turing.jl</code> provides a unified framework: the same model definition can be used for both <code>MLE</code> optimization and <code>MCMC</code> sampling. The results should match our manual <code>MLE</code> implementation, demonstrating the consistency between different computational approaches, although small numerical differences may arise due to optimization tolerances.</p>
</section>
<section id="comparison-of-approaches" class="level3">
<h3 class="anchored" data-anchor-id="comparison-of-approaches">Comparison of approaches</h3>
<p>All three approaches provide similar point estimates, but differ in their treatment of uncertainty:</p>
<div id="cell-linear-regression-comparison" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="annotated-cell-19"><pre class="sourceCode numberSource julia code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode julia"><span id="annotated-cell-19-1"><a href="#annotated-cell-19-1"></a><span class="kw">function</span> <span class="fu">create_regression_comparison_plot</span>()</span>
<span id="annotated-cell-19-2"><a href="#annotated-cell-19-2"></a>    fig <span class="op">=</span> <span class="fu">Figure</span>(size <span class="op">=</span> (<span class="fl">1200</span>, <span class="fl">800</span>))</span>
<span id="annotated-cell-19-3"><a href="#annotated-cell-19-3"></a></span>
<span id="annotated-cell-19-4"><a href="#annotated-cell-19-4"></a>    <span class="co"># Data and fits subplot</span></span>
<span id="annotated-cell-19-5"><a href="#annotated-cell-19-5"></a>    ax1 <span class="op">=</span> <span class="fu">Axis</span>(fig[<span class="fl">1</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span>],</span>
<span id="annotated-cell-19-6"><a href="#annotated-cell-19-6"></a>        xlabel <span class="op">=</span> <span class="st">"x"</span>,</span>
<span id="annotated-cell-19-7"><a href="#annotated-cell-19-7"></a>        ylabel <span class="op">=</span> <span class="st">"y"</span>,</span>
<span id="annotated-cell-19-8"><a href="#annotated-cell-19-8"></a>        title <span class="op">=</span> <span class="st">"Regression Results: All Methods"</span>)</span>
<span id="annotated-cell-19-9"><a href="#annotated-cell-19-9"></a></span>
<span id="annotated-cell-19-10"><a href="#annotated-cell-19-10"></a>    <span class="co"># Plot data</span></span>
<span id="annotated-cell-19-11"><a href="#annotated-cell-19-11"></a>    <span class="fu">scatter!</span>(ax1, x_reg, y_reg, color <span class="op">=</span> <span class="op">:</span>black, markersize <span class="op">=</span> <span class="fl">6</span>, label <span class="op">=</span> <span class="st">"Data"</span>)</span>
<span id="annotated-cell-19-12"><a href="#annotated-cell-19-12"></a></span>
<span id="annotated-cell-19-13"><a href="#annotated-cell-19-13"></a>    <span class="co"># Plot true relationship</span></span>
<span id="annotated-cell-19-14"><a href="#annotated-cell-19-14"></a>    x_line <span class="op">=</span> <span class="fu">range</span>(<span class="fu">minimum</span>(x_reg), <span class="fu">maximum</span>(x_reg), length <span class="op">=</span> <span class="fl">100</span>)</span>
<span id="annotated-cell-19-15"><a href="#annotated-cell-19-15"></a>    y_true <span class="op">=</span> true_params.intercept <span class="op">.+</span> true_params.slope <span class="op">.*</span> x_line</span>
<span id="annotated-cell-19-16"><a href="#annotated-cell-19-16"></a>    <span class="fu">lines!</span>(ax1, x_line, y_true, color <span class="op">=</span> <span class="op">:</span>green, linewidth <span class="op">=</span> <span class="fl">3</span>,</span>
<span id="annotated-cell-19-17"><a href="#annotated-cell-19-17"></a>        label <span class="op">=</span> <span class="st">"True: y = </span><span class="sc">$</span>(true_params.intercept)<span class="st"> + </span><span class="sc">$</span>(true_params.slope)<span class="st">x"</span>)</span>
<span id="annotated-cell-19-18"><a href="#annotated-cell-19-18"></a></span>
<span id="annotated-cell-19-19"><a href="#annotated-cell-19-19"></a>    <span class="co"># Plot least squares fit</span></span>
<span id="annotated-cell-19-20"><a href="#annotated-cell-19-20"></a>    y_ols <span class="op">=</span> ols_fit.intercept <span class="op">.+</span> ols_fit.slope <span class="op">.*</span> x_line</span>
<span id="annotated-cell-19-21"><a href="#annotated-cell-19-21"></a>    <span class="fu">lines!</span>(ax1, x_line, y_ols, color <span class="op">=</span> <span class="op">:</span>blue, linewidth <span class="op">=</span> <span class="fl">2</span>,</span>
<span id="annotated-cell-19-22"><a href="#annotated-cell-19-22"></a>        label <span class="op">=</span> <span class="st">"Least squares"</span>)</span>
<span id="annotated-cell-19-23"><a href="#annotated-cell-19-23"></a></span>
<span id="annotated-cell-19-24"><a href="#annotated-cell-19-24"></a>    <span class="co"># Plot MLE fit  </span></span>
<span id="annotated-cell-19-25"><a href="#annotated-cell-19-25"></a>    y_mle <span class="op">=</span> mle_estimates.intercept <span class="op">.+</span> mle_estimates.slope <span class="op">.*</span> x_line</span>
<span id="annotated-cell-19-26"><a href="#annotated-cell-19-26"></a>    <span class="fu">lines!</span>(ax1, x_line, y_mle, color <span class="op">=</span> <span class="op">:</span>red, linewidth <span class="op">=</span> <span class="fl">2</span>, linestyle <span class="op">=</span> <span class="op">:</span>dash,</span>
<span id="annotated-cell-19-27"><a href="#annotated-cell-19-27"></a>        label <span class="op">=</span> <span class="st">"MLE"</span>)</span>
<span id="annotated-cell-19-28"><a href="#annotated-cell-19-28"></a></span>
<span id="annotated-cell-19-29"><a href="#annotated-cell-19-29"></a>    <span class="co"># Plot Bayesian posterior mean and credible interval</span></span>
<span id="annotated-cell-19-30"><a href="#annotated-cell-19-30"></a>    y_bayes <span class="op">=</span> bayesian_estimates.intercept.mean <span class="op">.+</span> bayesian_estimates.slope.mean <span class="op">.*</span> x_line</span>
<span id="annotated-cell-19-31"><a href="#annotated-cell-19-31"></a>    <span class="fu">lines!</span>(ax1, x_line, y_bayes, color <span class="op">=</span> <span class="op">:</span>purple, linewidth <span class="op">=</span> <span class="fl">2</span>, linestyle <span class="op">=</span> <span class="op">:</span>dot,</span>
<span id="annotated-cell-19-32"><a href="#annotated-cell-19-32"></a>        label <span class="op">=</span> <span class="st">"Bayesian mean"</span>)</span>
<span id="annotated-cell-19-33"><a href="#annotated-cell-19-33"></a></span>
<span id="annotated-cell-19-34"><a href="#annotated-cell-19-34"></a>    <span class="co"># Calculate 95% credible interval for predictions</span></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-19" data-target-annotation="1">1</button><span id="annotated-cell-19-35" class="code-annotation-target"><a href="#annotated-cell-19-35"></a>    n_pred_samples <span class="op">=</span> <span class="fu">min</span>(<span class="fl">1000</span>, <span class="fu">length</span>(α_samples))</span>
<span id="annotated-cell-19-36"><a href="#annotated-cell-19-36"></a>    y_pred_samples <span class="op">=</span> <span class="fu">zeros</span>(<span class="fu">length</span>(x_line), n_pred_samples)</span>
<span id="annotated-cell-19-37"><a href="#annotated-cell-19-37"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span>n_pred_samples</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-19" data-target-annotation="2">2</button><span id="annotated-cell-19-38" class="code-annotation-target"><a href="#annotated-cell-19-38"></a>        y_pred_samples[<span class="op">:</span>, i] <span class="op">=</span> α_samples[i] <span class="op">.+</span> β_samples[i] <span class="op">.*</span> x_line</span>
<span id="annotated-cell-19-39"><a href="#annotated-cell-19-39"></a>    <span class="cf">end</span></span>
<span id="annotated-cell-19-40"><a href="#annotated-cell-19-40"></a></span>
<span id="annotated-cell-19-41"><a href="#annotated-cell-19-41"></a>    <span class="co"># Calculate percentiles for credible band</span></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-19" data-target-annotation="3">3</button><span id="annotated-cell-19-42" class="code-annotation-target"><a href="#annotated-cell-19-42"></a>    y_lower <span class="op">=</span> [<span class="fu">quantile</span>(y_pred_samples[i, <span class="op">:</span>], <span class="fl">0.025</span>) for i <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu">length</span>(x_line)]</span>
<span id="annotated-cell-19-43"><a href="#annotated-cell-19-43"></a>    y_upper <span class="op">=</span> [<span class="fu">quantile</span>(y_pred_samples[i, <span class="op">:</span>], <span class="fl">0.975</span>) for i <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu">length</span>(x_line)]</span>
<span id="annotated-cell-19-44"><a href="#annotated-cell-19-44"></a></span>
<span id="annotated-cell-19-45"><a href="#annotated-cell-19-45"></a>    <span class="co"># Add credible interval band</span></span>
<span id="annotated-cell-19-46"><a href="#annotated-cell-19-46"></a>    <span class="fu">band!</span>(ax1, x_line, y_lower, y_upper, color <span class="op">=</span> (<span class="op">:</span>purple, <span class="fl">0.2</span>),</span>
<span id="annotated-cell-19-47"><a href="#annotated-cell-19-47"></a>        label <span class="op">=</span> <span class="st">"95% credible interval"</span>)</span>
<span id="annotated-cell-19-48"><a href="#annotated-cell-19-48"></a></span>
<span id="annotated-cell-19-49"><a href="#annotated-cell-19-49"></a>    <span class="fu">axislegend</span>(ax1, position <span class="op">=</span> <span class="op">:</span>lt)</span>
<span id="annotated-cell-19-50"><a href="#annotated-cell-19-50"></a></span>
<span id="annotated-cell-19-51"><a href="#annotated-cell-19-51"></a>    <span class="co"># Posterior distributions subplot</span></span>
<span id="annotated-cell-19-52"><a href="#annotated-cell-19-52"></a>    ax2 <span class="op">=</span> <span class="fu">Axis</span>(fig[<span class="fl">2</span>, <span class="fl">1</span>],</span>
<span id="annotated-cell-19-53"><a href="#annotated-cell-19-53"></a>        xlabel <span class="op">=</span> L<span class="st">"Intercept </span><span class="sc">$</span><span class="st">\</span>alpha<span class="sc">$</span><span class="st">",</span></span>
<span id="annotated-cell-19-54"><a href="#annotated-cell-19-54"></a><span class="st">        </span>ylabel<span class="st"> = "</span>Density<span class="st">",</span></span>
<span id="annotated-cell-19-55"><a href="#annotated-cell-19-55"></a>        title <span class="op">=</span> <span class="st">"Parameter Uncertainty"</span>)</span>
<span id="annotated-cell-19-56"><a href="#annotated-cell-19-56"></a></span>
<span id="annotated-cell-19-57"><a href="#annotated-cell-19-57"></a>    <span class="fu">hist!</span>(ax2, <span class="fu">vec</span>(α_samples), bins <span class="op">=</span> <span class="fl">30</span>, normalization <span class="op">=</span> <span class="op">:</span>pdf,</span>
<span id="annotated-cell-19-58"><a href="#annotated-cell-19-58"></a>        color <span class="op">=</span> (<span class="op">:</span>purple, <span class="fl">0.6</span>), label <span class="op">=</span> <span class="st">"Posterior"</span>)</span>
<span id="annotated-cell-19-59"><a href="#annotated-cell-19-59"></a>    <span class="fu">vlines!</span>(ax2, [true_params.intercept], color <span class="op">=</span> <span class="op">:</span>green, linewidth <span class="op">=</span> <span class="fl">2</span>,</span>
<span id="annotated-cell-19-60"><a href="#annotated-cell-19-60"></a>        label <span class="op">=</span> <span class="st">"True value"</span>)</span>
<span id="annotated-cell-19-61"><a href="#annotated-cell-19-61"></a>    <span class="fu">vlines!</span>(ax2, [mle_estimates.intercept], color <span class="op">=</span> <span class="op">:</span>red, linewidth <span class="op">=</span> <span class="fl">2</span>,</span>
<span id="annotated-cell-19-62"><a href="#annotated-cell-19-62"></a>        label <span class="op">=</span> <span class="st">"MLE"</span>)</span>
<span id="annotated-cell-19-63"><a href="#annotated-cell-19-63"></a>    <span class="fu">axislegend</span>(ax2, position <span class="op">=</span> <span class="op">:</span>rt)</span>
<span id="annotated-cell-19-64"><a href="#annotated-cell-19-64"></a></span>
<span id="annotated-cell-19-65"><a href="#annotated-cell-19-65"></a>    ax3 <span class="op">=</span> <span class="fu">Axis</span>(fig[<span class="fl">2</span>, <span class="fl">2</span>],</span>
<span id="annotated-cell-19-66"><a href="#annotated-cell-19-66"></a>        xlabel <span class="op">=</span> L<span class="st">"Slope </span><span class="sc">$</span><span class="st">\</span>beta<span class="sc">$</span><span class="st">",</span></span>
<span id="annotated-cell-19-67"><a href="#annotated-cell-19-67"></a><span class="st">        </span>ylabel<span class="st"> = "</span>Density<span class="st">")</span></span>
<span id="annotated-cell-19-68"><a href="#annotated-cell-19-68"></a></span>
<span id="annotated-cell-19-69"><a href="#annotated-cell-19-69"></a>    <span class="fu">hist!</span>(ax3, <span class="fu">vec</span>(β_samples), bins <span class="op">=</span> <span class="fl">30</span>, normalization <span class="op">=</span> <span class="op">:</span>pdf,</span>
<span id="annotated-cell-19-70"><a href="#annotated-cell-19-70"></a>        color <span class="op">=</span> (<span class="op">:</span>purple, <span class="fl">0.6</span>))</span>
<span id="annotated-cell-19-71"><a href="#annotated-cell-19-71"></a>    <span class="fu">vlines!</span>(ax3, [true_params.slope], color <span class="op">=</span> <span class="op">:</span>green, linewidth <span class="op">=</span> <span class="fl">2</span>)</span>
<span id="annotated-cell-19-72"><a href="#annotated-cell-19-72"></a>    <span class="fu">vlines!</span>(ax3, [mle_estimates.slope], color <span class="op">=</span> <span class="op">:</span>red, linewidth <span class="op">=</span> <span class="fl">2</span>)</span>
<span id="annotated-cell-19-73"><a href="#annotated-cell-19-73"></a></span>
<span id="annotated-cell-19-74"><a href="#annotated-cell-19-74"></a>    <span class="cf">return</span> fig</span>
<span id="annotated-cell-19-75"><a href="#annotated-cell-19-75"></a><span class="kw">end</span></span>
<span id="annotated-cell-19-76"><a href="#annotated-cell-19-76"></a></span>
<span id="annotated-cell-19-77"><a href="#annotated-cell-19-77"></a>fig_comparison <span class="op">=</span> <span class="fu">create_regression_comparison_plot</span>()</span>
<span id="annotated-cell-19-78"><a href="#annotated-cell-19-78"></a>fig_comparison</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-annotation">
<dl class="code-annotation-container-hidden code-annotation-container-grid">
<dt data-target-cell="annotated-cell-19" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-19" data-code-lines="35" data-code-annotation="1">Use subset of MCMC samples for computational efficiency</span>
</dd>
<dt data-target-cell="annotated-cell-19" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-19" data-code-lines="38" data-code-annotation="2">For each x-value, calculate predicted y using different parameter samples</span>
</dd>
<dt data-target-cell="annotated-cell-19" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-19" data-code-lines="42" data-code-annotation="3">Calculate 2.5th and 97.5th percentiles for 95% credible interval</span>
</dd>
</dl>
</div>
<div class="cell-output cell-output-display" data-execution_count="1">
<div>
<figure class="figure">
<p><img src="probability-stats-examples_files/figure-html/linear-regression-comparison-output-1.svg" id="linear-regression-comparison" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>This comparison reveals several important points:</p>
<ul>
<li><strong>Point estimates agree</strong>: All methods produce nearly identical parameter estimates</li>
<li><strong>Uncertainty quantification differs</strong>: Only the Bayesian approach provides full uncertainty distributions and credible intervals</li>
<li><strong>Prediction uncertainty</strong>: The 95% credible interval shows the range of plausible regression lines given parameter uncertainty</li>
<li><strong>Computational cost varies</strong>: Least squares is fastest, <code>MLE</code> requires optimization, <code>MCMC</code> is most expensive</li>
<li><strong>Interpretability</strong>: Bayesian posteriors directly quantify “how certain should we be about these parameter values?”</li>
</ul>
<p>The Bayesian approach excels when we need to propagate parameter uncertainty through subsequent calculations or decision processes. The credible interval demonstrates how parameter uncertainty translates into prediction uncertainty—a crucial consideration for risk assessment and decision-making under uncertainty.</p>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        const annoteTargets = window.document.querySelectorAll('.code-annotation-anchor');
        for (let i=0; i<annoteTargets.length; i++) {
          const annoteTarget = annoteTargets[i];
          const targetCell = annoteTarget.getAttribute("data-target-cell");
          const targetAnnotation = annoteTarget.getAttribute("data-target-annotation");
          const contentFn = () => {
            const content = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
            if (content) {
              const tipContent = content.cloneNode(true);
              tipContent.classList.add("code-annotation-tip-content");
              return tipContent.outerHTML;
            }
          }
          const config = {
            allowHTML: true,
            content: contentFn,
            onShow: (instance) => {
              selectCodeLines(instance.reference);
              instance.reference.classList.add('code-annotation-active');
              window.tippy.hideAll();
            },
            onHide: (instance) => {
              unselectCodeLines();
              instance.reference.classList.remove('code-annotation-active');
            },
            maxWidth: 300,
            delay: [50, 0],
            duration: [200, 0],
            offset: [5, 10],
            arrow: true,
            appendTo: function(el) {
              return el.parentElement.parentElement.parentElement;
            },
            interactive: true,
            interactiveBorder: 10,
            theme: 'quarto',
            placement: 'right',
            popperOptions: {
              modifiers: [
              {
                name: 'flip',
                options: {
                  flipVariations: false, // true by default
                  allowedAutoPlacements: ['right'],
                  fallbackPlacements: ['right', 'top', 'top-start', 'top-end', 'bottom', 'bottom-start', 'bottom-end', 'left'],
                },
              },
              {
                name: 'preventOverflow',
                options: {
                  mainAxis: false,
                  altAxis: false
                }
              }
              ]        
            }      
          };
          window.tippy(annoteTarget, config); 
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../notebooks/index.html" class="pagination-link" aria-label="Overview">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Overview</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../notebooks/extremes-examples.html" class="pagination-link" aria-label="Extreme Value Theory Examples ✏️">
        <span class="nav-page-text">Extreme Value Theory Examples ✏️</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p><span class="faux-block"><i class="fa-brands fa-creative-commons" aria-label="creative-commons"></i> 2024–2025 <a href="https://orcid.org/0000-0002-3428-2224">James Doss-Gollin</a></span> <span class="faux-block"></span></p>
</div>   
    <div class="nav-footer-center">
<p>All content licensed under <a href="../chapters/about/license.html"><i class="fa-brands fa-creative-commons" aria-label="creative-commons"></i> <i class="fa-brands fa-creative-commons-by" aria-label="creative-commons-by"></i> Creative Commons CC BY-NC 4.0</a></p>
</div>
    <div class="nav-footer-right">
<p><span class="faux-block">Made with Julia and <a href="https://quarto.org/">Quarto</a>.</span> <span class="faux-block"><a href="https://github.com/dossgollin-lab/climate-risk-management-reference">View the source on <i class="fa-brands fa-github" aria-label="github"></i> GitHub</a></span></p>
</div>
  </div>
</footer>




</body></html>